---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```
PCA (http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/118-principal-component-analysis-in-r-prcomp-vs-princomp/)


General methods for principal component analysis

There are two general methods to perform PCA in R :

    Spectral decomposition which examines the covariances / correlations between variables
    Singular value decomposition which examines the covariances / correlations between individuals

The function princomp() uses the spectral decomposition approach. The functions prcomp() and PCA()[FactoMineR] use the singular value decomposition (SVD).

!!According to the R help, SVD has slightly better numerical accuracy. Therefore, the function prcomp() is preferred compared to princomp().

Package for PCA visualization

We'll use the factoextra R package to create a ggplot2-based elegant visualization.

    You can install it from CRAN:

Demo data sets

We'll use the data sets decathlon2 [in factoextra], which has been already described at: PCA - Data format.

Briefly, it contains:

    Active individuals (rows 1 to 23) and active variables (columns 1 to 10), which are used to perform the principal component analysis
    Supplementary individuals (rows 24 to 27) and supplementary variables (columns 11 to 13), which coordinates will be predicted using the PCA information and parameters obtained with active individuals/variables.


Other readings:

http://www.sthda.com/english/wiki/print.php?id=204
http://www.sthda.com/english/articles/31-principal-component-methods-in-r-practical-guide/112-pca-principal-component-analysis-essentials/
http://www.sthda.com/english/wiki/factoextra-r-package-easy-multivariate-data-analyses-and-elegant-visualization

install.packages(c("FactoMineR", "factoextra"))


# Bioconductor; BiocManager::install("xx")

library(limma)
library(Glimma)
library(edgeR)
library(Mus.musculus)
library(DESeq2)
library(topGO)

# CRAN; install.packages("xx")

library("FactoMineR")
library(factoextra)
library(dplyr)
library(tibble)
library(ggplot2)
library(ggpubr)
library(RColorBrewer)
library(tidyverse)
library(gplots)
library("pheatmap")
library(plotrix)
library(ggrepel)


```{r}
setwd("~/Sophieguo/Projects/MouseExperiment/LCM_Jingxu-Xiaoli_2015/results/Ranalysis")
save(x,x.norm,v.norm,efit,file ="LCM.GAL.RData")
load("LCM.GAL.RData")

#read count data and make a Limma dataset

counts <-read.delim("master_list_of_gene_counts_MIN.sense.UNNORMQUANT.REORDERED.txt", header=T,stringsAsFactors=FALSE)

counts.ZT0<-counts%>%
  dplyr::select(-starts_with("gene"))%>%
column_to_rownames(var="id")%>%
  as.matrix(.)

sample.ZT0<-data.frame(sampleID=colnames(counts.ZT0))%>%
  mutate(sleep=factor(ifelse(grepl("SS",sampleID),"SS",ifelse(grepl("SD",sampleID),"SD","ZT0")),levels=c("ZT0","SS","SD")),
         time=factor(ifelse(grepl("0hr",sampleID),"Z0",
                            ifelse(grepl("3hr",sampleID),"3hr",
                            ifelse(grepl("6hr",sampleID),"6hr",
                            ifelse(grepl("9hr",sampleID),"9hr","12hr")))),
                            levels=c("Z0","3hr","6hr","9hr","12hr")))%>%
  unite(sleep.time,sleep,time,sep=".",remove=F)%>%
  mutate(sleep.time=factor(sleep.time,levels=c("ZT0.Z0","SS.3hr","SS.6hr","SS.9hr","SS.12hr","SD.3hr","SD.6hr","SD.9hr","SD.12hr")))%>%
  column_to_rownames(var="sampleID")

library(biomaRt)

mouseensembl<-useMart(host = "aug2017.archive.ensembl.org", biomart="ensembl", dataset="mmusculus_gene_ensembl")

att<-listAttributes(mouseensembl)

#add more gene identifiers
count_gene<-getBM(attributes = c("ensembl_gene_id","entrezgene","description","gene_biotype","transcript_length"),
                mart=mouseensembl)%>%
  dplyr::rename(.,id=ensembl_gene_id)%>%
  dplyr::rename(.,EntrezID=entrezgene)%>%
  distinct(id,.keep_all = TRUE)%>%
 left_join(counts[,c(1,54,55)],.,by="id") 

x <- DGEList(counts=counts.ZT0, samples=sample.ZT0, genes=count_gene) #creat DGElist object x

table(rowSums(x$counts==0)==52) #25807/37681 genes have non-zero total counts across all 52 samples

keep <- rowMeans(x$counts) >=10 
x2 <- x[keep,]#13856/37681 has a mean counts>=10 across all samples

design <- model.matrix(~0+x$samples$sleep.time)
colnames(design) <- levels(x$samples$sleep.time)

contrast <- makeContrasts(
  SDvsSS_3hr = SD.3hr-SS.3hr, 
  SDvsSS_6hr = SD.6hr-SS.6hr, 
  SDvsSS_9hr = SD.9hr-SS.9hr, 
  SDvsSS_12hr = SD.12hr-SS.12hr, levels=design)

x.norm<- calcNormFactors(x2, method = "TMM") 
v.norm<-voom(x.norm,design=design,plot=FALSE) #renormalize after filtering
vfit <- lmFit(v.norm, design)

efit<- eBayes(contrasts.fit(vfit, contrasts=contrast)) 

DEG.SDvSS<-topTable(efit,coef=1:4,n=Inf,p.value=0.05) #263 DEG with FDR<5% for SDvSS.young animals

DEG.2<-topTable(efit,coef=1:4,n=Inf)%>%
    mutate(aveFC=rowMeans(.[,8:11]))


keep.DEG<-DEG.2%>%
  arrange(id)%>%
  unlist(.$adj.P.Val)<0.05

keep.DEG<-unlist(DEG.2%>%arrange(id)%>%.$adj.P.Val)<0.05

v.DEG<-v.norm[keep.DEG,]

contrast.time <- makeContrasts(
  SS.3hrvsZ0 = SS.3hr-ZT0.Z0, 
  SS.6hrvsZ0 = SS.6hr-ZT0.Z0, 
  SS.9hrvsZ0 = SS.9hr-ZT0.Z0, 
  SS.12hrvsZ0 = SS.12hr-ZT0.Z0, 
  SD.3hrvsZ0 = SD.3hr-ZT0.Z0,
  SD.6hrvsZ0 = SD.6hr-ZT0.Z0,
  SD.9hrvsZ0 = SD.9hr-ZT0.Z0,
  SD.12hrvsZ0 = SD.12hr-ZT0.Z0,
  levels=design)

vfit.time <- lmFit(v.DEG, design)

efit.time.SDvSSDEG<- eBayes(contrasts.fit(vfit.time, contrasts=contrast.time)) #identify DEG between SS and ZT0 among SDvSS.DEGs (n=263)

efit.time<- eBayes(contrasts.fit(vfit, contrasts=contrast.time)) #identify DEG between SS and ZT0 among all genes (n=13856)


DEG.time.SDvSSDEG<-topTable(efit.time.SDvSSDEG,coef=1:4,n=Inf,p.value=0.05) #21 DEG with FDR<5% for SS vs. ZT0; 47 DEG with FDR<10% for SS vs. ZT0 when using 263 DEGs from SD vs. SS

DEG.time<-topTable(efit.time,coef=1:4,n=Inf,p.value=0.05) #13 DEG with FDR<5% for SS vs. ZT0; 19 DEG with FDR<10% for SS vs. ZT0 when use all 13856 genes


keep.sample<-colnames(v.norm)!="F5.SD9hr"

x.rm<-x.norm[,keep.sample]
design.rm <- model.matrix(~0+x.rm$samples$sleep.time)
colnames(design.rm) <- levels(x.rm$samples$sleep.time)

v.rm<-voom(x.rm,design=design.rm,plot=FALSE) #renormalize after filtering
vfit.rm <- lmFit(v.rm, design.rm)

efit.rm<- eBayes(contrasts.fit(vfit.rm, contrasts=contrast)) 

DEG.SDvSS.rm<-topTable(efit,coef=1:4,n=Inf,p.value=0.05) #263 DEG with FDR<5% for SDvSS.young animals



DEG.combine<-topTable(efit.time,coef=1:4,n=Inf)%>%
    mutate(aveFC=rowMeans(.[,8:11]))%>%
dplyr::select(id,starts_with("SS"),aveFC,P.Value,adj.P.Val)%>%
  left_join(DEG.2,.,by="id",suffix=c(".SDvSS",".SSvZ0"))%>%
  left_join(.,cbind(id=efit$gene$id,data.frame(efit$p.value),stringsAsFactors=F),by="id",suffix=c(".FC",".pval"))%>%
  left_join(.,cbind(id=efit.time$gene$id,data.frame(efit.time$p.value,stringsAsFactors=F)[,1:4]),by="id",suffix=c(".FC",".pval"))%>%

  mutate(dir.SDvSS.all=ifelse(adj.P.Val.SDvSS>=0.05,"NS",ifelse(aveFC.SDvSS>0,"SD.up","SD.down")),dir.SSvZ0.all=ifelse(adj.P.Val.SSvZ0>=0.05,"NS",ifelse(aveFC.SSvZ0>0,"SSvZ0.up","SSvZ0.down")))%>%
  mutate(class.SDvSS=ifelse(adj.P.Val.SDvSS>=0.05,"NS",ifelse(SDvsSS_3hr.pval<0.05&SDvsSS_6hr.pval<0.05&SDvsSS_9hr.pval<0.05&SDvsSS_12hr.pval<0.05,"all4",ifelse(SDvsSS_9hr.pval>=0.05&SDvsSS_12hr.pval>=0.05,"early",ifelse(SDvsSS_3hr.pval>=0.05&SDvsSS_6hr.pval>=0.05,"late",ifelse((-log10(SDvsSS_3hr.pval)-log10(SDvsSS_6hr.pval))>(-log10(SDvsSS_9hr.pval)-log10(SDvsSS_12hr.pval)),"early.trend","late.trend"))))))%>%
  mutate(dir.class.SDvSS=ifelse(class.SDvSS=="NS","NS",                        ifelse(class.SDvSS=="all4"&aveFC.SDvSS>0,"SD.up.all",     ifelse(class.SDvSS=="all4"&aveFC.SDvSS<0,"SD.down.all",          ifelse(class.SDvSS%in%c("early","early.trend")&((SDvsSS_3hr.FC+SDvsSS_6hr.FC)/2)>0,"SD.up.early",                               ifelse(class.SDvSS%in%c("early","early.trend")&((SDvsSS_3hr.FC+SDvsSS_6hr.FC)/2)<0,"SD.down.early",    ifelse(class.SDvSS%in%c("late","late.trend")&((SDvsSS_9hr.FC+SDvsSS_12hr.FC)/2)>0,"SD.up.late",                               ifelse(class.SDvSS%in%c("late","late.trend")&((SDvsSS_9hr.FC+SDvsSS_12hr.FC)/2)<0,"SD.down.late","na"))))))))%>%
  mutate(mismatch=ifelse(dir.SDvSS.all=="SD.up"&dir.class.SDvSS=="SD.down.late","mismatch","same"))

write.table(DEG.combine,"DEG.combine.txt",sep="\t",row.names = F)

DEG.time<-topTable(efit.time,coef=1:4,n=Inf,p.value=0.05) #13 DEG with FDR<5% for SS vs. ZT0; 19 DEG with FDR<10% for SS vs. ZT0 when use all 13856 genes

# volcano plot

png(filename = "volcano.SDvSS.LCM.png",height = 4.5,width =5,res = 300,units = "in",type="cairo")

DEG.combine%>%
  dplyr::select(id, geneSymbol,adj.P.Val.SDvSS,aveFC.SDvSS,dir.SDvSS.all)%>%
  mutate(logFDR=-log10(adj.P.Val.SDvSS))%>%
ggscatter(.,x="aveFC.SDvSS",y="logFDR",color="dir.SDvSS.all",palette=c("grey50","green4","red3"),size=1.5,label="geneSymbol",label.select=list(criteria="`x`>0&`y`>6|`x`<0&`y`>4"),repel=T,xlim=c(-3.5,3.5),ylim=c(0,11),ylab=expression('-log'[10]*'FDR'),xlab=expression('Average log'[2]*'fold change of SD vs. SS'),title="DEGs of SD vs.SS in galanin cells")+
  geom_hline(yintercept=1.3,linetype="dashed",color="grey")+
  geom_vline(xintercept=0,linetype="dashed",color="grey")+
  annotate("text", x = c(-2.5,-2.7), y = c(9,7.7), label = c("79", "184"), color=c("green4","red"), size=8, fontface="bold")+
  annotate("segment", x = -2, xend = -2, y = 9.4, yend = 8.5, colour = "green4", size=2, alpha=1, arrow=arrow(length = unit(0.3, "cm")))+
  annotate("segment", x = -2, xend =-2, y = 7.3, yend = 8, colour = "red", size=2, alpha=1, arrow=arrow(length = unit(0.3, "cm")))+
  theme(plot.title=element_text(face="bold",size=16,hjust=0.5),legend.position="none",axis.title=element_text(face="bold",size=14),axis.text=element_text(face="bold",size=10))+
scale_x_continuous(breaks=c(-3,-2,-1,0,1,2,3))+scale_y_continuous(breaks=c(1.3,5,10))

dev.off()


# heatmap with DEG.SDvSS

test<-topTable(efit.time,coef=1:8,n=Inf)%>%
  dplyr::select(id,ends_with("Z0"))%>%
  left_join(.,DEG.combine,by="id")%>%
  filter(dir.SDvSS.all!="NS")%>%
  dplyr::select(geneSymbol,ends_with("Z0"),dir.SDvSS.all,class.SDvSS)

SSvsZ0.up<-test%>%
  filter(dir.SDvSS.all=="SD.up")%>%
  .[,1:5]%>%
  mutate(Z0=rep(0,184))%>%
  column_to_rownames(.,var="geneSymbol")%>%
  .[,c(5,1,2,3,4)]

SDvsZ0.up<-test%>%
    filter(dir.SDvSS.all=="SD.up")%>%
  .[,c(1,6:9)]%>%
  mutate(Z0=rep(0,184))%>%
  column_to_rownames(.,var="geneSymbol")%>%
  .[,c(5,1,2,3,4)]

heatmap.up<-cbind(t(scale(t(as.matrix(SSvsZ0.up)))),t(scale(t(as.matrix(SDvsZ0.up))))) #z-scale row-wise for SS and SD separately and then cbind them together
heatmap.2(heatmap.up,scale="none",dendrogram="row",Rowv=TRUE,Colv=NULL, trace="none", na.rm = T)

png(filename = "uniqY.SStimeHM.png",height = 30,width = 20,res = 150,units = "cm",type="cairo")
par(mar=c(5.1,0.5,4.1,0.5))
my_palette <- colorRampPalette(c("blue", "grey88", "red"))(n = 100)
heatmap.2(x.yuniq,scale="none",dendrogram ="row",Rowv=TRUE,Colv=NULL, trace="none", na.rm = T,col=my_palette, labCol=c("Y.Z0","Y.3hvsZ0","Y.6hvsZ0","Y.9hvsZ0","Y.12hvsZ0","O.Z0","O.3hvsZ0","O.6hvsZ0","O.9hvsZ0","O.12hsZ0"),cexCol=1.5,srtCol=35,offsetCol=-0.5,labRow = FALSE,xlab = "", ylab = "",lhei =c(0.1,0.9),lwid = c(0.2,0.8),colsep = 5, sepcolor = "white", main="Unique temporal sleep genes in Young")
dev.off()



# heatmap of SS vs. Z0 and SD vs. Z0 of all DEG.SDvsSS (263)
rowan=data.frame(Direction=test$dir.SDvSS.all)
rownames(rowan)=test$geneSymbol
ann.color = list(Direction=c(SD.up="red3",SD.down="green4"))


png(filename ="DEG.263.heatmap.png",height = 6,width =4,res = 300,units = "in",type="cairo")


SSSD.matrix<-test[,1:9]%>%
    column_to_rownames(.,var="geneSymbol")%>%
  as.matrix()%>%
  pheatmap(.,scale="row",cluster_col=FALSE,color=(brewer.pal(8,"BuPu")),legend=F,annotation_row =rowan, annotation_colors = ann.color, border_color= NA,labels_col = c("3","6","9","12","3","6","9","12"), show_rownames=FALSE,gaps_col=4,fontsize = 10,treeheight_row=0)

dev.off()

# compare DEG.LCM and DEG.mPFC

library(VennDiagram)

DEG.LCMvsmPFC<-DEG.combine%>%
  dplyr::select(id,geneSymbol,adj.P.Val.SDvSS,dir.SDvSS.all)%>%
  left_join(.,DEG.combine.all[,c(1,12,34)],by="id")%>%
  filter(!is.na(FDR.Y))%>%
  mutate(dir.mPFC=ifelse(FDR.Y<0.05&aveFC.SDvSS.Y>0,"SD.up",ifelse(FDR.Y<0.05&aveFC.SDvSS.Y<0,"SD.down","NS")))%>%
  rename(FDR.LCM=adj.P.Val.SDvSS,dir.LCM=dir.SDvSS.all,FDR.mPFC.Y=FDR.Y)%>%
  mutate(group=ifelse(FDR.LCM>=0.05,"NS",ifelse(dir.LCM=="SD.up"&dir.mPFC=="SD.up","common.up",ifelse(dir.LCM=="SD.down"&dir.mPFC=="SD.down","common.down",ifelse(dir.LCM=="SD.up","LCM.only.up","LCM.only.down")))))


venn.diagram(list(DEG.LCMvsmPFC[DEG.LCMvsmPFC$dir.LCM=="SD.up",]$id,DEG.LCMvsmPFC[DEG.LCMvsmPFC$dir.mPFC=="SD.up",]$id), "SDup.LCMvs.mPFC.tiff", fill=brewer.pal(5,"Set1")[c(1,5)], cex = 3, main="SD-up genes", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=3, cat.cex=3,category.names=c("Gal","mPFC"),cat.pos=c(-50,55))

venn.diagram(list(DEG.LCMvsmPFC[DEG.LCMvsmPFC$dir.LCM=="SD.down",]$id,DEG.LCMvsmPFC[DEG.LCMvsmPFC$dir.mPFC=="SD.down",]$id), "SDdown.LCMvs.mPFC.tiff", fill=brewer.pal(5,"Set1")[c(3,2)], cex = 3, main="SD-down genes", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=3, cat.cex=3,category.names=c("Gal","mPFC"),cat.pos=c(-50,55))


#GO for common genes between LCM.Gal and mPFC (young mice)

write.csv(DEG.LCMvsmPFC%>%filter(group!="NS"),"DEG.LCMvsmPFC.csv")

DEG.mPFCvsLCM<-DEG.LCMvsmPFC%>%
  mutate(group.mPFC=ifelse(FDR.mPFC.Y>=0.05,"NS",ifelse(dir.LCM=="SD.up"&dir.mPFC=="SD.up","common.up",ifelse(dir.LCM=="SD.down"&dir.mPFC=="SD.down","common.down",ifelse(dir.mPFC=="SD.up","mPFC.only.up","mPFC.only.down")))))

write.csv(DEG.mPFCvsLCM%>%filter(group.mPFC!="NS"),"DEG.mPFCvsLCM.csv")

# plot Hspa8 and Hspa5
x.LCM=x.norm
v.LCM=v.norm


CPM.LCM<-cbind(x.LCM$genes,as.data.frame(v.LCM$E))%>%
  filter(geneSymbol%in%c("Hspa5","Hspa8","Hsph1","Cirbp","Rbm3","Cck","Crh","Tac1","Pdyn","Fos"))%>%
  dplyr::select(geneSymbol,ends_with("hr"))%>%
  mutate(B2.SD0hr=B2.CTL0hr,B11.SD0hr=B11.CTL0hr,C10.SD0hr=C10.CTL0hr,D9.SD0hr=D9.CTL0hr,E8.SD0hr=E8.CTL0hr,F7.SD0hr=F7.CTL0hr)%>%
  dplyr::rename(B2.SS0hr=B2.CTL0hr,B11.SS0hr=B11.CTL0hr,C10.SS0hr=C10.CTL0hr,D9.SS0hr=D9.CTL0hr,E8.SS0hr=E8.CTL0hr,F7.SS0hr=F7.CTL0hr)%>%
  gather(sample,CPM,-geneSymbol)%>%
  mutate(sleep=ifelse(grepl("SS",sample),"SS","SD"),time=ifelse(grepl("0hr",sample),0,ifelse(grepl("3hr",sample),3,ifelse(grepl("6hr",sample),6,ifelse(grepl("9hr",sample),9,12)))))%>%
  mutate(sleep=factor(sleep,levels=c("SD","SS")))%>%
  mutate(experiment="Gal.LCM")

# fetch gene expression from the mPFC dataset
x.mPFC=x_norm
v.mPFC=v.norm
  
CPM.mPFC<-cbind(x.mPFC$genes,as.data.frame(v.mPFC$E))%>%
  filter(geneSymbol%in%c("Hspa5","Hspa8","Hsph1","Cirbp","Rbm3","Cck","Crh","Tac1","Pdyn","Fos"))%>%
  dplyr::select(geneSymbol,starts_with("Young"))%>%
  dplyr::rename(YZ0.SS.B2=Young.Z0.66902.B2, YZ0.SS.G4=Young.Z0.66919.G4, YZ0.SS.D8=Young.Z0.66938.D8, YZ0.SS.F4=Young.Z0.66972.F4, YZ0.SS.E7=Young.Z0.66988.E7, YZ0.SS.E11=Young.Z0.67008.E11)%>%
  mutate(YZ0.SE.B2=YZ0.SS.B2, YZ0.SD.G4=YZ0.SS.G4, YZ0.SD.D8=YZ0.SS.D8, YZ0.SD.F4=YZ0.SS.F4, YZ0.SD.E7=YZ0.SS.E7, YZ0.SD.E11=YZ0.SS.E11)%>%
  gather(sample,CPM,-geneSymbol)%>%
  mutate(sleep=ifelse(grepl("SS",sample),"SS","SD"),time=ifelse(grepl("Z0",sample),0,ifelse(grepl("3hr",sample),3,ifelse(grepl("6hr",sample),6,ifelse(grepl("9hr",sample),9,12)))))%>%
  mutate(sleep=factor(sleep,levels=c("SD","SS")))%>%
  mutate(experiment="mPFC")


png(filename ="Fos.LCMvmPFC.png",height = 3,width =4,res = 300,units = "in",type="cairo")
rbind(CPM.LCM,CPM.mPFC)%>%
  filter(geneSymbol=="Fos")%>%
  ggline(.,x="time",y="CPM",color="sleep",size=1,facet.by="experiment",add="mean_se",ylab=expression('log'[2]*'count per million'),xlab="duration of sleep or SD (h)",title="c-Fos")+
  theme(legend.position="none",plot.title=element_text(face="bold",hjust=0.5,size=14),axis.title=element_text(size=12),strip.text =element_text(face="bold",size=12,color="white"), strip.background = element_rect(fill="grey50"))
  dev.off()

  png(filename ="Fos.LCMvmPFC.boxplot.png",height = 3,width =4,res = 300,units = "in",type="cairo")
  
rbind(CPM.LCM,CPM.mPFC)%>%
  filter(geneSymbol=="Fos")%>%
  ggboxplot(.,x="time",y="CPM",color="sleep",add="jitter",facet.by="experiment",ylab=expression('log'[2]*'count per million'),xlab="duration of sleep or SD (h)",title="c-Fos")+
  theme(legend.position="none",plot.title=element_text(face="bold",hjust=0.5,size=14),axis.title=element_text(size=12),strip.text =element_text(face="bold",size=12,color="white"), strip.background = element_rect(fill="grey50"))
  
  dev.off()
  

png(filename ="Hspa8.LCMvmPFC.png",height = 3,width =4,res = 300,units = "in",type="cairo")
rbind(CPM.LCM,CPM.mPFC)%>%
  filter(geneSymbol=="Hspa8")%>%
  ggline(.,x="time",y="CPM",color="sleep",size=1,facet.by="experiment",add="mean_se",ylab=expression('log'[2]*'count per million'),xlab="duration of sleep or SD (h)",title="Hspa8")+
  theme(legend.position="none",plot.title=element_text(face="bold",hjust=0.5,size=14),axis.title=element_text(size=12),strip.text =element_text(face="bold",size=12,color="white"), strip.background = element_rect(fill="grey50"))
  dev.off()


png(filename ="Hsph1.LCMvmPFC.png",height = 3,width =4,res = 300,units = "in",type="cairo")
rbind(CPM.LCM,CPM.mPFC)%>%
  filter(geneSymbol=="Hsph1")%>%
  ggline(.,x="time",y="CPM",color="sleep",size=1,facet.by="experiment",add="mean_se",ylab=expression('log'[2]*'count per million'),xlab="duration of sleep or SD (h)",title="Hsph1")+
  theme(legend.position="none",plot.title=element_text(face="bold",hjust=0.5,size=14),axis.title=element_text(size=12),strip.text =element_text(face="bold",size=12,color="white"), strip.background = element_rect(fill="grey50"))
  dev.off()
  
png(filename ="Rbm3.LCMvmPFC.png",height = 3,width =4,res = 300,units = "in",type="cairo")
rbind(CPM.LCM,CPM.mPFC)%>%
  filter(geneSymbol=="Rbm3")%>%
  ggline(.,x="time",y="CPM",color="sleep",size=1,facet.by="experiment",add="mean_se",ylab=expression('log'[2]*'count per million'),xlab="duration of sleep or SD (h)",title="Rbm3")+
  theme(legend.position="none",plot.title=element_text(face="bold",hjust=0.5,size=14),axis.title=element_text(size=12),strip.text =element_text(face="bold",size=12,color="white"), strip.background = element_rect(fill="grey50"))
  dev.off()
  
  
png(filename ="Cirbp.LCMvmPFC.png",height = 3,width =4,res = 300,units = "in",type="cairo")
rbind(CPM.LCM,CPM.mPFC)%>%
  filter(geneSymbol=="Cirbp")%>%
  ggline(.,x="time",y="CPM",color="sleep",size=1,facet.by="experiment",add="mean_se",ylab=expression('log'[2]*'count per million'),xlab="duration of sleep or SD (h)",title="Cirbp")+
  theme(legend.position="none",plot.title=element_text(face="bold",hjust=0.5,size=14),axis.title=element_text(size=12),strip.text =element_text(face="bold",size=12,color="white"), strip.background = element_rect(fill="grey50"))
  dev.off()  
  
  CPM.LCM%>%
  filter(geneSymbol%in%c("Tac1","Pdyn"))%>%
  ggline(.,x="time",y="CPM",color="sleep",facet.by="geneSymbol",add="mean_se")

rbind(CPM.LCM,CPM.mPFC)%>%
  filter(geneSymbol=="Rbm3")%>%
  ggline(.,x="time",y="CPM",color="sleep",facet.by="experiment",add="mean_se")
  
png(filename ="Tac1Pdyn.png",height = 3,width =4,res = 300,units = "in",type="cairo")

CPM.LCM%>%
  filter(time!=0)%>%
  mutate(sleep=factor(sleep,levels=c("SS","SD")))%>%
  filter(geneSymbol%in%c("Tac1","Pdyn"))%>%
ggviolin(.,x="sleep",y="CPM",fill="sleep",palette = c("#00AFBB","#FC4E07"), add=c("boxplot"), add.params = list(fill = "white"), facet.by="geneSymbol", ylab=expression('log'[2]*'count per million'),xlab="")+
  theme(legend.position="none",strip.text = element_text(face="bold",size=16,color="white"), axis.title.y = element_text(face="bold",size=12),strip.background = element_rect(color="white",fill = "grey50"), panel.background=element_rect(fill = NA,color = "blue",colour="white"))

dev.off()  


ggviolin(.,x="age",y="p.val",fill = "class",palette = c("#00AFBB", "#E7B800", "#FC4E07"), add="boxplot",add.params = list(fill = "white"), xlab="",ylab=expression('-log'[10]*' P-values'), title="Wake Genes")+
  facet_wrap(vars(class,ngene),strip.position = "bottom")+
  stat_compare_means(label="p.signif",paired = TRUE,label.x=1.4,label.y=40)+
theme(legend.position="none",strip.text = element_text(face="bold",size=12,color="white"), axis.text.x = element_text(face="bold",size=11),strip.background = element_rect(color="white",fill = brewer.pal(9,"PuBu")[8]), plot.title=element_text(face="bold",size=18,hjust=0.5), axis.title.y=element_text(face="bold",size=14))+
  scale_y_continuous(expand=c(0,1))


  #DEG before and after controling four markers

#use expression of Tac1, Pdyn as covariates
control.Tac1Pdyn<-cbind(x.norm$genes,as.data.frame(v.norm$E))%>%
  filter(geneSymbol%in%c("Tac1","Pdyn"))%>%
  dplyr::select(-id,-geneCoordinate,-EntrezID,-description,-gene_biotype,-transcript_length)%>%
  gather(sample,CPM,-geneSymbol)%>%
  mutate(sample = factor(sample, unique(sample)))%>%
  spread(geneSymbol,CPM)

control.design=model.matrix(~0+control.Tac1Pdyn$Tac1+control.Tac1Pdyn$Pdyn+x$samples$sleep.time)
colnames(control.design) <- c("Tac1","Pdyn","ZT0","SS.3hr","SS.6hr","SS.9hr","SS.12hr","SD.3hr","SD.6hr","SD.9hr","SD.12hr")

control.contrast <- makeContrasts(
  SDvsSS_3hr = SD.3hr-SS.3hr, 
  SDvsSS_6hr = SD.6hr-SS.6hr, 
  SDvsSS_9hr = SD.9hr-SS.9hr, 
  SDvsSS_12hr = SD.12hr-SS.12hr, levels=control.design)

vfit.control <- lmFit(v.norm, control.design)

efit.control<- eBayes(contrasts.fit(vfit.control, contrasts=control.contrast)) 

control.DEG<-topTable(efit.control,coef=1:4,n=Inf,p.value=0.05) #184 DEG with FDR<5% after controling for cck, crh, tac1 and pdyn


DEG.time<-topTable(efit.time,coef=1:4,n=Inf,p.value=0.05) #13 DEG with FDR<5% for SS vs. ZT0; 19 DEG with FDR<10% for SS vs. ZT0 when use all 13856 genes

# volcano plot

png(filename = "volcano.controlTac1Pdyn.png",height = 4.5,width =5,res = 300,units = "in",type="cairo")


topTable(efit.control,coef=1:4,n=Inf)%>%
  mutate(aveFC=rowMeans(.[,8:11]))%>%
  mutate(dir=ifelse(adj.P.Val>0.05,"NS",ifelse(aveFC>0,"up","down")),logFDR=-log10(adj.P.Val))%>%
ggscatter(.,x="aveFC",y="logFDR",color="dir",palette=c("green4","grey50","red3"),size=1.5,label="geneSymbol",label.select=list(criteria="`x`>0&`y`>6|`x`<0&`y`>3.5"),repel=T,xlim=c(-3.5,3.5),ylim=c(0,11),ylab=expression('-log'[10]*'FDR'),xlab=expression('Average log'[2]*'fold change of SD vs. SS'),title="DEGs after control for Tac1 and Pdyn")+
  geom_hline(yintercept=1.3,linetype="dashed",color="grey")+
  geom_vline(xintercept=0,linetype="dashed",color="grey")+
  annotate("text", x = c(-2.5,-2.7), y = c(9,7.7), label = c("81", "163"), color=c("green4","red"), size=8, fontface="bold")+
  annotate("segment", x = -2, xend = -2, y = 9.4, yend = 8.5, colour = "green4", size=2, alpha=1, arrow=arrow(length = unit(0.3, "cm")))+
  annotate("segment", x = -2, xend =-2, y = 7.3, yend = 8, colour = "red", size=2, alpha=1, arrow=arrow(length = unit(0.3, "cm")))+
  theme(plot.title=element_text(face="bold",size=16,hjust=0.5),legend.position="none",axis.title=element_text(face="bold",size=14),axis.text=element_text(face="bold",size=10))+
scale_x_continuous(breaks=c(-3,-2,-1,0,1,2,3))+scale_y_continuous(breaks=c(1.3,5,10))

dev.off()


test.heat<-cbind(x$genes,as.data.frame(v.norm$E))%>%
  filter(geneSymbol%in%c("Fos","Hspa8","Hsph1","Hspa5","Dnajb1"))%>%
  dplyr::select(-id,-geneCoordinate,-EntrezID,-description,-gene_biotype,-transcript_length)%>%
  gather(sample,CPM,-geneSymbol)%>%
  mutate(sample = factor(sample, unique(sample)))%>%
  spread(geneSymbol,CPM)

test2.design=model.matrix(~0+test.heat$Fos+x$samples$sleep.time)
colnames(test2.design) <- c("Fos","ZT0","SS.3hr","SS.6hr","SS.9hr","SS.12hr","SD.3hr","SD.6hr","SD.9hr","SD.12hr")

test.contrast <- makeContrasts(
  SDvsSS_3hr = SD.3hr-SS.3hr, 
  SDvsSS_6hr = SD.6hr-SS.6hr, 
  SDvsSS_9hr = SD.9hr-SS.9hr, 
  SDvsSS_12hr = SD.12hr-SS.12hr, levels=test2.design)

vfit.test <- lmFit(v.norm, test2.design)

efit.test<- eBayes(contrasts.fit(vfit.test, contrasts=test.contrast)) 

test2.DEG<-topTable(efit.test,coef=1:4,n=Inf,p.value=0.05) #0 DEG with FDR<5% after controling for "Fos","Hspa8","Hsph1","Hspa5","Dnajb1"



test<-cbind(x$genes,as.data.frame(v.norm$E))%>%
  filter(geneSymbol%in%c("Cck","Crh","Tac1","Pdyn"))%>%
  dplyr::select(-id,-geneCoordinate,-EntrezID,-description,-gene_biotype,-transcript_length)%>%
  gather(sample,CPM,-geneSymbol)%>%
  mutate(sample = factor(sample, unique(sample)))%>%
  spread(geneSymbol,CPM)

test2.design=model.matrix(~0+test$Pdyn+test$Tac1+x$samples$sleep.time)
colnames(test2.design) <- c("Pdyn","Tac1","ZT0","SS.3hr","SS.6hr","SS.9hr","SS.12hr","SD.3hr","SD.6hr","SD.9hr","SD.12hr")

test.contrast <- makeContrasts(
  SDvsSS_3hr = SD.3hr-SS.3hr, 
  SDvsSS_6hr = SD.6hr-SS.6hr, 
  SDvsSS_9hr = SD.9hr-SS.9hr, 
  SDvsSS_12hr = SD.12hr-SS.12hr, levels=test2.design)

vfit.test <- lmFit(v.norm, test2.design)

efit.test<- eBayes(contrasts.fit(vfit.test, contrasts=test.contrast)) 

test2.DEG<-topTable(efit.test,coef=1:4,n=Inf,p.value=0.05) #226 DEG with FDR<5% after controling for cck, crh, tac1 and pdyn





library(VennDiagram)

#DEG before and after controling four markers

venn.diagram(list(DEG.SDvSS$id,test.DEG$id), "BandA.controlmarker.tiff", fill=brewer.pal(4,"Set1")[3:4], cex = 3, main="Control Cck, Crh, Tac1, Pdyn", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=3, cat.cex=3,category.names=c("before","after"),cat.pos=c(-50,55))

venn.diagram(list(DEG.SDvSS$id,test2.DEG$id), "BandA.control3marker.tiff", fill=brewer.pal(4,"Set1")[3:4], cex = 3, main="Control Cck, Tac1, Pdyn", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=3, cat.cex=3,category.names=c("before","after"),cat.pos=c(-50,55))

venn.diagram(list(DEG.SDvSS$id,test2.DEG$id), "BandA.control2marker.tiff", fill=brewer.pal(4,"Set1")[3:4], cex = 3, main="Control Tac1, Pdyn", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=3, cat.cex=3,category.names=c("before","after"),cat.pos=c(-50,55))


lost.id<-data.frame(id=setdiff(DEG.SDvSS$id,test.DEG$id))%>%
  left_join(.,DEG.2,by="id")

gain.id<-data.frame(id=setdiff(test.DEG$id,DEG.SDvSS$id))%>%
    left_join(.,DEG.2,by="id")

lost2.id<-data.frame(id=setdiff(DEG.SDvSS$id,test2.DEG$id))%>%
  left_join(.,DEG.2,by="id")

gain2.id<-data.frame(id=setdiff(test2.DEG$id,DEG.SDvSS$id))%>%
    left_join(.,DEG.2,by="id")


#plot marker genes

marker<-c("Tac1","Pdyn","Crh","Cck","Dnahc6","Wdr62","Chaf1a","Eif2ak2","Slco1a4","Col15a1","Oprk1")

  
marker.plot<-cbind(gene=v.norm$genes$geneSymbol,as.data.frame(v.norm$E))%>%
  filter(gene%in%marker)%>%
  rename(B2.SS0hr=B2.CTL0hr,B11.SS0hr=B11.CTL0hr,C10.SS0hr=C10.CTL0hr,D9.SS0hr=D9.CTL0hr,E8.SS0hr=E8.CTL0hr,F7.SS0hr=F7.CTL0hr)%>%
  mutate(B2.SD0hr=B2.SS0hr,B11.SD0hr=B11.SS0hr,C10.SD0hr=C10.SS0hr,D9.SD0hr=D9.SS0hr,E8.SD0hr=E8.SS0hr,F7.SD0hr=F7.SS0hr)%>%
    gather(sampleID,CPM,-gene)%>%
  mutate(sleep=ifelse(grepl("SS",sampleID),"SS","SD"),
         time=ifelse(grepl("0hr",sampleID),0,ifelse(grepl("3hr",sampleID),3,ifelse(grepl("6hr",sampleID),6,ifelse(grepl("9hr",sampleID),9,12)))))%>%
  mutate(gene=factor(gene,levels=c("Tac1","Pdyn","Crh","Cck","Dnahc6","Wdr62","Chaf1a","Eif2ak2","Slco1a4","Col15a1","Oprk1")),sleep=factor(sleep,levels=c("SS","SD")))%>%
  spread(gene,CPM)

 exp.palette<-c(brewer.pal(9,"OrRd")[c(6,9)],brewer.pal(9,"GnBu")[c(6,9)])

ggboxplot(marker.plot,x="time",y="Wdr62",add="jitter",color="sleep")   
 
ggscatter(marker.plot,x="Tac1",y="Dnahc6",color="sleep")

   
cir.plot<-function(i){plot<-cir.core.plot%>%
  filter(geneSymbol==levels(cir.core.plot$geneSymbol)[i])%>%
  ggline(.,x="time",y="CPM",color="age.sleep",palette =exp.palette, linetype="age.sleep",add ="mean_se",size=0.8,plot_type="l",xlab ="",ylab = "",title=levels(cir.core.plot$geneSymbol)[i])+scale_linetype_manual(values=c("solid","longdash","solid","longdash"))+theme(legend.title=element_blank(),legend.position="right",plot.title = element_text(size=10,face="bold",hjust = 0.5), axis.text=element_text(size=8))+ scale_y_continuous(labels = scales::number_format(accuracy = 0.1))
  return(plot)}

cir.pp<-ggarrange(cir.plot(1),cir.plot(2),cir.plot(3),cir.plot(4),cir.plot(5),cir.plot(6),cir.plot(7),cir.plot(8),cir.plot(9),cir.plot(10),cir.plot(11),cir.plot(12),ncol = 4, nrow =3,common.legend = TRUE, legend = "right")

png(filename = "circadian.core.png",height = 6,width =8,res = 300,units = "in",type="cairo")
annotate_figure(cir.pp,top = text_grob("Circadian core clock genes", color = "black", face = "bold", size = 14),bottom = text_grob("Zeitgeber time (ZT)", color = "black",hjust = 0.5, face = "bold", size = 12),left = text_grob("Gene expression (logCPM)", color = "black", face = "bold",rot = 90, size = 12),fig.lab.size=14,fig.lab.face="bold")
dev.off()




marker.FC<-DEG.2%>%
  filter(geneSymbol%in%c("Cck","Crh","Tac1","Pdyn"))

png(filename = "markerFC.png",height =5,width =5,res = 300,units = "in",type="cairo")
ggscatter(marker.FC,x="AveExpr",y="aveFC",label="geneSymbol")
dev.off()

#PCA with marker genes (neuropeptide, transcription factor, neuropeptide receptor, and DEGs [SS-top50; SD-top50])

PCAmarker<-read.delim("PCAgenelist.txt",header=T,row.names = NULL,stringsAsFactors = F)

marker.genes<-data.frame(v.norm$E)%>%
  rownames_to_column(var="id")%>%
  left_join(x$genes,.,by="id")%>%
  left_join(data.frame(geneSymbol=PCAmarker[,7]),.,by="geneSymbol")%>% 
  filter(!is.na(id))%>%
  distinct(id,.keep_all=T)%>%
  column_to_rownames(var="geneSymbol")

marker.pca <- PCA(t(as.matrix(marker.genes[,7:58])), graph =F)

fviz_pca_ind(marker.pca, axes=c(1,2),
             label = "ind", # individual labels
             repel = F,
             habillage = x.norm$samples$sleep, # color by groups
             palette = c("#00AFBB", "blue", "#FC4E07"),
             addEllipses = F, # Concentration ellipses,
             mean.point=F
             )

#k-means
set.seed(123)
km.mark <- kmeans(scale(t(as.matrix(marker.genes[,7:58]))), 2, nstart = 1)

fviz_cluster(km.mark, data = scale(t(as.matrix(marker.genes[,7:58]))),
             palette = c("#00AFBB","#E7B800"),
             ggtheme = theme_minimal(),
             main = "Partitioning Clustering Plot"
             )

#Hierarchical clustering
res.mark <- hcut(t(as.matrix(marker.genes[,7:58])), k = 6, stand = TRUE)
# Visualize
png(filename = "markergenes.HC.png",height =5,width =10,res = 300,units = "in",type="cairo")

fviz_dend(res.mark, rect = TRUE, cex = 0.5,
          k_colors = c("#00AFBB","green","violetred","#2E9FDF", "#E7B800", "#FC4E07"))
dev.off()


png(filename = "markergenes.png",height =10,width =4,res = 300,units = "in",type="cairo")

cbind(data.frame(marker.cluster=res.mark$cluster),t(marker.genes[,7:58]))%>%
  group_by(marker.cluster)%>%
  summarise_if(is.numeric, mean, na.rm = TRUE)%>%
  mutate(cluster="cluster")%>%
  unite(mark,c(cluster,marker.cluster))%>%
  gather(gene,mean,-mark)%>%
  spread(mark,mean)%>%
  column_to_rownames(var="gene")%>%
  as.matrix()%>%
  pheatmap(.,treeheight_col=10)

dev.off()

png(filename = "markergenes.scaled.png",height =10,width =4,res = 300,units = "in",type="cairo")

cbind(data.frame(marker.cluster=res.mark$cluster),t(marker.genes[,7:58]))%>%
  group_by(marker.cluster)%>%
  summarise_if(is.numeric, mean, na.rm = TRUE)%>%
  mutate(cluster="cluster")%>%
  unite(mark,c(cluster,marker.cluster))%>%
  gather(gene,mean,-mark)%>%
  spread(mark,mean)%>%
  column_to_rownames(var="gene")%>%
  as.matrix()%>%
  pheatmap(.,scale="row",treeheight_col=10)

dev.off()

test.group6<-cbind(data.frame(marker.cluster=res.mark$cluster),t(marker.genes[,7:58]))%>%
  mutate(group.6=ifelse(marker.cluster==6,"group6","rest"),group.4=ifelse(marker.cluster==4,"group4","rest"))%>%
  group_by(group.6)%>%  
  summarise_if(is.numeric, mean, na.rm = TRUE)%>%
  gather(gene,mean,-group.6,-marker.cluster)%>%
  dplyr::select(-marker.cluster)%>%
  spread(group.6,mean)%>%
  mutate(delta=group6-rest,FC=group6/rest)

test.group4<-cbind(data.frame(marker.cluster=res.mark$cluster),t(marker.genes[,7:58]))%>%
  mutate(group.6=ifelse(marker.cluster==6,"group6","rest"),group.4=ifelse(marker.cluster==4,"group4","rest"))%>%
  group_by(group.4)%>%  
  summarise_if(is.numeric, mean, na.rm = TRUE)%>%
  gather(gene,mean,-group.4,-marker.cluster)%>%
  dplyr::select(-marker.cluster)%>%
  spread(group.4,mean)%>%
  mutate(delta=group4-rest,FC=group4/rest)




VLPO.gene<-data.frame(v.norm$E)%>%
  rownames_to_column(var="id")%>%
  left_join(x$genes,.,by="id")%>%
  left_join(data.frame(geneSymbol=PCAmarker[,6]),.,by="geneSymbol")%>% 
  filter(!is.na(id))%>%
  distinct(id,.keep_all=T)%>%
  column_to_rownames(var="geneSymbol")

VLPO.pca <- PCA(t(as.matrix(VLPO.gene[,7:58])), graph =F)
fviz_eig(VLPO.pca)

fviz_pca_ind(VLPO.pca, axes=c(1,2),
             label = "ind", # individual labels
             repel = F,
             habillage = x.norm$samples$sleep, # color by groups
             palette = c("#00AFBB", "blue", "#FC4E07"),
             addEllipses = F, # Concentration ellipses,
             mean.point=F
             )

fviz_pca_biplot(VLPO.pca, repel = TRUE,
                col.var = "#2E9FDF", # Variables color
                col.ind = "#696969",  # Individuals color
                label = "var"
                )

fviz_contrib(VLPO.pca, choice = "var", axes = 2, top = 6) # %contribution to Dim3 from top10 variables


#k-means
set.seed(123)
km.res <- kmeans(scale(t(as.matrix(VLPO.gene[,7:58]))), 4, nstart = 1)

fviz_cluster(km.res, data = scale(t(as.matrix(VLPO.gene[,7:58]))),
             palette = c("#00AFBB","#2E9FDF", "#E7B800", "#FC4E07"),
             ggtheme = theme_minimal(),
             main = "Partitioning Clustering Plot"
             )

#Hierarchical clustering
res <- hcut(t(as.matrix(VLPO.gene[,7:58])), k = 5, stand = TRUE)
# Visualize
fviz_dend(res, rect = TRUE, cex = 0.5,
          k_colors = c("#00AFBB","violetred","#2E9FDF", "#E7B800", "#FC4E07"))

#six genes (fos, gal, Cck, Crh, Tac1, Pdyn)
select.gene<-data.frame(v.norm$E)%>%
  rownames_to_column(var="id")%>%
  left_join(x$genes,.,by="id")%>%
  left_join(data.frame(geneSymbol=PCAmarker[,5]),.,by="geneSymbol")%>% 
  filter(!is.na(id))%>%
  distinct(geneSymbol,.keep_all=T)%>%
  column_to_rownames(var="geneSymbol")

#Hierarchical clustering
res.sel<- hcut(t(as.matrix(select.gene[,7:58])), k = 6, stand = TRUE)
# Visualize
png(filename = "slectgenes.HC.png",height =5,width =10,res = 300,units = "in",type="cairo")

fviz_dend(res.sel, rect = TRUE, cex = 0.5,
          k_colors = c("#00AFBB","green","violetred","#2E9FDF", "#E7B800", "#FC4E07"))
dev.off()


png(filename = "slectgenes.scaled.png",height =10,width =4,res = 300,units = "in",type="cairo")

cbind(data.frame(marker.cluster=res.sel$cluster),t(select.gene[,7:58]))%>%
  group_by(marker.cluster)%>%
  summarise_if(is.numeric, mean, na.rm = TRUE)%>%
  mutate(cluster="cluster")%>%
  unite(mark,c(cluster,marker.cluster))%>%
  gather(gene,mean,-mark)%>%
  spread(mark,mean)%>%
  column_to_rownames(var="gene")%>%
  as.matrix()%>%
  pheatmap(.,scale="row",treeheight_col=10)

dev.off()

png(filename = "slectgenes.png",height =10,width =4,res = 300,units = "in",type="cairo")

cbind(data.frame(marker.cluster=res.sel$cluster),t(select.gene[,7:58]))%>%
  group_by(marker.cluster)%>%
  summarise_if(is.numeric, mean, na.rm = TRUE)%>%
  mutate(cluster="cluster")%>%
  unite(mark,c(cluster,marker.cluster))%>%
  gather(gene,mean,-mark)%>%
  spread(mark,mean)%>%
  column_to_rownames(var="gene")%>%
  as.matrix()%>%
  pheatmap(.,scale="none",treeheight_col=10)

dev.off()




set.seed(123)
km.res.select <- kmeans(scale(t(as.matrix(select.gene[,7:58]))), 2, nstart = 1)

fviz_cluster(km.res.select, data = scale(t(as.matrix(select.gene[,7:58]))),
             palette = c("#00AFBB","#E7B800"),
             ggtheme = theme_minimal(),
             main = "Partitioning Clustering Plot"
             )

#Hierarchical clustering
res.select <- hcut(t(as.matrix(select.gene[,7:58])), k = 3, stand = TRUE)
# Visualize
fviz_dend(res.select, rect = TRUE, cex = 0.5,
          k_colors = c("#00AFBB","#E7B800", "#FC4E07"))


select.pca <- PCA(t(as.matrix(select.gene[,7:58])), graph =F)
fviz_eig(select.pca)

fviz_pca_ind(select.pca, axes=c(1,2),
             label = "ind", # individual labels
             repel = F,
             habillage = x.norm$samples$sleep, # color by groups
             palette = c("#00AFBB", "blue", "#FC4E07"),
             addEllipses = F, # Concentration ellipses,
             mean.point=F
             )

fviz_pca_biplot(select.pca, repel = TRUE,
                col.var = "#2E9FDF", # Variables color
                col.ind = "#696969",  # Individuals color
                label = "var"
                )

fviz_contrib(select.pca, choice = "var", axes = 1, top = 6) # %contribution to Dim3 from top10 variables


#combine cluster information from VLPO.gene and select.gene

test<-cbind(data.frame(VLPO.cluster=res$cluster),data.frame(select.cluster=res.select$cluster),t(VLPO.gene[,7:58]))

png(filename = "VLPOgenes.png",height =20,width =4,res = 300,units = "in",type="cairo")

cbind(data.frame(VLPO.cluster=res$cluster),t(VLPO.gene[,7:58]))%>%
  group_by(VLPO.cluster)%>%
  summarise_if(is.numeric, mean, na.rm = TRUE)%>%
  mutate(cluster="cluster")%>%
  unite(VLPO,c(cluster,VLPO.cluster))%>%
  gather(gene,mean,-VLPO)%>%
  spread(VLPO,mean)%>%
  column_to_rownames(var="gene")%>%
  as.matrix()%>%
  pheatmap(.,scale = "row",treeheight_col=10)

dev.off()

,data.frame(select.cluster=res.select$cluster)



png(filename = "selectgenes.png",height =4,width =4,res = 300,units = "in",type="cairo")

cbind(data.frame(select.cluster=res.select$cluster),t(select.gene[,7:58]))%>%
  group_by(select.cluster)%>%
  summarise_if(is.numeric, mean, na.rm = TRUE)%>%
  mutate(cluster="cluster")%>%
  unite(sel,c(cluster,select.cluster))%>%
  gather(gene,mean,-sel)%>%
  spread(sel,mean)%>%
  column_to_rownames(var="gene")%>%
  as.matrix()%>%
  pheatmap(.,scale = "row",treeheight_col=10)
dev.off()



NP.gene<-data.frame(v.norm$E)%>%
  rownames_to_column(var="id")%>%
  left_join(x$genes,.,by="id")%>%
  left_join(data.frame(geneSymbol=PCAmarker[,1]),.,by="geneSymbol")%>% 
  filter(!is.na(id))%>%
  distinct(id,.keep_all=T)%>%
  column_to_rownames(var="geneSymbol")

NP.pca <- PCA(t(as.matrix(NP.gene[,7:58])), graph =F)
fviz_eig(NP.pca)

fviz_pca_ind(NP.pca, axes=c(1,2),
             label = "ind", # individual labels
             repel = TRUE,
             habillage = x.norm$samples$sleep, # color by groups
             palette = c("#00AFBB", "blue", "#FC4E07"),
             addEllipses = F, # Concentration ellipses,
             mean.point=F
             )

fviz_contrib(NP.pca, choice = "var", axes = 4, top = 10) # %contribution to Dim3 from top10 variables

NP.gene.t<-NP.gene[,7:58]%>%
  rownames_to_column(var="geneSymbol")%>%
  gather(sample,CPM,-geneSymbol)%>%
  spread(geneSymbol,CPM)%>%
  mutate(sleep=ifelse(grepl("SS",sample),"SS",ifelse(grepl("SD",sample),"SD","ZT0")))%>%
  mutate(sleep=factor(sleep,levels=c("ZT0","SS","SD")))%>%
  separate(sample,c("well","con"))
  
  
ggscatter(NP.gene.t,x="Slc31a1",y="Slc17a6",color="sleep",palette = c("#00AFBB", "blue", "#FC4E07"),label="well",repel=T)


TF.gene<-data.frame(v.norm$E)%>%
  rownames_to_column(var="id")%>%
  left_join(x$genes,.,by="id")%>%
  left_join(data.frame(geneSymbol=PCAmarker[,2]),.,by="geneSymbol")%>% 
  filter(!is.na(id))%>%
  distinct(geneSymbol,.keep_all=T)%>%
  column_to_rownames(var="geneSymbol")

TF.pca <- PCA(t(as.matrix(TF.gene[,7:58])), graph =F)
fviz_eig(TF.pca)

fviz_pca_ind(TF.pca, axes=c(1,2),
             label = "ind", # individual labels
             repel = F,
             habillage = x.norm$samples$sleep, # color by groups
             palette = c("#00AFBB", "blue", "#FC4E07"),
             addEllipses = F, # Concentration ellipses,
             mean.point=F
             )

fviz_contrib(TF.pca, choice = "var", axes = 4, top = 10) # %contribution to Dim3 from top10 variables

TF.gene.t<-TF.gene[,7:58]%>%
  rownames_to_column(var="geneSymbol")%>%
  gather(sample,CPM,-geneSymbol)%>%
  spread(geneSymbol,CPM)%>%
  mutate(sleep=ifelse(grepl("SS",sample),"SS",ifelse(grepl("SD",sample),"SD","ZT0")))%>%
  mutate(sleep=factor(sleep,levels=c("ZT0","SS","SD")))%>%
  separate(sample,c("well","con"))
  
  
ggscatter(TF.gene.t,x="Fos",y="Gabpa",color="sleep",palette = c("#00AFBB", "blue", "#FC4E07"),label="well",repel=T)


SS.gene<-data.frame(v.norm$E)%>%
  rownames_to_column(var="id")%>%
  left_join(x$genes,.,by="id")%>%
  left_join(data.frame(geneSymbol=PCAmarker[,4]),.,by="geneSymbol")%>% 
  filter(!is.na(id))%>%
  distinct(geneSymbol,.keep_all=T)%>%
  column_to_rownames(var="geneSymbol")

SS.pca <- PCA(t(as.matrix(SS.gene[,7:58])), graph =F)
fviz_eig(SS.pca)

fviz_pca_ind(SS.pca, axes=c(1,2),
             label = "ind", # individual labels
             repel = F,
             habillage = x.norm$samples$sleep, # color by groups
             palette = c("#00AFBB", "blue", "#FC4E07"),
             addEllipses = F, # Concentration ellipses,
             mean.point=F
             )

fviz_contrib(SS.pca, choice = "var", axes = 2, top = 10) # %contribution to Dim3 from top10 variables

SS.gene.t<-SS.gene[,7:58]%>%
  rownames_to_column(var="geneSymbol")%>%
  gather(sample,CPM,-geneSymbol)%>%
  spread(geneSymbol,CPM)%>%
  mutate(sleep=ifelse(grepl("SS",sample),"SS",ifelse(grepl("SD",sample),"SD","ZT0")))%>%
  mutate(sleep=factor(sleep,levels=c("ZT0","SS","SD")))%>%
  separate(sample,c("well","con"))
  
  
ggscatter(SS.gene.t,x="Rbm3",y="Cirbp",color="sleep",palette = c("#00AFBB", "blue", "#FC4E07"),label="well",repel=T)


SDvSS.gene<-data.frame(v.norm$E)%>%
  rownames_to_column(var="id")%>%
  left_join(x$genes,.,by="id")%>%
  left_join(data.frame(geneSymbol=PCAmarker[,4]),.,by="geneSymbol")%>% 
  filter(!is.na(id))%>%
  distinct(geneSymbol,.keep_all=T)%>%
  column_to_rownames(var="geneSymbol")

SDvSS.pca <- PCA(t(as.matrix(SDvSS.gene[,7:58])), graph =F)
fviz_eig(SDvSS.pca)

fviz_pca_ind(SDvSS.pca, axes=c(1,2),
             label = "ind", # individual labels
             repel = F,
             habillage = x.norm$samples$sleep, # color by groups
             palette = c("#00AFBB", "blue", "#FC4E07"),
             addEllipses = F, # Concentration ellipses,
             mean.point=F
             )

fviz_contrib(SDvSS.pca, choice = "var", axes = 1, top = 40) # %contribution to Dim3 from top10 variables

SDvSS.gene.t<-SDvSS.gene[,7:58]%>%
  rownames_to_column(var="geneSymbol")%>%
  gather(sample,CPM,-geneSymbol)%>%
  spread(geneSymbol,CPM)%>%
  mutate(sleep=ifelse(grepl("SS",sample),"SS",ifelse(grepl("SD",sample),"SD","ZT0")))%>%
  mutate(sleep=factor(sleep,levels=c("ZT0","SS","SD")))%>%
  separate(sample,c("well","con"))
  
PCA.gene.combine<-SDvSS.gene.t%>%
  left_join(.,NP.gene.t,by="well")%>%
  left_join(.,TF.gene.t,by="well")

ggscatter(PCA.gene.combine,x="Cck",y="Gal",color="sleep",palette = c("#00AFBB", "blue", "#FC4E07"),label="well",repel=T)


ggscatter(PCA.gene.combine,x="Slc17a6",y="Fos",color="sleep",palette = c("#00AFBB", "blue", "#FC4E07"),label="well",repel=T)



#plotMDS

install.packages("scatterplot3d") # Install
library("scatterplot3d") # load


mds<-plotMDS(v.norm$E,ndim=3)
toplot<-data.frame(Dim1=mds$cmdscale.out[,1],Dim3=mds$cmdscale.out[,3],Dim2=mds$cmdscale.out[,2],Group = factor(x.norm$samples$sleep.time),sleep=factor(x.norm$samples$sleep),time=factor(x.norm$samples$time))


ggscatter(toplot, x="Dim1",y="Dim2",color="sleep",shape="time",size=3.5, palette=brewer.pal(3,"Set1"),ellipse=T, ellipse.type = "convex",ellipse.alpha = 0,xlab="MDS dim1", ylab="MDS dim2",title="2B") +
theme(legend.position="none",panel.background = element_rect(fill="white", linetype="solid", color="black"), plot.title = element_text(face="bold",hjust=-0.14))

palette<-brewer.pal(9, "Set1")[c(7,1,2,3,4)]
colors <- palette[as.numeric(toplot$time)]

shapes = c(10,17,2) 
shapes <- shapes[as.numeric(toplot$sleep)]


scatterplot3d(toplot[,1:3], pch = shapes, cex.symbols=1.5,color=colors, grid=F,angle = 60, cex.lab=1.5,xlab = "MDS dim1",ylab = "MDS dim3",zlab = "MDS dim2",mar=c(3,2,1,3))


res.pca <- PCA(t(v.norm$E), graph =F)
fviz_eig(res.pca)

fviz_pca_ind(res.pca,
             col.ind = "cos2", # Color by the quality of representation
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping

)

fviz_pca_ind(res.pca, axes=c(1,3),
             label = "none", # hide individual labels
             habillage = x.norm$samples$sleep, # color by groups
             palette = c("#00AFBB", "#E7B800", "#FC4E07"),
             addEllipses = T, # Concentration ellipses,
             mean.point=F
             )

fviz_contrib(res.pca, choice = "var", axes = 3, top = 10) # %contribution to Dim3 from top10 variables


```











data(decathlon2)
decathlon2.active <- decathlon2[1:23, 1:10]
res.pca <- prcomp(decathlon2.active, scale = TRUE)# use prcomp 
fviz_eig(res.pca) #Visualize eigenvalues

res.pca <- PCA(decathlon2.active, graph = FALSE)

res.pca <- PCA(t(assay(vsd)), graph = FALSE)
fviz_eig(res.pca)

fviz_pca_ind(res.pca,
             col.ind = "cos2", # Color by the quality of representation
             gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"),
             repel = TRUE     # Avoid text overlapping
             )


fviz_contrib(res.pca, choice = "var", axes = 1, top = 7) # %contribution to Dim1 from top7 variables
head(res.pca$ind$cos2,12) #Cos2 of individuals (quality of the representation) on all dims (by default PCA calculates 5 dimensions) 
fviz_contrib(res.pca, choice = "ind", axes = 1) # %contribution to Dim1 of all individuals
fviz_contrib(res.pca, choice = "ind", axes = 2) # %contribution to Dim2 of all individuals




useful resources:
1. scRNA-seq course from UCambridge
https://hemberg-lab.github.io/scRNA.seq.course/

2. papers

3. GMO data
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE125428 
RNA-seq experiment on five mouse brain regions--hypothalamus, hippocampus, neocortex, cerebellum, and the amygdala --comparing expression in virgin females to primiparous females three weeks after pup weaning
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE125441
REM sleep's unique associations with corticosterone regulation, apoptotic pathways and behavior in chronic stress in mice
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE125065
Single cell transcriptomic analysis of the lateral hypothalamic area reveals molecularly distinct populations of inhibitory and excitatory neurons
    Mickelsen LE, Bolisetty M, Chimileski BR, Fujita A et al. Single-cell transcriptomic analysis of the lateral hypothalamic area reveals molecularly distinct populations of inhibitory and excitatory neurons. Nat Neurosci 2019 Mar 11. PMID: 30858605
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE113576   
Molecular, Spatial and Functional Single-Cell Profiling of the Hypothalamic Preoptic Region
Moffitt JR, Bambah-Mukku D, Eichhorn SW, Vaughn E et al. Molecular, spatial, and functional single-cell profiling of the hypothalamic preoptic region. Science 2018 Nov 16;362(6416). PMID: 30385464
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE79108
  Chung S, Weber F, Zhong P, Tan CL et al. Identification of preoptic sleep neurons using retrograde labelling and gene profiling. Nature 2017 May 25;545(7655):477-481. PMID: 28514446

age related
https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE99791
astrocyte

Using the GEOquery Package
https://www.bioconductor.org/packages/devel/bioc/vignettes/GEOquery/inst/doc/GEOquery.html


if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("DESeq2", version = "3.8")

source("https://bioc.ism.ac.jp/biocLite.R")
biocLite("apeglm")

browseVignettes("DESeq2")
http://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html#time-course-experiments
8 Removing hidden batch effects

    8.1 Using SVA with DESeq2
    8.2 Using RUV with DESeq2

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("Rsamtools", version = "3.8")


https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf

```{r}
counts.dupZT0<-counts.raw%>%
   rename(B2.SS0hr=B2.CTL0hr,B11.SS0hr=B11.CTL0hr,C10.SS0hr=C10.CTL0hr,D9.SS0hr=D9.CTL0hr,E8.SS0hr=E8.CTL0hr,F7.SS0hr=F7.CTL0hr)%>%
   mutate(B2.SD0hr=B2.SS0hr,B11.SD0hr=B11.SS0hr,C10.SD0hr=C10.SS0hr,D9.SD0hr=D9.SS0hr,E8.SD0hr=E8.SS0hr,F7.SD0hr=F7.SS0hr)%>%
  dplyr::select(-starts_with("gene"))%>%
column_to_rownames(var="id")%>%
  as.matrix(.)

sample.dupZT0<-data.frame(sampleID=colnames(counts.dupZT0))%>%
  mutate(sleep=factor(ifelse(grepl("SS",sampleID),"SS","SD"),levels=c("SS","SD")),
         time=factor(ifelse(grepl("0hr",sampleID),"Z0",
                            ifelse(grepl("3hr",sampleID),"3hr",
                            ifelse(grepl("6hr",sampleID),"6hr",
                            ifelse(grepl("9hr",sampleID),"9hr","12hr")))),
                            levels=c("Z0","3hr","6hr","9hr","12hr")))%>%
  column_to_rownames(var="sampleID")


dds.dupZT0<- DESeqDataSetFromMatrix(countData = counts.dupZT0,
                              colData =sample.dupZT0,
                              design = ~ time+sleep)


keep <- rowSums(counts(dds.dupZT0)) >= 10 
dds.dupZT0 <- dds.dupZT0[keep,]#22016/37681 has a total >=10 counts across all samples

colData(dds.dupZT0)$group<-factor(paste0(sample.dupZT0$sleep,sample.dupZT0$time))
design(dds.dupZT0) <- ~ group
dds.dupZT0.deseq <- DESeq(dds.dupZT0)
resultsNames(dds.dupZT0.deseq)
summary(results(dds.dupZT0.deseq, contrast=c("group", "SD3hr", "SS3hr"),alpha=0.05))
summary(results(dds.dupZT0.deseq, contrast=c("group", "SD6hr", "SS6hr"),alpha=0.05))
summary(results(dds.dupZT0.deseq, contrast=c("group", "SD9hr", "SS9hr"),alpha=0.05))
summary(results(dds.dupZT0.deseq, contrast=c("group", "SD12hr", "SS12hr"),alpha=0.05))
summary(results(dds.dupZT0.deseq, contrast=c("group", "SDZ0", "SSZ0"),alpha=0.05))



LRT.dupZT0 <- DESeq(dds.dupZT0, test="LRT", reduced=~time)
res.LRT.dupZT0 <- results(LRT.dupZT0, alpha=0.05)
sum(res.LRT.dupZT0$padj < 0.05, na.rm=TRUE)


summary(res.LRT.dupZT0)




#test droplevel
counts.raw <-read.delim("master_list_of_gene_counts_MIN.sense.UNNORMQUANT.REORDERED.txt", header=T,stringsAsFactors=FALSE)


counts <-read.delim("master_list_of_gene_counts_MIN.sense.UNNORMQUANT.REORDERED.txt", header=T,stringsAsFactors=FALSE)%>%
dplyr::select(-starts_with("gene"))%>%
column_to_rownames(var="id")%>%
  as.matrix(.)

sample<-data.frame(sampleID=colnames(counts))%>%
  mutate(sleep=factor(ifelse(grepl("SS",sampleID),"SS",ifelse(grepl("SD",sampleID),"SD","CTL")),levels=c("SS","SD","CTL")),
         time=factor(ifelse(grepl("0hr",sampleID),"Z0",
                            ifelse(grepl("3hr",sampleID),"3hr",
                            ifelse(grepl("6hr",sampleID),"6hr",
                            ifelse(grepl("9hr",sampleID),"9hr","12hr")))),
                            levels=c("Z0","3hr","6hr","9hr","12hr")))%>%
  column_to_rownames(var="sampleID")
  
  
  
sample.rmZT0<-sample%>%
filter(sleep!="CTL")%>%
mutate(sleep=droplevels(sleep),time=droplevels(time))

m1 <- model.matrix(~ time+sleep+time:sleep, sample.rmZT0)


dds.rmZT0<- DESeqDataSetFromMatrix(countData = counts[,7:52],
                              colData = sample.rmZT0,
                              design = ~ time+sleep+time:sleep)

keep <- rowSums(counts(dds.rmZT0)) >= 10 
dds.rmZT0 <- dds.rmZT0[keep,] #21516/37681 has a total >=10 counts across all samples

LRT.rmZT0 <- DESeq(dds.rmZT0, test="LRT", reduced=~time)
res.LRT.rmZT0 <- results(LRT.rmZT0, alpha=0.05)
sum(res.LRT.rmZT0$padj < 0.05, na.rm=TRUE)

LRT.rmZT0.table<-as.data.frame(res.LRT.rmZT0)%>%
  rownames_to_column(var="id")%>%
  filter(!is.na(padj))
LRT.dupZT0.table<-as.data.frame(res.LRT.dupZT0)%>%
  rownames_to_column(var="id")%>%
  filter(!is.na(padj))

test=intersect(LRT.rmZT0.table[LRT.rmZT0.table$padj<0.05,]$id,LRT.dupZT0.table[LRT.dupZT0.table$padj<0.05,]$id)

comp.limma=intersect(LRT.rmZT0.table[LRT.rmZT0.table$padj<0.05,]$id,DEG.SDvSS$id)

comb.DEseq.limma<-DEG.2%>%
  left_join(.,LRT.rmZT0.table,by="id")%>%
  mutate(group=ifelse(adj.P.Val<0.05&padj<0.05,"both",ifelse(adj.P.Val<0.05,"limma",ifelse(padj<0.05,"DEseq","na"))))



write.csv(comb.DEseq.limma, file="comb.DEseq.limma.csv" )

#DEG change with time
counts.SS<-counts.raw%>%
   rename(B2.SS0hr=B2.CTL0hr,B11.SS0hr=B11.CTL0hr,C10.SS0hr=C10.CTL0hr,D9.SS0hr=D9.CTL0hr,E8.SS0hr=E8.CTL0hr,F7.SS0hr=F7.CTL0hr)%>%
   mutate(B2.SD0hr=B2.SS0hr,B11.SD0hr=B11.SS0hr,C10.SD0hr=C10.SS0hr,D9.SD0hr=D9.SS0hr,E8.SD0hr=E8.SS0hr,F7.SD0hr=F7.SS0hr)%>%
  dplyr::select(id,contains("SS"))%>%
column_to_rownames(var="id")%>%
  as.matrix(.)

sample.time<-data.frame(sampleID=colnames(counts.SS))%>%
  mutate(time=factor(ifelse(grepl("0hr",sampleID),"Z0",
                            ifelse(grepl("3hr",sampleID),"3hr",
                            ifelse(grepl("6hr",sampleID),"6hr",
                            ifelse(grepl("9hr",sampleID),"9hr","12hr")))),
                            levels=c("Z0","3hr","6hr","9hr","12hr")))%>%
  column_to_rownames(var="sampleID")


dds.SS<- DESeqDataSetFromMatrix(countData = counts.SS,
                              colData =sample.time,
                              design = ~ time)


keep <- rowSums(counts(dds.SS)) >= 10 
dds.SS<- dds.SS[keep,]#20383/37681 has a total >=10 counts across all samples

dds.SS.deseq <- DESeq(dds.SS)
resultsNames(dds.SS.deseq)

summary(results(dds.SS.deseq, name="time_3hr_vs_Z0",alpha=0.05))
summary(results(dds.SS.deseq, name="time_6hr_vs_Z0",alpha=0.05))
summary(results(dds.SS.deseq, name="time_9hr_vs_Z0",alpha=0.05))
summary(results(dds.SS.deseq, name="time_12hr_vs_Z0",alpha=0.05))


LRT.time <- DESeq(dds.SS, test="LRT", reduced=~1)
res.LRT.time <- results(LRT.time, alpha=0.05)
sum(res.LRT.time$padj < 0.1, na.rm=TRUE)
summary(res.LRT.time)





library(biomaRt)

mouseensembl<-useMart(host = "aug2017.archive.ensembl.org", biomart="ensembl", dataset="mmusculus_gene_ensembl")

att<-listAttributes(mouseensembl)

#add more gene identifiers
count_gene<-getBM(attributes = c("ensembl_gene_id","entrezgene","description","gene_biotype","transcript_length"),
                mart=mouseensembl)%>%
  dplyr::rename(.,id=ensembl_gene_id)%>%
  dplyr::rename(.,EntrezID=entrezgene)%>%
  distinct(id,.keep_all = TRUE)%>%
 left_join(counts.raw[,c(1,54,55)],.,by="id") 


mcols(dds) <- DataFrame(mcols(dds), count_gene)

keep <- rowSums(counts(dds)) >= 10 
dds <- dds[keep,]#22016/37681 has a total >=10 counts across all samples

degs <- DESeq(dds)
res <- results(degs)
summary(res)
sum(res$padj < 0.05, na.rm=TRUE)

#LRT

degs.LRT <- DESeq(dds, test="LRT", reduced=~time)
res.LRT <- results(degs.LRT, alpha=0.05)
sum(res.LRT$padj < 0.05, na.rm=TRUE)

write.csv( as.data.frame(res.LRT), file="noZT0.fullmodel.396.csv" )


summary(res.LRT)

head(res.LRT[order(res.LRT$padj),], 10)




#Use DEseq2 

counts <-read.delim("master_list_of_gene_counts_MIN.sense.UNNORMQUANT.REORDERED.txt", header=T,stringsAsFactors=FALSE)

counts.addZT0<-counts%>%
  mutate(B2.SD0hr=B2.CTL0hr,B11.SD0hr=B11.CTL0hr,C10.SD0hr=C10.CTL0hr,D9.SD0hr=D9.CTL0hr,E8.SD0hr=E8.CTL0hr,F7.SD0hr=F7.CTL0hr)%>%
  rename(B2.SS0hr=B2.CTL0hr,B11.SS0hr=B11.CTL0hr,C10.SS0hr=C10.CTL0hr,D9.SS0hr=D9.CTL0hr,E8.SS0hr=E8.CTL0hr,F7.SS0hr=F7.CTL0hr)%>%
  dplyr::select(-starts_with("gene"))%>%
column_to_rownames(var="id")%>%
  as.matrix(.)

count_sample<-data.frame(sampleID=colnames(counts.addZT0))%>%
  mutate(sleep=factor(ifelse(grepl("SS",sampleID),"SS","SD"),levels=c("SS","SD")),
         time=factor(ifelse(grepl("0hr",sampleID),"Z0",
                            ifelse(grepl("3hr",sampleID),"3hr",
                            ifelse(grepl("6hr",sampleID),"6hr",
                            ifelse(grepl("9hr",sampleID),"9hr","12hr")))),
                            levels=c("Z0","3hr","6hr","9hr","12hr")))%>%
  column_to_rownames(var="sampleID")

table(colnames(counts.addZT0)==rownames(count_sample)) #check if colnames of the count matrix is the same as the rownames of the sample information


library(biomaRt)

mouseensembl<-useMart(host = "aug2017.archive.ensembl.org", biomart="ensembl", dataset="mmusculus_gene_ensembl")

att<-listAttributes(mouseensembl)

#add more gene identifiers
count_gene<-getBM(attributes = c("ensembl_gene_id","entrezgene","description","gene_biotype","transcript_length"),
                mart=mouseensembl)%>%
  dplyr::rename(.,id=ensembl_gene_id)%>%
  dplyr::rename(.,EntrezID=entrezgene)%>%
  distinct(id,.keep_all = TRUE)%>%
 left_join(counts[,c(1,54,55)],.,by="id") 

library("DESeq2")
dds <- DESeqDataSetFromMatrix(countData = counts.addZT0,
                              colData = count_sample,
                              design = ~ time+sleep)

mcols(dds) <- DataFrame(mcols(dds), count_gene)

```


Pre-filtering

While it is not necessary to pre-filter low count genes before running the DESeq2 functions, there are two reasons which make pre-filtering useful: by removing rows in which there are very few reads, we reduce the memory size of the dds data object, and we increase the speed of the transformation and testing functions within DESeq2. Here we perform a minimal pre-filtering to keep only rows that have at least 10 reads total. Note that more strict filtering to increase power is automatically applied via independent filtering on the mean of normalized counts within the results function.


```{r}
keep <- rowSums(counts(dds)) >= 10 
dds <- dds[keep,]#22016/37681 has a total >=10 counts across all samples

degs <- DESeq(dds)
res <- results(degs)
summary(res)
sum(res$padj < 0.05, na.rm=TRUE)



counts.SSZT0<-counts%>%
   rename(B2.SS0hr=B2.CTL0hr,B11.SS0hr=B11.CTL0hr,C10.SS0hr=C10.CTL0hr,D9.SS0hr=D9.CTL0hr,E8.SS0hr=E8.CTL0hr,F7.SS0hr=F7.CTL0hr)%>%
  dplyr::select(-starts_with("gene"))%>%
column_to_rownames(var="id")%>%
  as.matrix(.)

sample.SSZT0<-data.frame(sampleID=colnames(counts.SSZT0))%>%
  mutate(sleep=factor(ifelse(grepl("SS",sampleID),"SS","SD"),levels=c("SS","SD")),
         time=factor(ifelse(grepl("0hr",sampleID),"Z0",
                            ifelse(grepl("3hr",sampleID),"3hr",
                            ifelse(grepl("6hr",sampleID),"6hr",
                            ifelse(grepl("9hr",sampleID),"9hr","12hr")))),
                            levels=c("Z0","3hr","6hr","9hr","12hr")))%>%
  column_to_rownames(var="sampleID")

table(colnames(counts.SSZT0)==rownames(sample.SSZT0)) #check if colnames of the count matrix is the same as the rownames of the sample information

m1 <- model.matrix(~ time+sleep, sample.SSZT0)



dds.SSZT0 <- DESeqDataSetFromMatrix(countData = counts.SSZT0,
                              colData = sample.SSZT0,
                              design = ~ time+sleep)

mcols(dds.SSZT0) <- DataFrame(mcols(dds.SSZT0), count_gene)



keep.2 <- rowSums(counts(dds.SSZT0)) >= 10 
dds.SSZT0 <- dds.SSZT0[keep.2,]#21788/37681 has a total >=10 counts across all samples(without duplicate ZT0 samples)

degs.2 <- DESeq(dds.SSZT0 )

vsd <- vst(dds.SSZT0)

plotPCA(vsd, intgroup = c("sleep"))

sampleDists <- dist(t(assay(vsd)))

sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( vsd$sleep, vsd$time, sep = " - " )
colnames(sampleDistMatrix) <- NULL
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
pheatmap(sampleDistMatrix,
         clustering_distance_rows = sampleDists,
         clustering_distance_cols = sampleDists,
         col = colors)



res.2 <- results(degs.2, alpha=0.05)
summary(res.2)
resultsNames(degs.2)
resLFC <- lfcShrink(degs.2, coef="sleep_SD_vs_SS", type="apeglm")


plotMA(resLFC, ylim=c(-1,1))

```
```{r}
counts.ZT0<-counts%>%
  dplyr::select(-starts_with("gene"))%>%
column_to_rownames(var="id")%>%
  as.matrix(.)

sample.ZT0<-data.frame(sampleID=colnames(counts.ZT0))%>%
  mutate(sleep=factor(ifelse(grepl("SS",sampleID),"SS",ifelse(grepl("SD",sampleID),"SD","ZT0")),levels=c("ZT0","SS","SD")),
         time=factor(ifelse(grepl("0hr",sampleID),"Z0",
                            ifelse(grepl("3hr",sampleID),"3hr",
                            ifelse(grepl("6hr",sampleID),"6hr",
                            ifelse(grepl("9hr",sampleID),"9hr","12hr")))),
                            levels=c("Z0","3hr","6hr","9hr","12hr")))%>%
  column_to_rownames(var="sampleID")

table(colnames(counts.SSZT0)==rownames(sample.SSZT0)) #check if colnames of the count matrix is the same as the rownames of the sample information

m1 <- model.matrix(~ time+sleep, sample.ZT0)

dds.ZT0 <- DESeqDataSetFromMatrix(countData = counts.ZT0,
                              colData = sample.ZT0,
                              design = ~ sleep)

vsd <- vst(dds.ZT0)
rld <- rlog(dds.ZT0)

plotPCA(vsd, intgroup = c("sleep","time"))
plotPCA(rld, intgroup = c("sleep","time"))

pcaData <- plotPCA(vsd, intgroup = c("sleep","time"), returnData=TRUE)
percentVar <- round(100 * attr(pcaData, "percentVar"))
ggplot(pcaData, aes(PC1, PC2, color=time, shape=sleep)) +
  geom_point(size=3) +
  xlab(paste0("PC1: ",percentVar[1],"% variance")) +
  ylab(paste0("PC2: ",percentVar[2],"% variance")) + 
  coord_fixed()


```

```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("ReportingTools", version = "3.8")

browseVignettes("ReportingTools")

```

The key point to remember about designs with interaction terms is that, unlike for a design ~genotype + condition, where the condition effect represents the overall effect controlling for differences due to genotype, by adding genotype:condition, the main condition effect only represents the effect of condition for the reference level of genotype (I, or whichever level was defined by the user as the reference level). The interaction terms genotypeII.conditionB and genotypeIII.conditionB give the difference between the condition effect for a given genotype and the condition effect for the reference genotype.


Likelihood ratio test

DESeq2 offers two kinds of hypothesis tests: the Wald test, where we use the estimated standard error of a log2 fold change to test if it is equal to zero, and the likelihood ratio test (LRT). The LRT examines two models for the counts, a full model with a certain number of terms and a reduced model, in which some of the terms of the full model are removed. The test determines if the increased likelihood of the data using the extra terms in the full model is more than expected if those extra terms are truly zero.

The LRT is therefore useful for testing multiple terms at once, for example testing 3 or more levels of a factor at once, or all interactions between two variables. The LRT for count data is conceptually similar to an analysis of variance (ANOVA) calculation in linear regression, except that in the case of the Negative Binomial GLM, we use an analysis of deviance (ANODEV), where the deviance captures the difference in likelihood between a full and a reduced model.

The likelihood ratio test can be performed by specifying test="LRT" when using the DESeq function, and providing a reduced design formula, e.g. one in which a number of terms from design(dds) are removed. The degrees of freedom for the test is obtained from the difference between the number of parameters in the two models. A simple likelihood ratio test, if the full design was ~condition would look like:

dds <- DESeq(dds, test="LRT", reduced=~1)
res <- results(dds)

If the full design contained other variables, such as a batch variable, e.g. ~batch + condition then the likelihood ratio test would look like:

dds <- DESeq(dds, test="LRT", reduced=~batch)
res <- results(dds)

When you use the LRT, using the 'name' argument only changes the columns of the results which have to do with the LFC's. For the current release of DESeq2 (v1.4), when you use the 'contrast' argument, the p-values are replaced with Wald test p-values. In the development branch, I am working on having more specific control of whether or not to replace the LRT p-values with Wald p-values. For now, you can build whatever results table you like by combining the LFC columns for all contrasts with the LRT statistic, p-values and adjusted p-value columns.

```{r}

counts.woZT0<-counts%>%
   dplyr::select(-ends_with("0hr"),-starts_with("gene"))%>%
column_to_rownames(var="id")%>%
  as.matrix(.)

sample.woZT0<-data.frame(sampleID=colnames(counts.woZT0))%>%
  mutate(sleep=factor(ifelse(grepl("SS",sampleID),"SS","SD"),levels=c("SS","SD")),
         time=factor(ifelse(grepl("3hr",sampleID),"3hr",
                            ifelse(grepl("6hr",sampleID),"6hr",
                            ifelse(grepl("9hr",sampleID),"9hr","12hr"))),
                            levels=c("3hr","6hr","9hr","12hr")))%>%
  column_to_rownames(var="sampleID")

table(colnames(counts.woZT0)==rownames(sample.woZT0)) #check if colnames of the count matrix is the same as the rownames of the sample information

dds.woZT0 <- DESeqDataSetFromMatrix(countData = counts.woZT0,
                              colData = sample.woZT0,
                              design = ~ time+sleep)

mcols(dds.woZT0) <- DataFrame(mcols(dds.woZT0), count_gene)

keep.3 <- rowSums(counts(dds.woZT0)) >= 10 
dds.woZT0 <- dds.woZT0[keep.3,]#21516/37681 has a total >=10 counts across all samples(without ZT0 samples)

degs.3 <- DESeq(dds.woZT0)
resultsNames(degs.3)


res.3 <- results(degs.3, alpha=0.05)
summary(res.3)
head(res.3[order(res.3$padj),], 10)


#LRT

degs3.LRT <- DESeq(dds.woZT0, test="LRT", reduced=~time)
res3.LRT <- results(degs3.LRT, alpha=0.05)
summary(res3.LRT)

head(res3.LRT[order(res3.LRT$padj),], 10)


counts.SSZT0<-counts%>%
   rename(B2.SS0hr=B2.CTL0hr,B11.SS0hr=B11.CTL0hr,C10.SS0hr=C10.CTL0hr,D9.SS0hr=D9.CTL0hr,E8.SS0hr=E8.CTL0hr,F7.SS0hr=F7.CTL0hr)%>%
  dplyr::select(-starts_with("gene"))%>%
column_to_rownames(var="id")%>%
  as.matrix(.)

sample.SSZT0<-data.frame(sampleID=colnames(counts.SSZT0))%>%
  mutate(sleep=factor(ifelse(grepl("SS",sampleID),"SS","SD"),levels=c("SS","SD")),
         time=factor(ifelse(grepl("0hr",sampleID),"Z0",
                            ifelse(grepl("3hr",sampleID),"3hr",
                            ifelse(grepl("6hr",sampleID),"6hr",
                            ifelse(grepl("9hr",sampleID),"9hr","12hr")))),
                            levels=c("Z0","3hr","6hr","9hr","12hr")))%>%
  unite(sleep.time,sleep,time,sep=".",remove=F)%>%
  column_to_rownames(var="sampleID")

table(colnames(counts.SSZT0)==rownames(sample.SSZT0)) #check if colnames of the count matrix is the same as the rownames of the sample information

m1 <- model.matrix(~ time+sleep, sample.SSZT0)



dds.SSZT0 <- DESeqDataSetFromMatrix(countData = counts.SSZT0,
                              colData = sample.SSZT0,
                              design = ~ time+sleep)

mcols(dds.SSZT0) <- DataFrame(mcols(dds.SSZT0), count_gene)


keep.2 <- rowSums(counts(dds.SSZT0)) >= 10 
dds.SSZT0 <- dds.SSZT0[keep.2,]#21788/37681 has a total >=10 counts across all samples(without duplicate ZT0 samples)



degs2.LRT<-DESeq(dds.SSZT0, test="LRT", reduced=~time)
res2.LRT <- results(degs2.LRT, alpha=0.05)
summary(res2.LRT)
dt.deseq2 <- as.numeric(res2.LRT$padj<0.05)


glMDPlot(res2.LRT, counts=counts(dds.SSZT0), status=dt.deseq2,main="LRT.SSZT0", groups=sample.SSZT0$sleep.time, transform=FALSE,
samples=colnames(counts.SSZT0), anno=mcols(dds.SSZT0),side.main="id")

#export result
resOrderedDF <- as.data.frame(res2.LRT[order(res2.LRT$pvalue),])

dds.pair <- DESeqDataSetFromMatrix(countData = counts.SSZT0,
                              colData = sample.SSZT0,
                              design = ~ sleep.time)

write.csv(resOrderedDF, file = "results.csv")




```
Removing hidden batch effects

Suppose we did not know that there were different cell lines involved in the experiment, only that there was treatment with dexamethasone. The cell line effect on the counts then would represent some hidden and unwanted variation that might be affecting many or all of the genes in the dataset. We can use statistical methods designed for RNA-seq from the sva package (Leek 2014) or the RUVSeq package (Risso et al. 2014) in Bioconductor to detect such groupings of the samples, and then we can add these to the DESeqDataSet design, in order to account for them.

The SVA package uses the term surrogate variables for the estimated variables that we want to account for in our analysis, while the RUV package uses the terms factors of unwanted variation with the acronym "Remove Unwanted Variation" explaining the package title. We first use SVA to find hidden batch effects and then RUV following.

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("sva", version = "3.8")

```{r}
library("sva")

dat  <- counts(degs2.LRT, normalized = TRUE)
idx  <- rowMeans(dat) > 1
dat  <- dat[idx, ]
mod  <- model.matrix(~ time+sleep, colData(degs2.LRT))
mod0 <- model.matrix(~   1, colData(degs2.LRT))
svseq <- svaseq(dat, mod, mod0, n.sv = 2)


par(mfrow = c(2, 1), mar = c(3,5,3,1))
for (i in 1:2) {
  stripchart(svseq$sv[, i] ~ degs2.LRT$time, vertical = TRUE, main = paste0("SV", i))
  abline(h = 0)
}

ddssva <- dds.SSZT0
ddssva$SV1 <- svseq$sv[,1]
ddssva$SV2 <- svseq$sv[,2]
design(ddssva) <- ~ SV1 + SV2+time + sleep

ddssva.LRT<-DESeq(ddssva, test="LRT", reduced=~ SV1 + SV2+time)
res.ddssva.LRT <- results(ddssva.LRT, alpha=0.05)
summary(res.ddssva.LRT)


dt.deseq2 <- as.numeric(res.ddssva.LRT$padj<0.05)


glMDPlot(res.ddssva.LRT, counts=counts(ddssva.LRT), status=dt.deseq2,main="LRT.SSZT0.SVA", groups=sample.SSZT0$sleep.time, transform=FALSE,
samples=colnames(counts.SSZT0), anno=mcols(dds.SSZT0)[,c(1,3,4,5)],side.main="id")


```







```{r}
save(x,x_norm,v.norm,vfit,efit.main,design,DEG.combine.all,DEG.combine.all.int,CV.DEGs,FCandGO.id,FC.FDRandFC,FC.meanFC.pairedwilcox, OvY.combine.result,c.sc.SDvSS.comb,c.sc.OvY.comb,file ="RNAseq_mPFC.RData")
load("RNAseq_mPFC.RData")

#12/3/18, needed update "Rcpp", "install.packages("Rcpp",  dependencies = TRUE, repos = "http://cran.us.r-project.org")"

#as of 3/21/18, a new un-normalized count table that only contains sense reads is created by Dimitra; 37681 genes
counts <-read.delim("master_list_of_gene_counts_MIN.sense.UNNORMQUANT.REORDERED.txt", header=T,stringsAsFactors=FALSE)

counts.addZT0<-counts%>%
  mutate(B2.SD0hr=B2.CTL0hr,B11.SD0hr=B11.CTL0hr,C10.SD0hr=C10.CTL0hr,D9.SD0hr=D9.CTL0hr,E8.SD0hr=E8.CTL0hr,F7.SD0hr=F7.CTL0hr)%>%
  rename(B2.SS0hr=B2.CTL0hr,B11.SS0hr=B11.CTL0hr,C10.SS0hr=C10.CTL0hr,D9.SS0hr=D9.CTL0hr,E8.SS0hr=E8.CTL0hr,F7.SS0hr=F7.CTL0hr)%>%
  dplyr::select(-starts_with("gene"))


library(biomaRt)

mouseensembl<-useMart(host = "aug2017.archive.ensembl.org", biomart="ensembl", dataset="mmusculus_gene_ensembl")

att<-listAttributes(mouseensembl)

#add more gene identifiers
count_gene<-getBM(attributes = c("ensembl_gene_id","entrezgene","description","gene_biotype","transcript_length"),
                mart=mouseensembl)%>%
  dplyr::rename(.,id=ensembl_gene_id)%>%
  dplyr::rename(.,EntrezID=entrezgene)%>%
  distinct(id,.keep_all = TRUE)%>%
 left_join(counts[,c(1,54,55)],.,by="id") 


#creat a data frame that contains the sample info; add more columns to describe the conditions

count_sample<-data.frame(sampleID=colnames(counts.addZT0[,2:59]))%>%
  mutate(sleep=factor(ifelse(grepl("SS",sampleID),"SS","SD"),levels=c("SS","SD")),
         time=factor(ifelse(grepl("0hr",sampleID),"Z0",
                            ifelse(grepl("3hr",sampleID),"3hr",
                            ifelse(grepl("6hr",sampleID),"6hr",
                            ifelse(grepl("9hr",sampleID),"9hr","12hr")))),
                            levels=c("Z0","3hr","6hr","9hr","12hr")))%>%
  unite(sleep.time,sleep,time,sep=".",remove=F)%>%
  mutate(sleep.time=factor(sleep.time, levels=c("SS.Z0","SS.3hr","SS.6hr","SS.9hr","SS.12hr","SD.Z0","SD.3hr","SD.6hr","SD.9hr","SD.12hr")))


x <- DGEList(counts=as.matrix(counts.addZT0[,2:59]), samples=count_sample, genes=count_gene) #creat DGElist object x

table(rowSums(x$counts==0)==58) #11874/37681 genes have zero counts across all 58 samples (duplicate zt0 samples, so 52+6=58)



cpm <- cpm(x)
lcpm <- cpm(x, log=TRUE)
#Transformations from the raw-scale to counts per million (CPM), log2-counts per million (log-CPM). Transformations use a prior count of 0.25 (default) to avoid taking the log of zero. RPKM values are just as easily calculated as CPM values using the rpkm function in edgeR if gene lengths are available.



design <- model.matrix(~0+x$samples$sleep.time)
colnames(design) <- levels(x$samples$sleep.time)

contrast <- makeContrasts(
  SDvsSS_3hr = SD.3hr-SS.3hr, 
  SDvsSS_6hr = SD.6hr-SS.6hr, 
  SDvsSS_9hr = SD.9hr-SS.9hr, 
  SDvsSS_12hr = SD.12hr-SS.12hr, levels=design)



x1<- calcNormFactors(x, method = "TMM")
v<-voom(x1,design=design,plot=TRUE)
aveexp<-rowMeans(v$E)

keep_genes<-aveexp>-1.4 #filter average CPM calculated based on the normalized data; 2^-1.4x15=5.6 (sutract the 0.5 prior counts added) is equivalent to 5 counts for a library size of 15 million reads 
x2 <- x[keep_genes,, keep.lib.sizes=FALSE] #15507 genes left; keep the original library sizes



dge <- estimateDisp(x1, design, robust=TRUE)
plotBCV(dge) #Plot the genewise biological coefficient of variation (BCV) against gene abundance (in log2 counts per million)
dge2 <- estimateDisp(x2, design, robust=TRUE)
plotBCV(dge2)
test<-as.data.frame(cbind(dge$genes$id,dge$genes$geneSymbol,dge$tagwise.dispersion,dge$AveLogCPM,aveexp))
#f0<-rowSums(x$counts==0)<108
#x.f0 <- x[f0,, keep.lib.sizes=TRUE]


x1<- calcNormFactors(x, method = "TMM")
v<-voom(x1,design=design,plot=TRUE)
aveexp<-rowMeans(v$E)

keep_genes<-aveexp>-1.4 #filter average CPM calculated based on the normalized data; 2^-1.4x15=5.6 (sutract the 0.5 prior counts added) is equivalent to 5 counts for a library size of 15 million reads 
x2 <- x[keep_genes,, keep.lib.sizes=FALSE] #14656 genes left; keep the original library sizes
x_norm<- calcNormFactors(x2, method = "TMM") 
v.norm<-voom(x_norm,design=design,plot=FALSE) #renormalize after filtering
vfit <- lmFit(v.norm, design)

efit<- eBayes(contrasts.fit(vfit, contrasts=contrast)) 

#DEGs with F-test over four timepoints FDR<1%

DEG.SDvSS<-topTable(efit,coef=1:4,n=Inf,p.value=0.05) #5158 DEG with FDR<1% for SDvSS.young animals

DEG.LCM<-topTable(efit,coef=1:4,n=Inf)%>%
    mutate(aveFC=rowMeans(.[,8:11]))

write.table(DEG.LCM,"DEG.LCM.txt",sep="\t",row.names = F)


#distribution plots before and after normalization

boxplot(cpm(x,log=T)[,55:108],ylab="Log2 counts per million",main="logCPMs_unnormalized",col=c(rep("black",6),rep("cyan",6),rep("cyan4",6),rep("aquamarine",6),rep("aquamarine4",6),rep("coral",6),rep("coral4",6),rep("hotpink",6),rep("hotpink4",6)))
abline(h=mean(cpm(x,log=T)),col="blue")

boxplot(v.norm$E[,55:108],ylab="Log2 counts per million",main="logCPMs_filtered&normalized",col=c(rep("black",6),rep("cyan",6),rep("cyan4",6),rep("aquamarine",6),rep("aquamarine4",6),rep("coral",6),rep("coral4",6),rep("hotpink",6),rep("hotpink4",6)))
abline(h=median(v.norm$E),col="blue")




```

```{r}
#plotMDS

install.packages("scatterplot3d") # Install
library("scatterplot3d") # load


mds<-plotMDS(v.norm$E,ndim=3)
toplot<-data.frame(Dim1=mds$cmdscale.out[,1],Dim3=mds$cmdscale.out[,3],Dim2=mds$cmdscale.out[,2],Group = factor(x_norm$samples$age_sleep),Age=factor(x_norm$samples$age),sleep=factor(x_norm$samples$sleep),timepoint=factor(x_norm$samples$timepoint))

palette<-brewer.pal(12, "Paired")[c(2,4,6,8,10,12)]
colors <- palette[as.numeric(toplot$Group)]

shapes = c(19, 17) 
shapes <- shapes[as.numeric(toplot$Age)]

old.par <- par(no.readonly=T) #save default par settings
par(old.par) #restore par settings after modification

png(filename = "MDS3d.png",height =4,width =5,res = 300,units = "in",type="cairo")

scatterplot3d(toplot[,1:3], pch = shapes, cex.symbols=1.5,color=colors, grid=F,angle = -15, cex.lab=1.5,xlab = "MDS dim1",ylab = "MDS dim3",zlab = "MDS dim2",mar=c(3,2,1,3))

title("(b)", adj = 0, line = 0)
dev.off()

png(filename = "MDS2d.png",height =4,width =4.3,res = 300,units = "in",type="cairo")

ggscatter(toplot, x="Dim1",y="Dim2",color="Group",shape="Age",size=3.5, palette=brewer.pal(7, "Set1")[c(1:5,7)],ellipse=T, ellipse.type = "convex",ellipse.alpha = 0,xlab="MDS dim1", ylab="MDS dim2",title="2B") +
theme(legend.position="none",panel.background = element_rect(fill="white", linetype="solid", color="black"), plot.title = element_text(face="bold",hjust=-0.14))

dev.off()

#separate Y from O for the 2D MDS plot
png(filename = "MDS2d.Y.png",height =4,width =7,res = 300,units = "in",type="cairo")

ggscatter(toplot, x="Dim1",y="Dim2",color="Group",shape="Age",size=3.5,facet.by="Age", palette=palette,ellipse=T, ellipse.alpha = 0.2,xlab="MDS dim1", ylab="MDS dim2",title="(a)") +
theme(legend.position="right",panel.background = element_rect(fill="white", linetype="solid", color="black"), plot.title = element_text(face="bold",hjust=-0.1),strip.text=element_text(size = 14, face="bold"),axis.title=element_text(size = 12, face="bold"))

dev.off()

#optional legend for the 3D scatter plot
legend("right",bty = "n", legend = levels(toplot$Group),inset = 0,cex=2.5,
      col = palette, pch = c(rep(19,3),rep(17,3)))

#MDS plot without Z0 samples

toplot.rmZ0<-toplot%>%filter(timepoint!="Z0")%>%mutate(Group=factor(Group,levels=c("Y.SS","Y.SD","O.SS","O.SD")))

palette<-brewer.pal(4, "Set1")[c(1:4)]
colors.rmZo <- palette[as.numeric(toplot.rmZ0$Group)]

shapes = c(19, 17) 
shapes.rmZo <- shapes[as.numeric(toplot.rmZ0$Age)]

png(filename = "MDS3d.rmZ0.png",height =4,width =5,res = 300,units = "in",type="cairo")

scatterplot3d(toplot.rmZ0[,1:3], pch = shapes.rmZo, cex.symbols=1.5,color=colors.rmZo, grid=F,angle = -20, xlab=NULL,ylab=NULL,zlab=NULL,mar=c(3,2,1,3))

legend("right",bty = "n", legend = levels(toplot.rmZ0$Group),inset = 0.05,
      col = palette, pch = c(rep(19,2),rep(17,2)))

dev.off()


png(filename = "MDS2d.rmzo.png",height =4,width =6.5,res = 300,units = "in",type="cairo")

ggscatter(toplot.rmZ0, x="Dim1",y="Dim2",color="Group",shape="Age",size=3.5,facet.by="Age", palette=brewer.pal(4, "Set1")[c(1:4)],ellipse=T, ellipse.alpha = 0.2,xlab="MDS dim1", ylab="MDS dim2",title="2D MDS plot") +
theme(legend.position="none",panel.background = element_rect(fill="white", linetype="solid", color="black"), plot.title = element_text(face="bold",hjust=0.5),strip.text = element_text(face="bold", size=12),
strip.background = element_rect(fill="white", colour="black"),axis.title.x = element_text(face="bold",size=10),axis.title.y = element_text(face="bold",size=10))

dev.off()

```

```{r}
#contrasts for DEG
#two main tests: 1. age effect at Z0,3,6,9,12 during undisturbed sleep; 2. SD vs SS for Y and O separately; 3. differences between sleep and ZT0 (averaged across four time points) for determination of sleep genes


contrasts.main<-makeContrasts(
  OvY.SS3h=O.SS.3hr-Y.SS.3hr,
  OvY.SS6h=O.SS.6hr-Y.SS.6hr,
  OvY.SS9h=O.SS.9hr-Y.SS.9hr,
  OvY.SS12h=O.SS.12hr-Y.SS.12hr,
  OvY.SD3h=O.SD.3hr-Y.SD.3hr,
  OvY.SD6h=O.SD.6hr-Y.SD.6hr,
  OvY.SD9h=O.SD.9hr-Y.SD.9hr,
  OvY.SD12h=O.SD.12hr-Y.SD.12hr,
  OvY.Z0=O.Z0.Z0-Y.Z0.Z0,
  SDvsSS_3hr.Y = Y.SD.3hr-Y.SS.3hr, 
   SDvsSS_6hr.Y = Y.SD.6hr-Y.SS.6hr,
   SDvsSS_9hr.Y = Y.SD.9hr-Y.SS.9hr,
   SDvsSS_12hr.Y = Y.SD.12hr-Y.SS.12hr,
  SDvsSS_3hr.O = O.SD.3hr-O.SS.3hr, 
   SDvsSS_6hr.O = O.SD.6hr-O.SS.6hr,
   SDvsSS_9hr.O = O.SD.9hr-O.SS.9hr,
   SDvsSS_12hr.O = O.SD.12hr-O.SS.12hr,
  aveYSS.vZ0=(Y.SS.3hr+Y.SS.6hr+Y.SS.9hr+Y.SS.12hr)/4-Y.Z0.Z0,
  aveOSS.vZ0=(O.SS.3hr+O.SS.6hr+O.SS.9hr+O.SS.12hr)/4-O.Z0.Z0,
  SSvZ0.Y3h=Y.SS.3hr-Y.Z0.Z0,
  SSvZ0.Y6h=Y.SS.6hr-Y.Z0.Z0,
  SSvZ0.Y9h=Y.SS.9hr-Y.Z0.Z0,
  SSvZ0.Y12h=Y.SS.12hr-Y.Z0.Z0,
  SSvZ0.O3h=O.SS.3hr-O.Z0.Z0,
  SSvZ0.O6h=O.SS.6hr-O.Z0.Z0,
  SSvZ0.O9h=O.SS.9hr-O.Z0.Z0,
  SSvZ0.O12h=O.SS.12hr-O.Z0.Z0,
  SDvZ0.Y3h=Y.SD.3hr-Y.Z0.Z0,
  SDvZ0.Y6h=Y.SD.6hr-Y.Z0.Z0,
  SDvZ0.Y9h=Y.SD.9hr-Y.Z0.Z0,
  SDvZ0.Y12h=Y.SD.12hr-Y.Z0.Z0,
  SDvZ0.O3h=O.SD.3hr-O.Z0.Z0,
  SDvZ0.O6h=O.SD.6hr-O.Z0.Z0,
  SDvZ0.O9h=O.SD.9hr-O.Z0.Z0,
  SDvZ0.O12h=O.SD.12hr-O.Z0.Z0,
  levels=design)

efit.main<- eBayes(contrasts.fit(vfit, contrasts=contrasts.main)) 

#DEGs with F-test over four timepoints FDR<1%

DEG.SDvSS.Y<-topTable(efit.main,coef=10:13,n=Inf,p.value=0.01) #5158 DEG with FDR<1% for SDvSS.young animals
DEG.SDvSS.O<-topTable(efit.main,coef=14:17,n=Inf,p.value=0.01) #3540 DEG with FDR<1% for SDvSS.old animals
DEG.OvY.SS<-topTable(efit.main,coef=1:4,n=Inf,p.value=0.01) #654 DEG with FDR<1% for Old vs. Young during sleep (SS)
DEG.OvY.SD<-topTable(efit.main,coef=5:8,n=Inf,p.value=0.01) #951 DEG with FDR<1% for Old vs. Young during sleep deprivation (SD)
DEG.OvY.ZT0<-topTable(efit.main,coef=9,n=Inf,p.value=0.01) #255 DEG with FDR<1% for Old vs. Young at ZT0 (7am)

#define sleep genes as SD.down genes with aveExp.SS (across 4 timepoints) vs. ZT0 > 0.1 (log2FC) for young and old separately; define wake genes as SD.up genes with avelog2FC (across 4 timepoints) >0.2

DEG.SDvSS.Y.full<-topTable(efit.main,coef=10:13,n=Inf) 
DEG.SDvSS.O.full<-topTable(efit.main,coef=14:17,n=Inf) 
DEG.OvY.SS.full<-topTable(efit.main,coef=1:4,n=Inf) 
DEG.OvY.SD.full<-topTable(efit.main,coef=5:8,n=Inf) 
DEG.OvY.ZT0.full<-topTable(efit.main,coef=9,n=Inf) 



aveFCvsZT0<-cbind(as.data.frame(efit.main$genes),as.data.frame(efit.main$coefficients))%>%
  dplyr::select(id,starts_with("ave"))

test<-DEG.SDvSS.O.full%>%
  dplyr::select(id,P.Value,adj.P.Val,starts_with("SDvsSS"))%>%
  left_join(DEG.SDvSS.Y.full,.,by="id",suffix=c(".Y",".O"))%>%
  dplyr::select(id,ENTREZID,description,starts_with("gene"),starts_with("P.Value"),starts_with("adj."),starts_with("SDvsSS"))%>%
  dplyr::rename(FDR.Y=adj.P.Val.Y,FDR.O=adj.P.Val.O)%>%
  mutate(aveFC.SDvSS.Y=rowMeans(.[,11:14]),aveFC.SDvSS.O=rowMeans(.[,15:18]))%>%
  left_join(.,aveFCvsZT0,by="id")%>%
  rename(aveFC.SSvZ0.Y=aveYSS.vZ0,aveFC.SSvZ0.O=aveOSS.vZ0)%>%
  mutate(DEG.class=ifelse(FDR.Y<0.01&FDR.O<0.01,"common.DEG",ifelse(FDR.Y<0.01,"Yspec.DEG",ifelse(FDR.O<0.01, "Ospec.DEG","NS"))), 
         wake.0=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0&FDR.O<0.01&aveFC.SDvSS.O>0,"common.wake",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0,"Yspec.wake",ifelse(FDR.O<0.01&aveFC.SDvSS.O>0,"Ospec.wake","na"))),
         sleep.0=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<0&FDR.O<0.01&aveFC.SDvSS.O<0,"common.sleep",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<0,"Yspec.sleep",ifelse(FDR.O<0.01&aveFC.SDvSS.O<0,"Ospec.sleep","na"))),
         wake.0.2=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0.2&FDR.O<0.01&aveFC.SDvSS.O>0.2,"common.wake",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0.2,"Yspec.wake",ifelse(FDR.O<0.01&aveFC.SDvSS.O>0.2,"Ospec.wake","na"))),
         sleep.0.2=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<(-0.2)&FDR.O<0.01&aveFC.SDvSS.O<(-0.2),"common.sleep",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<(-0.2),"Yspec.sleep",ifelse(FDR.O<0.01&aveFC.SDvSS.O<(-0.2),"Ospec.sleep","na"))))%>%
  mutate(sleep.0.Z0.Y=ifelse(sleep.0.2%in%c("common.sleep","Yspec.sleep")&aveFC.SSvZ0.Y>0,"SSvZ0>0","na"),    sleep.0.Z0.O=ifelse(sleep.0.2%in%c("common.sleep","Ospec.sleep")&aveFC.SSvZ0.O>0,"SSvZ0>0","na"), 
         sleep.0.1.Z0.Y=ifelse(sleep.0.2%in%c("common.sleep","Yspec.sleep")&aveFC.SSvZ0.Y>0.1,"SSvZ0>0.1","na"),
         sleep.0.1.Z0.O=ifelse(sleep.0.2%in%c("common.sleep","Ospec.sleep")&aveFC.SSvZ0.O>0.1,"SSvZ0>0.1","na"))%>%
  mutate(sleep.class=ifelse(sleep.0.1.Z0.Y!="na"&sleep.0.1.Z0.O!="na","common.sleep", ifelse(sleep.0.1.Z0.Y!="na","Yspec.sleep",ifelse(sleep.0.1.Z0.O!="na","Ospec.sleep","na"))), wake.class=wake.0.2)

write.table(test,"test.txt",sep="\t",row.names = F)
  

DEG.combine.all<-DEG.SDvSS.O.full%>%
  dplyr::select(id,P.Value,adj.P.Val,starts_with("SDvsSS"))%>%
  left_join(DEG.SDvSS.Y.full,.,by="id",suffix=c(".Y",".O"))%>%
  dplyr::select(id,starts_with("P.Value"),starts_with("adj."),starts_with("SDvsSS"))%>%
  left_join(.,DEG.OvY.SS.full,by="id")%>%
  dplyr::select(id,starts_with("P.Value"),starts_with("adj."),starts_with("SDvsSS"),starts_with("OvY"))%>%
  left_join(.,DEG.OvY.SD.full,by="id",suffix=c(".SS",".SD"))%>%
  dplyr::select(id,starts_with("P.Value"),starts_with("adj."),starts_with("SDvsSS"),starts_with("OvY"))%>%
  left_join(.,DEG.OvY.ZT0.full,by="id")%>%
  dplyr::select(id,ENTREZID,description,starts_with("gene"),starts_with("P.Value"),starts_with("adj."),starts_with("SDvsSS"),starts_with("OvY"),logFC)%>%
    dplyr::rename(FDR.Y=adj.P.Val.Y,FDR.O=adj.P.Val.O,FDR.SS.OvY=adj.P.Val.SS,FDR.SD.OvY=adj.P.Val.SD,FDR.Z0.OvY=adj.P.Val,P.Value.Z0=P.Value,OvY.Z0=logFC)%>%
  mutate(aveFC.SDvSS.Y=rowMeans(.[,17:20]),aveFC.SDvSS.O=rowMeans(.[,21:24]), aveFC.OvY.SS=rowMeans(.[,25:28]), aveFC.OvY.SD=rowMeans(.[,29:32]))%>%
  mutate(SD.up=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0&FDR.O<0.01&aveFC.SDvSS.O>0,"common",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0,"Yspec",ifelse(FDR.O<0.01&aveFC.SDvSS.O>0,"Ospec","na"))), SD.down=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<0&FDR.O<0.01&aveFC.SDvSS.O<0,"common",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<0,"Yspec",ifelse(FDR.O<0.01&aveFC.SDvSS.O<0,"Ospec","na"))),
         wake=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0.2&FDR.O<0.01&aveFC.SDvSS.O>0.2,"common",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0.2,"Yspec",ifelse(FDR.O<0.01&aveFC.SDvSS.O>0.2,"Ospec","na"))),
         sleep=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<(-0.2)&FDR.O<0.01&aveFC.SDvSS.O<(-0.2),"common",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<(-0.2),"Yspec",ifelse(FDR.O<0.01&aveFC.SDvSS.O<(-0.2),"Ospec","na"))))%>%
  mutate(dir=ifelse(wake!="na","wake",ifelse(sleep!="na","sleep","na")), class=ifelse(wake=="common"|sleep=="common","common",ifelse(wake=="Yspec"|sleep=="Yspec","Yspec",ifelse(wake=="Ospec"|sleep=="Ospec","Ospec","na"))))%>%
  left_join(.,aveFCvsZT0,by="id")

write.table(DEG.combine.all,"DEG.combine.all.txt",sep="\t",row.names = F)

test<-DEG.combine.all%>%
  mutate(FC0.5.Y=ifelse(FDR.Y<0.01&abs(aveFC.SDvSS.Y)>=0.5,">0.5",ifelse(FDR.Y<0.01&abs(aveFC.SDvSS.Y)<0.5,"<0.5","ns")), FC0.5.O=ifelse(FDR.O<0.01&abs(aveFC.SDvSS.O)>=0.5,">0.5",ifelse(FDR.O<0.01&abs(aveFC.SDvSS.O)<0.5,"<0.5","ns")))%>%
  group_by(FC0.5.Y, FC0.5.O)%>%
  summarise(n())


library(VennDiagram)

#DEG.Y vs. DEG.O

venn.diagram(list(DEG.combine.all[DEG.combine.all$FDR.Y<0.01,]$id,DEG.combine.all[DEG.combine.all$FDR.O<0.01,]$id), "SDvSS.FDR1.tiff", fill=brewer.pal(4,"Set1")[3:4], cex = 3, main="(c) DEGs between SD and SS", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=3, cat.cex=3,category.names=c("Young","Old"),cat.pos=c(-50,55))


#volcano plot

vol.Y<-DEG.combine.all%>%
  mutate(class.y=ifelse(FDR.Y>0.01,"ns",ifelse(abs(aveFC.SDvSS.Y)>0.2,"high","low")),class.o=ifelse(FDR.O>0.01,"ns",ifelse(abs(aveFC.SDvSS.O)>0.2,"high","low")))%>%
  mutate_at(vars(starts_with("FDR")),funs(-log10(.)))%>%
ggscatter(.,x="aveFC.SDvSS.Y",y="FDR.Y",size=0.8,color="class.y", palette=c("green3","gray50","gray80"),xlab=expression('Average log'[2]*'fold change'), ylab=expression('-log'[10]*'FDR'),title="DEGs in young", xlim=c(-3.5,3.5))+
geom_hline(yintercept=2, linetype="longdash", color = "black",size=0.5)+
  geom_vline(xintercept=c(-0.2,0.2), linetype="longdash", color = c("red","blue"),size=1)+
  scale_x_continuous(breaks=c(-3,-2,-1,-0.2,0.2,1,2,3))+scale_y_continuous(breaks=c(0,2,10,20,30))+
  theme(legend.position = "none",axis.title=element_text(size=14,face="bold"), axis.text.x=element_text(angle = 60,hjust = 1),plot.title=element_text(face="bold",size=16,hjust=0.5))+
  annotate("segment", x = -0.2, xend = -1.2, y = 30, yend = 30, colour = "red", size=1, alpha=0.6, arrow=arrow(length = unit(0.5, "cm")))+
  annotate("segment", x = 0.2, xend = 1.2, y = 30, yend = 30, colour = "blue", size=1, alpha=0.6, arrow=arrow(length = unit(0.5, "cm")))+
  annotate("text", x = c(-2.5,2.5), y = c(30,30), label = c("1858", "1944") , color=c("red","blue"), size=7, fontface="plain")+
  annotate("text", x = c(-2.5,2.5), y = c(25,25), label = c("sleep", "wake") , color=c("red","blue"), size=8, fontface="bold")


vol.O<-DEG.combine.all%>%
  mutate(class.y=ifelse(FDR.Y>0.01,"ns",ifelse(abs(aveFC.SDvSS.Y)>0.2,"high","low")),class.o=ifelse(FDR.O>0.01,"ns",ifelse(abs(aveFC.SDvSS.O)>0.2,"high","low")))%>%
  mutate_at(vars(starts_with("FDR")),funs(-log10(.)))%>%
ggscatter(.,x="aveFC.SDvSS.O",y="FDR.O",size=0.8,color="class.o", palette=c("purple3","gray50","gray80"),xlab=expression('Average log'[2]*'fold change'), ylab=expression('-log'[10]*'FDR'),title="DEGs in old",xlim=c(-3.5,3.5))+
geom_hline(yintercept=2, linetype="longdash", color = "black",size=0.5)+
  geom_vline(xintercept=c(-0.2,0.2), linetype="longdash", color = c("red","blue"),size=1)+
  scale_x_continuous(breaks=c(-3,-2,-1,-0.2,0.2,1,2,3))+scale_y_continuous(breaks=c(0,2,10,20,30))+
  theme(legend.position = "none",axis.title=element_text(size=14,face="bold"), axis.text.x=element_text(angle = 60,hjust = 1),plot.title=element_text(face="bold",size=16,hjust=0.5))+
  annotate("segment", x = -0.2, xend = -1.2, y = 30, yend = 30, colour = "red", size=1, alpha=0.6, arrow=arrow(length = unit(0.5, "cm")))+
  annotate("segment", x = 0.2, xend = 1.2, y = 30, yend = 30, colour = "blue", size=1, alpha=0.6, arrow=arrow(length = unit(0.5, "cm")))+
  annotate("text", x = c(-2.5,2.5), y = c(30,30), label = c("1254", "1278") , color=c("red","blue"), size=7, fontface="plain")+
  annotate("text", x = c(-2.5,2.5), y = c(25,25), label = c("sleep", "wake") , color=c("red","blue"), size=8, fontface="bold")


png(filename ="volcano.YO.png",height = 4,width =8,res = 300,units = "in",type="cairo")
ggarrange(vol.Y,vol.O,ncol=2,labels = c("(d)",""))
 dev.off()
  
  
#sleep genes and wake genes
venn.diagram(list(DEG.combine.all[DEG.combine.all$sleep%in%c("common","Yspec"),]$id,DEG.combine.all[DEG.combine.all$sleep%in%c("common","Ospec"),]$id), "sleep.FC0.2.tiff", fill=brewer.pal(9,"Reds")[c(4,8)], cex = 3, main="Sleep Genes", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=5, cat.cex=3,category.names=c("Young","Old"),cat.pos=c(-50,55))

venn.diagram(list(DEG.combine.all[DEG.combine.all$wake%in%c("common","Yspec"),]$id,DEG.combine.all[DEG.combine.all$wake%in%c("common","Ospec"),]$id), "wake.FC0.2.tiff", fill=brewer.pal(9,"Blues")[c(4,8)], cex = 3, main="Wake Genes", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=5, cat.cex=3,category.names=c("Young","Old"),cat.pos=c(-50,55))


#other venn not used in the paper
venn.diagram(list(test[test$FDR.Y<0.01&test$aveFC.SDvSS.Y<0,]$id,test[test$FDR.O<0.01&test$aveFC.SDvSS.O<0,]$id), "sleep.FC0.tiff", fill=brewer.pal(9,"Reds")[c(4,8)], cex = 3.5, main="Sleep.aveFC<0", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=5, cat.cex=3,category.names=c("Young","Old"),cat.pos=c(-50,55))


venn.diagram(list(test[test$FDR.Y<0.01&test$aveFC.SDvSS.Y>0,]$id,test[test$FDR.O<0.01&test$aveFC.SDvSS.O>0,]$id), "wake.FC0.tiff", fill=brewer.pal(9,"Blues")[c(4,8)], cex = 3.5, main="Wake.aveFC>0", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=5, cat.cex=3,category.names=c("Young","Old"),cat.pos=c(-50,55))



venn.diagram(list(test[test$sleep.0.Z0.Y=="SSvZ0>0",]$id,test[test$sleep.0.Z0.O=="SSvZ0>0",]$id), "sleep.SSvZ0.0.tiff", fill=brewer.pal(9,"Reds")[c(4,8)], cex = 3.5, main="Sleep.SSvZ0>0", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=5, cat.cex=3,category.names=c("Young","Old"),cat.pos=c(-50,55))


venn.diagram(list(test[test$sleep.0.1.Z0.Y=="SSvZ0>0.1",]$id,test[test$sleep.0.1.Z0.O=="SSvZ0>0.1",]$id), "sleep.SSvZ0.0.1.tiff", fill=brewer.pal(9,"Reds")[c(4,8)], cex = 3.5, main="Sleep.SSvZ0>0.1", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=5, cat.cex=3,category.names=c("Young","Old"),cat.pos=c(-50,55))


#comparing the P.Value between Y and O for the common, Y-specific and O-specific genes

DEG.combine.all%>%
  dplyr::select(id,dir,class,P.Value.Y,P.Value.O)%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  filter(dir!="na")%>%
  gather(age,p.val,P.Value.Y:P.Value.O)%>%
  mutate(age=ifelse(grepl("Y",age),"Y","O"))%>%
mutate(age=factor(age,levels=c("Y","O"),labels=c("Young","Old")),dir=factor(dir,levels=c("sleep","wake"),labels=c("Sleep Genes","Wake Genes")),class=factor(class,levels=c("common","Yspec","Ospec"),labels=c("Common","Y-specific","O-specific")))%>%
  ggviolin(.,x="age",y="p.val",fill = "age",add="boxplot",add.params = list(fill = "white"), facet.by=c("dir","class"),xlab="",ylab="-log10 (P.Value)")+
  stat_compare_means(label="p.format",paired = TRUE,label.x=1.3)+
theme(legend.position="none",strip.text = element_text(face="bold", size=12),axis.text.x = element_text(face="bold",size=11))

p1<-DEG.combine.all%>%
  dplyr::select(id,dir,class,P.Value.Y,P.Value.O)%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  filter(dir=="sleep")%>%
  gather(age,p.val,P.Value.Y:P.Value.O)%>%
  mutate(age=ifelse(grepl("Y",age),"Y","O"),ngene=class)%>%
mutate(age=factor(age,levels=c("Y","O"),labels=c("Young","Old")),dir=factor(dir,levels=c("sleep","wake"),labels=c("Sleep Genes","Wake Genes")),class=factor(class,levels=c("common","Yspec","Ospec"),labels=c("Common","Y-specific","O-specific")),ngene=factor(ngene,levels=c("common","Yspec","Ospec"),labels=c("n=982","n=876","n=272")))%>%
  ggviolin(.,x="age",y="p.val",fill = "class",palette = c("#00AFBB", "#E7B800", "#FC4E07"), add="boxplot",add.params = list(fill = "white"), xlab="",ylab=expression('-log'[10]*' P-values'), title="Sleep Genes")+
  facet_wrap(vars(class,ngene),strip.position = "bottom")+
  stat_compare_means(label="p.signif",paired = TRUE,label.x=1.4,label.y=40)+
theme(legend.position="none",strip.text = element_text(face="bold",size=12,color="white"), axis.text.x = element_text(face="bold",size=11),strip.background = element_rect(color="white",fill =brewer.pal(9,"YlOrBr")[8]), plot.title=element_text(face="bold",size=18,hjust=0.5), axis.title.y=element_text(face="bold",size=14))+
  scale_y_continuous(expand=c(0,1))

  
p2<-DEG.combine.all%>%
  dplyr::select(id,dir,class,P.Value.Y,P.Value.O)%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  filter(dir=="wake")%>%
  gather(age,p.val,P.Value.Y:P.Value.O)%>%
  mutate(age=ifelse(grepl("Y",age),"Y","O"),ngene=class)%>%
mutate(age=factor(age,levels=c("Y","O"),labels=c("Young","Old")),dir=factor(dir,levels=c("sleep","wake"),labels=c("Sleep Genes","Wake Genes")),class=factor(class,levels=c("common","Yspec","Ospec"),labels=c("Common","Y-specific","O-specific")),ngene=factor(ngene,levels=c("common","Yspec","Ospec"),labels=c("n=1075","n=869","n=203")))%>%
  ggviolin(.,x="age",y="p.val",fill = "class",palette = c("#00AFBB", "#E7B800", "#FC4E07"), add="boxplot",add.params = list(fill = "white"), xlab="",ylab=expression('-log'[10]*' P-values'), title="Wake Genes")+
  facet_wrap(vars(class,ngene),strip.position = "bottom")+
  stat_compare_means(label="p.signif",paired = TRUE,label.x=1.4,label.y=40)+
theme(legend.position="none",strip.text = element_text(face="bold",size=12,color="white"), axis.text.x = element_text(face="bold",size=11),strip.background = element_rect(color="white",fill = brewer.pal(9,"PuBu")[8]), plot.title=element_text(face="bold",size=18,hjust=0.5), axis.title.y=element_text(face="bold",size=14))+
  scale_y_continuous(expand=c(0,1))

png(filename ="Pval.compare.png",height = 5,width =10,res = 300,units = "in",type="cairo")

ggarrange(p1,p2,ncol=2,labels = c("(a)",""))

  dev.off()
  

 
bar1<-DEG.combine.all%>%
  dplyr::select(id,dir,class,P.Value.Y,P.Value.O)%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  filter(dir!="na")%>%
  mutate(deltaP=P.Value.O-P.Value.Y, class=factor(class,levels=c("common","Yspec","Ospec"),labels=c("Common","Y-specific","O-specific")))%>%
  mutate(deltaP.dir=ifelse(deltaP>0,"O-up","O-down"))%>%
group_by(dir,class,deltaP.dir)%>%
  summarise(n=n())%>%
  mutate(freq = n / sum(n)*100)%>%
filter(deltaP.dir=="O-down"&dir=="sleep")%>%
  ggbarplot(x="class",y="freq",fill="class",palette = c("#00AFBB", "#E7B800", "#FC4E07"),label=T,lab.pos="out",lab.size=4,lab.nb.digits=1,ylim=c(0,140), xlab="",ylab="%decrease in old")+theme(axis.title.y=element_text(size=8,face="bold"),axis.text.x=element_text(angle=25,hjust=1,size=7,face="bold"),axis.text.y=element_text(size=7),legend.position="none",plot.background = element_rect(fill = "transparent", color = NA))+
  scale_y_continuous(breaks=c(0,25,50,75,100))


bar2<-DEG.combine.all%>%
  dplyr::select(id,dir,class,P.Value.Y,P.Value.O)%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  filter(dir!="na")%>%
  mutate(deltaP=P.Value.O-P.Value.Y, class=factor(class,levels=c("common","Yspec","Ospec"),labels=c("Common","Y-specific","O-specific")))%>%
  mutate(deltaP.dir=ifelse(deltaP>0,"O-up","O-down"))%>%
group_by(dir,class,deltaP.dir)%>%
  summarise(n=n())%>%
  mutate(freq = n / sum(n)*100)%>%
filter(deltaP.dir=="O-down"&dir=="wake")%>%
  ggbarplot(x="class",y="freq",fill="class",palette = c("#00AFBB", "#E7B800", "#FC4E07"),label=T,lab.pos="out",lab.size=4,lab.nb.digits=1,ylim=c(0,140), xlab="",ylab="%decrease in old")+theme(axis.title.y=element_text(size=8,face="bold"),axis.text.x=element_text(angle=25,hjust=1,size=7,face="bold"),axis.text.y=element_text(size=7),legend.position="none",plot.background = element_rect(fill = "transparent", color = NA))+
  scale_y_continuous(breaks=c(0,25,50,75,100))


bar1_grob <- ggplotGrob(bar1)
bar2_grob <- ggplotGrob(bar2)

  
p1<-DEG.combine.all%>%
  dplyr::select(id,dir,class,P.Value.Y,P.Value.O)%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  filter(dir=="sleep")%>%
  mutate(deltaP=P.Value.O-P.Value.Y, class=factor(class,levels=c("common","Yspec","Ospec"),labels=c("Common","Y-specific","O-specific")))%>%
ggscatter(.,x="P.Value.Y",y="deltaP",size=1,color="class",palette = c("#00AFBB", "#E7B800", "#FC4E07"), ylim=c(-12,12),xlab=expression('-log'[10]*'P-values of young mice'),ylab=expression('Difference in -log'[10]*'P-values between old and young'),title="Sleep Genes")+
    geom_hline(yintercept = 0,linetype="dashed",color="black")+
    theme(plot.title=element_text(size=16,face="bold",hjust=0.5),legend.position="none",axis.title=element_text(size=14,face="bold"))+
  annotation_custom(grob = bar1_grob, xmin = 18, xmax = 40, 
                       ymin = 2, ymax = 12)

p2<-DEG.combine.all%>%
  dplyr::select(id,dir,class,P.Value.Y,P.Value.O)%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  filter(dir=="wake")%>%
  mutate(deltaP=P.Value.O-P.Value.Y, class=factor(class,levels=c("common","Yspec","Ospec"),labels=c("Common","Y-specific","O-specific")))%>%
ggscatter(.,x="P.Value.Y",y="deltaP",size=1,color="class",palette = c("#00AFBB", "#E7B800", "#FC4E07"), ylim=c(-12,12),xlab=expression('-log'[10]*'P-values of young mice'),ylab=expression('Difference in -log'[10]*'P-values between old and young'),title="Wake Genes")+
    geom_hline(yintercept = 0,linetype="dashed",color="black")+
    theme(plot.title=element_text(size=16,face="bold",hjust=0.5),legend.position="none",axis.title=element_text(size=14,face="bold"))+
  annotation_custom(grob = bar2_grob, xmin = 18, xmax = 40, 
                       ymin = 2, ymax = 12)

png(filename ="deltaPval.inset.png",height = 5,width =8,res = 300,units = "in",type="cairo")
ggarrange(p1,p2,ncol=2)

dev.off()
  
  
#calculate CV among biological replicate in each condition
CV.DEGs<-cbind(as.data.frame(v.norm$genes),as.data.frame(v.norm$E))%>%
        filter(.$id%in%DEG.combine.all[DEG.combine.all$class!="na",]$id)%>%
        dplyr::select(-starts_with("gene"),-ENTREZID,-description,-transcript_length)%>%
        gather(key=sample,value=logCPM,-id)%>%
        mutate(CPM=2^logCPM)%>%
  mutate(age=ifelse(grepl("Young",sample),"Y","O"),sleep=ifelse(grepl("SS",sample),"SS","SD"),time=ifelse(grepl("Z0",sample),"Z0",ifelse(grepl("3hr",sample),"3",ifelse(grepl("6hr",sample),"6",ifelse(grepl("9hr",sample),"9","12")))))%>%
  filter(time!="Z0")%>%
  unite(con,age:time,sep=".")%>%  
  unite(con.id,c("id","con"),sep="_")%>%
  group_by(con.id)%>%
  summarise(meanCPM=mean(CPM),sdCPM=sd(CPM))%>%
  mutate(CV=sdCPM/meanCPM)%>%
  mutate(logCV=log(CV))

png(filename ="CV.corr.png",height = 8,width =4.5,res = 300,units = "in",type="cairo")
CV.DEGs%>%
  separate(con.id,c("id","con"),sep="_")%>%
    dplyr::select(id,con,logCV)%>%
    separate(con,c("age","sleep","time"))%>%
    spread(age,logCV)%>%
    mutate(sleep=factor(sleep,levels=c("SS","SD")),time=factor(time,levels=c("3","6","9","12"),labels=c("3h","6h","9h","12h")))%>%
ggscatter(.,x="Y",y="O",size=0.5,facet.by=c("time","sleep"),ylim=c(-5,1),ellipse = F,ellipse.alpha=0.8,xlab="log transformed CV of the young", ylab="log transformed CV of the old",title="Correlation plots of CV between young and old mice")+geom_abline(slope=1,linetype="dashed",size=1)+stat_cor(method = "pearson", label.x = -4.5, label.y = 0.5)+theme(plot.title=element_text(face="bold",size=11,hjust=0.5),axis.title = element_text(face="bold",size=14))
  
 dev.off()


 png(filename ="deltaCV.png",height = 6,width =7,res = 300,units = "in",type="cairo")

 CV.DEGs%>%
  separate(con.id,c("id","con"),sep="_")%>%
    dplyr::select(id,con,logCV)%>%
    separate(con,c("age","sleep","time"))%>%
    spread(age,logCV)%>%
    mutate(deltaCV=O-Y,sleep=factor(sleep,levels=c("SS","SD")),time=factor(time,levels=c("3","6","9","12"),labels=c("3h","6h","9h","12h")))%>%
   mutate(deltaCV.dir=ifelse(deltaCV>0,"O-up","O-down"))%>%
   group_by(sleep,time,deltaCV.dir)%>%
   summarise(n=n())%>%
  mutate(freq = n / sum(n)*100)%>%
   mutate(deltaCV.dir=factor(deltaCV.dir,levels=c("O-down","O-up"),labels=c("Young","Old")))%>%
ggbarplot(.,x="time",y="freq",fill="deltaCV.dir", facet.by="sleep",ylim=c(0,100),size=2,label=T,lab.pos="in",lab.size=6,lab.nb.digits=0,xlab="Hours of spontaneous sleep or sleep deprivation", ylab="Percentage of genes with increased CV",title="Shift in CV between young and old mice")+geom_hline(yintercept = 50,linetype="longdash",size=1,color="black")+theme(plot.title=element_text(face="bold",size=14,hjust=0.5),axis.title.x = element_text(face="bold",size=12),axis.title.y = element_text(face="bold",size=14),strip.text = element_text(face="bold", size=14), legend.title = element_blank(),legend.position = "bottom")
   dev.off()
  
  
bar3<-DEG.combine.all%>%
  dplyr::select(id,dir,class,starts_with("aveFC.SDvSS"))%>%
  filter(dir!="na")%>%
  mutate(deltaFC=abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y), class=factor(class,levels=c("common","Yspec","Ospec"),labels=c("Common","Y-specific","O-specific")))%>%
  mutate(deltaFC.dir=ifelse(deltaFC>0,"O-up","O-down"))%>%
group_by(dir,class,deltaFC.dir)%>%
  summarise(n=n())%>%
  mutate(freq = n / sum(n)*100)%>%
filter(deltaFC.dir=="O-down"&dir=="sleep")%>%
  ggbarplot(x="class",y="freq",fill="class",palette = c("#00AFBB", "#E7B800", "#FC4E07"),label=T,lab.pos="out",lab.size=4,lab.nb.digits=1,ylim=c(0,140), xlab="",ylab="%decrease in old")+theme(axis.title.y=element_text(size=8,face="bold"),axis.text.x=element_text(angle=25,hjust=1,size=7,face="bold"),axis.text.y=element_text(size=7),legend.position="none",plot.background = element_rect(fill = "transparent", color = NA))+
  scale_y_continuous(breaks=c(0,25,50,75,100))


bar4<-DEG.combine.all%>%
  dplyr::select(id,dir,class,starts_with("aveFC.SDvSS"))%>%
  filter(dir!="na")%>%
  mutate(deltaFC=abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y), class=factor(class,levels=c("common","Yspec","Ospec"),labels=c("Common","Y-specific","O-specific")))%>%
  mutate(deltaFC.dir=ifelse(deltaFC>0,"O-up","O-down"))%>%
group_by(dir,class,deltaFC.dir)%>%
  summarise(n=n())%>%
  mutate(freq = n / sum(n)*100)%>%
filter(deltaFC.dir=="O-down"&dir=="wake")%>%
  ggbarplot(x="class",y="freq",fill="class",palette = c("#00AFBB", "#E7B800", "#FC4E07"),label=T,lab.pos="out",lab.size=4,lab.nb.digits=1,ylim=c(0,140), xlab="",ylab="%decrease in old")+theme(axis.title.y=element_text(size=8,face="bold"),axis.text.x=element_text(angle=25,hjust=1,size=7,face="bold"),axis.text.y=element_text(size=7),legend.position="none",plot.background = element_rect(fill = "transparent", color = NA))+
  scale_y_continuous(breaks=c(0,25,50,75,100))


bar3_grob <- ggplotGrob(bar3)
bar4_grob <- ggplotGrob(bar4)

  
p3<-DEG.combine.all%>%
dplyr::select(id,dir,class,starts_with("aveFC.SDvSS"))%>%
  filter(dir=="sleep")%>%
  mutate(deltaFC=abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y), class=factor(class,levels=c("common","Yspec","Ospec"),labels=c("Common","Y-specific","O-specific")))%>%
ggscatter(.,x="aveFC.SDvSS.Y",y="deltaFC",size=1,color="class",palette = c("#00AFBB", "#E7B800", "#FC4E07"), ylim=c(-1.2,1.5),xlab=expression('Average log'[2]*'FC (SD/SS) of young mice'),ylab=expression('Difference in absolute average log'[2]*'FC (old-young)'),title="Sleep Genes")+
    geom_hline(yintercept = 0,linetype="dashed",color="black")+
    theme(plot.title=element_text(size=16,face="bold",hjust=0.5),legend.position="none",axis.title=element_text(size=13,face="bold"))+
  annotation_custom(grob = bar3_grob, xmin = -1.9, xmax = -0.7, 
                       ymin = 0.3, ymax = 1.5)

p4<-DEG.combine.all%>%
dplyr::select(id,dir,class,starts_with("aveFC.SDvSS"))%>%
  filter(dir=="wake")%>%
  mutate(deltaFC=abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y), class=factor(class,levels=c("common","Yspec","Ospec"),labels=c("Common","Y-specific","O-specific")))%>%
ggscatter(.,x="aveFC.SDvSS.Y",y="deltaFC",size=1,color="class",palette = c("#00AFBB", "#E7B800", "#FC4E07"),ylim=c(-1.2,1.5), xlab=expression('Average log'[2]*'FC (SD/SS) of young mice'),ylab=expression('Difference in absolute average log'[2]*'FC (old-young)'),title="Wake Genes")+
    geom_hline(yintercept = 0,linetype="dashed",color="black")+
    theme(plot.title=element_text(size=16,face="bold",hjust=0.5),legend.position="none",axis.title=element_text(size=13,face="bold"))+
  annotation_custom(grob = bar4_grob, xmin = 1.2, xmax = 3, 
                       ymin = 0.3, ymax = 1.5)

png(filename ="deltaFC.inset.png",height = 5,width =8,res = 300,units = "in",type="cairo")
ggarrange(p3,p4,ncol=2,labels = c("(b)",""))
dev.off()


#interaction with sleep and wake genes combined from y and o
combine.slwake.id<-x_norm$genes$id%in%DEG.combine.all[DEG.combine.all$class!="na",]$id

v.int.slwake<-v.norm[combine.slwake.id,]

contr.int<-makeContrasts(
  SDvsSS_3hr.YvO=(Y.SD.3hr-Y.SS.3hr)-(O.SD.3hr-O.SS.3hr),
  SDvsSS_6hr.YvO=(Y.SD.6hr-Y.SS.6hr)-(O.SD.6hr-O.SS.6hr),
  SDvsSS_9hr.YvO=(Y.SD.9hr-Y.SS.9hr)-(O.SD.9hr-O.SS.9hr),
  SDvsSS_12hr.YvO=(Y.SD.12hr-Y.SS.12hr)-(O.SD.12hr-O.SS.12hr),
  ave.int=((Y.SD.3hr-Y.SS.3hr)+(Y.SD.6hr-Y.SS.6hr)+(Y.SD.9hr-Y.SS.9hr)+(Y.SD.12hr-Y.SS.12hr))-((O.SD.3hr-O.SS.3hr)+(O.SD.6hr-O.SS.6hr)+(O.SD.9hr-O.SS.9hr)+(O.SD.12hr-O.SS.12hr)),
  levels=design)

vfit.int.slwake<- eBayes(contrasts.fit(lmFit(v.int.slwake, design),contrasts=contr.int))

DEG.combine.all.int<-topTable(vfit.int.slwake,coef=c(1:4),n=Inf)%>%
  dplyr::select(id,adj.P.Val)%>%
  rename(FDR.int=adj.P.Val)%>%
  left_join(DEG.combine.all,.,by="id")#21 FDR<0.05 from combined sleep and wake genes 

write.table(DEG.combine.all.int,"DEG.combine.all.int.txt",sep="\t",row.names = F)

t.test<-cbind(as.data.frame(efit.main$genes),as.data.frame(efit.main$p.value))%>%
  dplyr::select(id,starts_with("SDvsSS"))%>%
  dplyr::rename_at(.vars = vars(starts_with("SDvsSS")),
            .funs = funs(sub("SDvsSS","p.val", .)))


int.ave.DEG<-topTable(vfit.int.slwake,coef=1:4,n=Inf,p.value=0.05)%>%
  left_join(.,DEG.combine.all[,c(1,17:24,34:43)],by="id")%>%#21 FDR<0.05 from combined sleep and wake genes 
  left_join(.,t.test,by="id")%>%
  dplyr::select(geneSymbol,adj.P.Val,dir,class, starts_with("SDvSS"),starts_with("aveFC.SDvSS."),starts_with("p.val_"))%>%
  dplyr::select(-ends_with("YvO"))%>%
  mutate(class=factor(class,levels=c("common","Yspec","Ospec"),labels=c("Common","Y.specific","O.specific")))%>%
arrange(dir,class)

rowan=data.frame(Class=int.ave.DEG$class,Direction=int.ave.DEG$dir)
rownames(rowan)=int.ave.DEG$geneSymbol
ann.color = list(Direction=c(sleep="hotpink",wake="midnightblue"),Class=c(Common="#00AFBB",O.specific= "#FC4E07",Y.specific = "#E7B800"),Age=c(Young="green",Old="purple"))

colan=data.frame(Age=factor(c(rep(c("Young","Old"),c(4,4)),c("Young","Old")),levels=c("Young","Old")))
rownames(colan)=colnames(int.ave.DEG[,5:14])
col.color=list(age=c(Young="green",Old="purple"))

FC_palette<-colorRampPalette(rev(brewer.pal(11,"Spectral")))(109)
colors.FC <- c(seq(-2.5,-0.51,length=20),seq(-0.5,-0.21,length=30),seq(-0.2,0.2,length=10),seq(0.21,0.5,length=30),seq(0.51,2.5,length=20))


png(filename ="inter.21gene.png",height = 19,width =8,res = 300,units = "in",type="cairo")

int.ave.DEG[,c(1,5:14)]%>%
  column_to_rownames(var="geneSymbol")%>%
  as.matrix(.)%>%
  pheatmap(.,cluster_rows = FALSE,cluster_col=FALSE,annotation_row =rowan, annotation_colors = ann.color,annotation_col =colan,breaks = colors.FC,color=FC_palette, border_color= NA,labels_col = c("3","6","9","12","3","6","9","12","aveFC","aveFC"), gaps_col=c(4,8),gaps_row=12,fontsize = 14,legend_breaks=c(-2,-0.2,0.2,2),legend_labels =c("-2","-0.2","0.2","2"))

dev.off()


#single cell marker; match genesymbol and output FDR and fold changes
scMarker.id<-read.delim("SCmarker.geneID.txt",sep="\t",header=T,stringsAsFactors = F)

scMarker.FDRandFC<-scMarker.id%>%  
  mutate(geneSymbol=gene)%>%
  left_join(.,DEG.combine.all[,c(3,5,12:16,33:37)],by="geneSymbol")

write.table(scMarker.FDRandFC,"scMarker.FDRandFC.txt",sep="\t",row.names = FALSE)



#replace ensembl ID from the DAVID result to gene symbol
FCandGO.id<-read.delim("FC.all.txt",sep="\t",header=T,stringsAsFactors = F)

# for all GO terms, convert ensembl id to genesymbol

FCandGO.id<-FCandGO.id%>%
  unite(FC.term,FC,Term,sep=".",remove=F)


GO.genename<-FCandGO.id%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,DEG.combine.all[,c(1,5)],by="id")%>%
  group_by(FC.term)%>%
  summarise(geneSymbol=paste(geneSymbol,collapse=","))%>%
  left_join(FCandGO.id,.,by="FC.term")

# for GO terms included in clusters, merge id within each cluster and convert ensembl id to genesymbol
FC.genename.pval<-FCandGO.id%>%
  filter(EC!="na")%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,DEG.combine.all[,c(1,5)],by="id")%>%
  group_by(FC)%>%
  distinct(geneSymbol,.keep_all=T)%>%
  group_by(FC)%>%
  summarise(n=n(),geneSymbol=paste(geneSymbol,collapse=","))


FC.genename.pval<-FCandGO.id%>%
  filter(EC!="na")%>%
  mutate(PValue.log=log2(PValue))%>%
  group_by(FC)%>%
  summarise(geoP=2^(mean(PValue.log)))%>%
  left_join(.,FC.genename.pval,by="FC")%>%
  mutate(logP=-log10(geoP))

FC.FDRandFC<-FCandGO.id%>%  
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,DEG.combine.all[,c(1,3,5,12:13,34:35,44:45)],by="id")%>%
  group_by(FC)%>%
  distinct(geneSymbol,.keep_all=T)

write.table(GO.genename,"GO.genename.txt",sep="\t",row.names = FALSE)
write.table(FC.genename.pval,"FC.genename.pval.txt",sep="\t",row.names = FALSE)
write.table(FC.FDRandFC,"FC.FDRandFC.txt",sep="\t",row.names = FALSE)



#specifically pool all cholesterol synthesis(GO:0006695)  and cholesterol metabolism (GO:0008203) genes

choles.id<-read.delim("choles.gene.txt",sep="\t",header=T,stringsAsFactors = F)

choles.FDRandFC<-choles.id%>%
  rename(id=ID)%>%
  left_join(.,DEG.combine.all[,c(1,3,5,12:13,34:35,44:45)],by="id")

write.table(choles.FDRandFC,"choles.FDRandFC.txt",sep="\t",row.names = FALSE)


#specifically pool all regulation of translation (GO:0006417)  and translation initiation (GO:0006413) genes

translation.gene<-read.delim("translation.gene.txt",sep="\t",header=T,stringsAsFactors = F)

translation.FDRandFC<-translation.gene%>%
  rename(id=ID)%>%
  left_join(.,DEG.combine.all[,c(1,3,5,12:13,34:35,44:45)],by="id")

write.table(translation.FDRandFC,"translation.FDRandFC.txt",sep="\t",row.names = FALSE)




#compare y and o with mean fold change (2^convert to %expression change) of all the genes involved in each functional cluster

png(filename ="wake_FC.aveFC.png",height = 4,width =8,res = 300,units = "in",type="cairo")

FC.FDRandFC%>%
  filter(EC!="na",dir=="wake")%>%
  dplyr::select(FC,class,id, starts_with("aveFC.SDvSS"))%>%
  gather(age,aveFC,aveFC.SDvSS.Y:aveFC.SDvSS.O)%>%
  mutate(age=factor(age,levels=c("aveFC.SDvSS.Y","aveFC.SDvSS.O"),labels=c("Young","Old")),aveFC=(2^aveFC-1)*100, FCname=factor(FC,levels=c("w.c.ER", "w.c.trans","w.c.endo", "w.c.dephos", "w.c.circadian", "w.c.FoxO", "w.c.cAMP", "w.c.MAPK", "w.c.Kinase", "w.c.hormone", "w.c.Dioxygenase", "w.c.fatcell", "w.y.actin", "w.y.GTPase", "w.o.lipid", "w.o.trans", "w.o.Tcell "), labels=c("ERstress", "Reg.transcription","Response to VEGF", "Dephosphorylation", "RhythmicProcess", "FoxO sigaling", "Response to cAMP", "MAPK signaling", "Kinase activity", "Cytokine activity", "Dioxygenase", "Fat cell differentiation", "Actin bindng", "GTPase activation","Lipid transport", "Regu.transcription", "Tcell receptor")))%>%
  ggbarplot(x="FCname",y="aveFC",color="age",palette=brewer.pal(9,"Blues")[c(6,8)],size=1,width=0.5,position=position_dodge(width=0.65),add="mean_se",xlab="",ylab="Relative expression change between SD and SS (%)",title="Expression changes between SD and SS for functions enriched among wake genes")+
  stat_compare_means(aes(group = age), paired=T,label="p.signif",label.y=100,hide.ns = TRUE)+
theme(legend.position=c(0.95,0.9),legend.title=element_blank(), axis.text.x = element_text(angle=35,hjust=1,size=8,face="bold",color=rep(c("black", "#E7B800", "#FC4E07"),c(12,2,3))),  plot.title=element_text(face="bold",size=10,hjust=0.5), axis.title.y=element_text(face="bold",size=8))

dev.off()


transmembrane<-FC.FDRandFC%>%
  filter(FC=="s.c.Signal"|FC=="s.o.Transmem"|FC=="s.y.Signal")%>%
ungroup(FC)%>%
  mutate(FC=rep("Transmem.combined"))
  


sleep.p1<-rbind(as.data.frame(FC.FDRandFC),as.data.frame(transmembrane))%>%
  filter(FC%in%c("s.c.DNA", "s.c.Signal", "s.c.Gprotein", "s.y.adhesion", "s.y.axon", "s.y.Signal", "s.y.kinase", "s.y.chloride", "s.o.Glycoprotein", "s.o.Transmem", "s.o.Calcium","Transmem.combined"))%>%
  dplyr::select(FC,class,id, starts_with("aveFC.SDvSS"))%>%
  gather(age,aveFC,aveFC.SDvSS.Y:aveFC.SDvSS.O)%>%
  mutate(age=factor(age,levels=c("aveFC.SDvSS.Y","aveFC.SDvSS.O"),labels=c("Young","Old")),aveFC=(2^aveFC-1)*100, FCname=factor(FC,levels=c("s.c.DNA", "s.c.Signal", "s.c.Gprotein", "s.y.adhesion", "s.y.axon", "s.y.Signal", "s.y.kinase", "s.y.chloride", "s.o.Glycoprotein", "s.o.Transmem", "s.o.Calcium","Transmem.combined"), labels=c("DNA repair", "Signal/transmembrane", "Gprotein receptor", "Cell adhesion", "Axon guidance", "Signal/Transmembrane", "Kinase activity", "Chloride channel", "Collagen/glycoprotein", "Transmembrane", "Calcium transport", "Transmem.combined")))%>%
  mutate(pos.FC=-1*aveFC)%>%
  ggbarplot(x="FCname",y="pos.FC",color="age",palette=brewer.pal(9,"Reds")[c(6,9)],size=1.5,width=0.5,position=position_dodge(width=0.65),ylim=c(0,40),add="mean_se",xlab="",ylab="Expression change between SS and SD (%)",title="Expression changes between SS and SD for functions enriched among sleep genes")+
    geom_hline(yintercept = 0,linetype="dashed")+
  stat_compare_means(aes(group = age), paired=T,label="p.signif",label.y=35,hide.ns = TRUE)+
theme(legend.direction = "horizontal",legend.position=c(0.85,0.95),legend.title=element_blank(), axis.text.x = element_text(angle=35,hjust=1,size=10,face="bold",color=rep(c("black", "#E7B800", "#FC4E07","blue"),c(3,5,3,1))),  plot.title=element_text(face="bold",size=12,hjust=0.5), axis.title.y=element_text(face="bold",size=8))



sleep.p2<-rbind(as.data.frame(FC.FDRandFC),as.data.frame(transmembrane))%>%
  filter(FC%in%c("s.c.DNA", "s.c.Signal", "s.c.Gprotein", "s.y.adhesion", "s.y.axon", "s.y.Signal", "s.y.kinase", "s.y.chloride", "s.o.Glycoprotein", "s.o.Transmem", "s.o.Calcium","Transmem.combined"))%>%
  dplyr::select(FC,class,id, aveYSS.vZ0,aveOSS.vZ0)%>%
  gather(age,aveFC,aveYSS.vZ0:aveOSS.vZ0)%>%
  mutate(age=factor(age,levels=c("aveYSS.vZ0","aveOSS.vZ0"),labels=c("Young","Old")),aveFC=(2^aveFC-1)*100, FCname=factor(FC,levels=c("s.c.DNA", "s.c.Signal", "s.c.Gprotein", "s.y.adhesion", "s.y.axon", "s.y.Signal", "s.y.kinase", "s.y.chloride", "s.o.Glycoprotein", "s.o.Transmem", "s.o.Calcium", "Transmem.combined"), labels=c("DNA repair", "Signal/transmembrane", "Gprotein receptor", "Cell adhesion", "Axon guidance", "Signal/Transmembrane", "Kinase activity", "Chloride channel", "Collagen/glycoprotein", "Transmembrane", "Calcium transport", "Transmem.combined")))%>%
  ggbarplot(x="FCname",y="aveFC",color="age",palette=brewer.pal(9,"Reds")[c(6,9)],size=1.5,width=0.5,position=position_dodge(width=0.65),add="mean_se",ylim=c(-10,40),xlab="",ylab="Expression change between SS and ZT0 (%)",title="Expression changes between SS and ZT0 for functions enriched among sleep genes")+
  stat_compare_means(aes(group = age), paired=T,label="p.signif",label.y=33,hide.ns = TRUE)+
  geom_hline(yintercept = 0,linetype="dashed")+
theme(legend.direction = "horizontal",legend.position=c(0.85,0.95),legend.title=element_blank(), axis.text.x = element_text(angle=35,hjust=1,size=10,face="bold",color=rep(c("black", "#E7B800", "#FC4E07","blue"),c(3,5,3,1))),  plot.title=element_text(face="bold",size=12,hjust=0.5), axis.title.y=element_text(face="bold",size=8))

png(filename ="sleep_FC.Figure4.png",height = 8,width =8,res = 300,units = "in",type="cairo")

ggarrange(sleep.p1,sleep.p2,nrow=2,labels=c("(a)","(b)"))

dev.off()


functioncluser.meanFC<-rbind(as.data.frame(FC.FDRandFC),as.data.frame(transmembrane))%>%
  filter(EC!="na")%>%
  dplyr::select(FC,class,id, starts_with("aveFC.SDvSS"), aveYSS.vZ0,aveOSS.vZ0)%>%
  mutate_at(vars(starts_with("ave")),funs(2^.))%>%
  group_by(FC)%>%
  summarise(aveSDvSS.Y=mean(aveFC.SDvSS.Y), aveSDvSS.O=mean(aveFC.SDvSS.O), aveSSvZ0.Y=mean(aveYSS.vZ0),aveSSvZ0.O=mean(aveOSS.vZ0),SDvSS.Y.se=sd(aveFC.SDvSS.Y)/sqrt(n()),SDvSS.O.se=sd(aveFC.SDvSS.O)/sqrt(n()),SSvZ0.Y.se=sd(aveYSS.vZ0)/sqrt(n()),SSvZ0.O.se=sd(aveOSS.vZ0)/sqrt(n()))
  
FC.meanFC.pairedwilcox<-rbind(as.data.frame(FC.FDRandFC),as.data.frame(transmembrane))%>%
  filter(EC!="na")%>%
  dplyr::select(FC,class,id, starts_with("aveFC.SDvSS"), aveYSS.vZ0,aveOSS.vZ0)%>%
  mutate_at(vars(starts_with("ave")),funs(2^.))%>%
  gather(con,aveFC,aveFC.SDvSS.Y:aveOSS.vZ0)%>%
  mutate(age=ifelse(grepl("Y",con),"young","old"),comparison=ifelse(grepl("SDvSS",con),"SDvSS","SSvZ0"))%>%
  dplyr::select(-con)%>%
  spread(comparison,aveFC)%>%
  group_by(FC)%>%
  summarise_at(vars(SDvSS:SSvZ0),funs(wilcox.test(.[age=="young"],.[age=="old"],paired = TRUE)$p.value))%>%
  left_join(functioncluser.meanFC,.,by="FC")%>%
  mutate(FCname=factor(FC,levels=c("w.c.ER", "w.c.trans","w.c.endo", "w.c.dephos", "w.c.circadian", "w.c.FoxO", "w.c.cAMP", "w.c.MAPK", "w.c.Kinase", "w.c.hormone", "w.c.Dioxygenase", "w.c.fatcell", "w.y.actin", "w.y.GTPase", "w.o.lipid", "w.o.trans", "w.o.Tcell ","s.c.DNA", "s.c.Signal", "s.c.Gprotein", "s.y.adhesion", "s.y.axon", "s.y.Signal", "s.y.kinase", "s.y.chloride", "s.o.Glycoprotein", "s.o.Transmem", "s.o.Calcium", "s.chol", "s.translation","Transmem.combined"), labels=c("ERstress", "Reg.transcription","Response to VEGF", "Dephosphorylation", "RhythmicProcess", "FoxO sigaling", "Response to cAMP", "MAPK signaling", "Kinase activity", "Cytokine activity", "Dioxygenase", "Fat cell differentiation", "Actin bindng", "GTPase activation","Lipid transport", "Regu.transcription", "Tcell receptor","DNA repair", "Signal/transmembrane", "Gprotein receptor", "Cell adhesion", "Axon guidance", "Signal/Transmembrane", "Kinase activity", "Chloride channel", "Collagen/glycoprotein", "Transmembrane", "Calcium transport", "Cholesterol synthesis", "Translation","Transmem.combined")))

write.table(FC.meanFC.pairedwilcox,"FC.meanFC.pairedwilcox.txt",sep="\t",row.names = FALSE)

```


```{r}


#cell type specific marker; gene set enrichment

SC.marker<-read.delim("singlecell_marker.txt",header = T,stringsAsFactors = F,row.names = NULL,na.strings = "NA")

SC.marker.long<-SC.marker%>%
  gather(.,key=geneset,value=geneSymbol)%>%
  left_join(.,v.norm$genes,by="geneSymbol")%>%
  filter(!is.na(ENTREZID))%>%
  group_by(geneset)%>%
  summarise(marker=list(unique(id)))
  
SC.marker.id<-SC.marker.long$marker
names(SC.marker.id)<-SC.marker.long$geneset
  
SC.marker.ind <-ids2indices(SC.marker.id, v.norm$gene$id)

contrasts.main<-makeContrasts(
  OvY.SS3h=O.SS.3hr-Y.SS.3hr,
  OvY.SS6h=O.SS.6hr-Y.SS.6hr,
  OvY.SS9h=O.SS.9hr-Y.SS.9hr,
  OvY.SS12h=O.SS.12hr-Y.SS.12hr,
  OvY.SD3h=O.SD.3hr-Y.SD.3hr,
  OvY.SD6h=O.SD.6hr-Y.SD.6hr,
  OvY.SD9h=O.SD.9hr-Y.SD.9hr,
  OvY.SD12h=O.SD.12hr-Y.SD.12hr,
  OvY.Z0=O.Z0.Z0-Y.Z0.Z0,
  SDvsSS_3hr.Y = Y.SD.3hr-Y.SS.3hr, 
   SDvsSS_6hr.Y = Y.SD.6hr-Y.SS.6hr,
   SDvsSS_9hr.Y = Y.SD.9hr-Y.SS.9hr,
   SDvsSS_12hr.Y = Y.SD.12hr-Y.SS.12hr,
  SDvsSS_3hr.O = O.SD.3hr-O.SS.3hr, 
   SDvsSS_6hr.O = O.SD.6hr-O.SS.6hr,
   SDvsSS_9hr.O = O.SD.9hr-O.SS.9hr,
   SDvsSS_12hr.O = O.SD.12hr-O.SS.12hr,
  aveYSS.vZ0=(Y.SS.3hr+Y.SS.6hr+Y.SS.9hr+Y.SS.12hr)/4-Y.Z0.Z0,
  aveOSS.vZ0=(O.SS.3hr+O.SS.6hr+O.SS.9hr+O.SS.12hr)/4-O.Z0.Z0,
  SSvZ0.Y3h=Y.SS.3hr-Y.Z0.Z0,
  SSvZ0.Y6h=Y.SS.6hr-Y.Z0.Z0,
  SSvZ0.Y9h=Y.SS.9hr-Y.Z0.Z0,
  SSvZ0.Y12h=Y.SS.12hr-Y.Z0.Z0,
  SSvZ0.O3h=O.SS.3hr-O.Z0.Z0,
  SSvZ0.O6h=O.SS.6hr-O.Z0.Z0,
  SSvZ0.O9h=O.SS.9hr-O.Z0.Z0,
  SSvZ0.O12h=O.SS.12hr-O.Z0.Z0,
  levels=design)

#camera results with single-cell marker (SC.marker.ind); name starts with c.sc. (stands for camera.single-cell)

c.sc.SDvSS.Y.3h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,10],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.Y.6h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,11],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.Y.9h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,12],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.Y.12h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,13],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.O.3h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,14],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.O.6h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,15],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.O.9h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,16],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.O.12h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,17],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")


c.sc.SDvSS.comb<-c.sc.SDvSS.Y.3h%>%
  left_join(.,c.sc.SDvSS.Y.6h,by="sc.geneset",suffix=c(".Y3h",".Y6h"))%>%
  left_join(.,c.sc.SDvSS.Y.9h,by="sc.geneset")%>%
  left_join(.,c.sc.SDvSS.Y.12h,by="sc.geneset",suffix=c(".Y9h",".Y12h"))%>%
  left_join(.,c.sc.SDvSS.O.3h,by="sc.geneset")%>%
  left_join(.,c.sc.SDvSS.O.6h,by="sc.geneset",suffix=c(".O3h",".O6h"))%>%
  left_join(.,c.sc.SDvSS.O.9h,by="sc.geneset")%>%
  left_join(.,c.sc.SDvSS.O.12h,by="sc.geneset",suffix=c(".O9h",".O12h"))%>%
  mutate(Direction.Y3h=ifelse(Direction.Y3h=="Up",1,-1),Direction.Y6h=ifelse(Direction.Y6h=="Up",1,-1),Direction.Y9h=ifelse(Direction.Y9h=="Up",1,-1),Direction.Y12h=ifelse(Direction.Y12h=="Up",1,-1),Direction.O3h=ifelse(Direction.O3h=="Up",1,-1),Direction.O6h=ifelse(Direction.O6h=="Up",1,-1),Direction.O9h=ifelse(Direction.O9h=="Up",1,-1),Direction.O12h=ifelse(Direction.O12h=="Up",1,-1))%>%
  mutate(dir.P.Y3h=-log10(FDR.Y3h)*Direction.Y3h,dir.P.Y6h=-log10(FDR.Y6h)*Direction.Y6h,dir.P.Y9h=-log10(FDR.Y9h)*Direction.Y9h,dir.P.Y12h=-log10(FDR.Y12h)*Direction.Y12h,dir.P.O3h=-log10(FDR.O3h)*Direction.O3h,dir.P.O6h=-log10(FDR.O6h)*Direction.O6h,dir.P.O9h=-log10(FDR.O9h)*Direction.O9h,dir.P.O12h=-log10(FDR.O12h)*Direction.O12h)%>%
  dplyr::select(sc.geneset,NGenes.Y3h,starts_with("dir.P."))%>%
  mutate(sc.geneset=ifelse(grepl("Tasic2016_Endothelial",sc.geneset),"Tasic2016_Endothelial",ifelse(grepl("Hrvatin2018_Endothelial",sc.geneset),"Hrvatin2018_Endothelial",sc.geneset)))%>% #shorten the names
  mutate(sc.geneset.n=paste0(sc.geneset,"(",NGenes.Y3h,")"))%>% #add gene number
  dplyr::select(-sc.geneset,-NGenes.Y3h)%>% 
  mutate(celltype=ifelse(grepl("OPC",sc.geneset.n),"OPC",ifelse(grepl("OLs",sc.geneset.n),"OL",ifelse(grepl("Microglia",sc.geneset.n),"Microglia",ifelse(grepl("neuron",sc.geneset.n),"neuron",ifelse(grepl("Astrocytes",sc.geneset.n),"Astrocyte",ifelse(grepl("Endothelial",sc.geneset.n),"Endo",ifelse(grepl("Inhi",sc.geneset.n),"Inhibitory","Excitatory"))))))))%>%
  mutate(celltype=factor(celltype,levels=c("Microglia","OL","OPC","Astrocyte","Endo","neuron","Inhibitory","Excitatory")))%>%
  arrange(celltype)

annotation.col=data.frame(celltype=c.sc.SDvSS.comb$celltype)
rownames(annotation.col)=c.sc.SDvSS.comb$sc.geneset.n

annotation.row=data.frame(comparison=rep(c("SDvSS.Y","SDvSS.O"),c(4,4)))
annotation.row$comparison=factor(annotation.row$comparison,levels=c("SDvSS.Y","SDvSS.O"))
rownames(annotation.row)=colnames(c.sc.SDvSS.comb)[1:8]

library("pheatmap")
my_palette.camera<-colorRampPalette(rev(brewer.pal(11,"RdBu")))(219)

colors.camera <- c(seq(-7,-5.1,length=20),seq(-5,-1.3,length=80),seq(-1.29,1.29,length=20),seq(1.3,5,length=80),seq(5.1,7,length=20))


png(filename = "scmarker.SDvSS.png",height = 7,width =12,res = 300,units = "in",type="cairo")

c.sc.SDvSS.comb%>%
  dplyr::select(-celltype)%>%
  column_to_rownames(.,var="sc.geneset.n")%>%
  t%>%
  pheatmap(., breaks = colors.camera, color=my_palette.camera, cluster_cols=FALSE,cluster_rows=FALSE,border_color= NA,annotation_row =annotation.row, annotation_col=annotation.col, gaps_row=4, gaps_col=12,labels_row=c("ZT3","ZT6","ZT9","ZT12","ZT3","ZT6","ZT9","ZT12"),fontsize = 14.5,legend_breaks=c(-5,-1.3,1.3,5),legend_labels =c("e-5","0.05","0.05","e-5"), main="(b) Enrichment of cell-type specific gene sets in sleep and wake regulation")

dev.off()


c.sc.OvY.SS.3h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,1],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SS.6h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,2],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SS.9h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,3],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SS.12h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,4],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SD.3h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,5],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SD.6h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,6],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SD.9h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,7],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SD.12h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,8],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")

c.sc.OvY.ZT0<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,9],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")

c.sc.OvY.comb<-c.sc.OvY.SS.3h%>%
  left_join(.,c.sc.OvY.SS.6h,by="sc.geneset",suffix=c(".SS3h",".SS6h"))%>%
  left_join(.,c.sc.OvY.SS.9h,by="sc.geneset")%>%
  left_join(.,c.sc.OvY.SS.12h,by="sc.geneset",suffix=c(".SS9h",".SS12h"))%>%
  left_join(.,c.sc.OvY.SD.3h,by="sc.geneset")%>%
  left_join(.,c.sc.OvY.SD.6h,by="sc.geneset",suffix=c(".SD3h",".SD6h"))%>%
  left_join(.,c.sc.OvY.SD.9h,by="sc.geneset")%>%
  left_join(.,c.sc.OvY.SD.12h,by="sc.geneset",suffix=c(".SD9h",".SD12h"))%>%
  left_join(.,c.sc.OvY.ZT0,by="sc.geneset")%>%
  mutate(Direction.SS3h=ifelse(Direction.SS3h=="Up",1,-1),Direction.SS6h=ifelse(Direction.SS6h=="Up",1,-1),Direction.SS9h=ifelse(Direction.SS9h=="Up",1,-1),Direction.SS12h=ifelse(Direction.SS12h=="Up",1,-1),Direction.SD3h=ifelse(Direction.SD3h=="Up",1,-1),Direction.SD6h=ifelse(Direction.SD6h=="Up",1,-1),Direction.SD9h=ifelse(Direction.SD9h=="Up",1,-1),Direction.SD12h=ifelse(Direction.SD12h=="Up",1,-1),Direction.ZT0=ifelse(Direction=="Up",1,-1))%>%
  mutate(dir.P.SS3h=-log10(PValue.SS3h)*Direction.SS3h,dir.P.SS6h=-log10(PValue.SS6h)*Direction.SS6h,dir.P.SS9h=-log10(PValue.SS9h)*Direction.SS9h,dir.P.SS12h=-log10(PValue.SS12h)*Direction.SS12h,dir.P.SD3h=-log10(PValue.SD3h)*Direction.SD3h,dir.P.SD6h=-log10(PValue.SD6h)*Direction.SD6h,dir.P.SD9h=-log10(PValue.SD9h)*Direction.SD9h,dir.P.SD12h=-log10(PValue.SD12h)*Direction.SD12h,dir.P.ZT0=-log10(PValue)*Direction.ZT0)%>%
  dplyr::select(sc.geneset,NGenes,starts_with("dir.P."))%>%
  mutate(sc.geneset=ifelse(grepl("Tasic2016_Endothelial",sc.geneset),"Tasic2016_Endothelial",ifelse(grepl("Hrvatin2018_Endothelial",sc.geneset),"Hrvatin2018_Endothelial",sc.geneset)))%>%
  mutate(sc.geneset.n=paste0(sc.geneset,"(",NGenes,")"))%>%
  dplyr::select(-sc.geneset,-NGenes)%>%
  mutate(celltype=ifelse(grepl("OPC",sc.geneset.n),"OPC",ifelse(grepl("OLs",sc.geneset.n),"OL",ifelse(grepl("Microglia",sc.geneset.n),"Microglia",ifelse(grepl("neuron",sc.geneset.n),"neuron",ifelse(grepl("Astrocytes",sc.geneset.n),"Astrocyte",ifelse(grepl("Endothelial",sc.geneset.n),"Endo",ifelse(grepl("Inhi",sc.geneset.n),"Inhibitory","Excitatory"))))))))%>%
  mutate(celltype=factor(celltype,levels=c("Microglia","OL","OPC","Astrocyte","Endo","neuron","Inhibitory","Excitatory")))%>%
  arrange(celltype)

annotation.col=data.frame(celltype=c.sc.OvY.comb$celltype)
rownames(annotation.col)=c.sc.OvY.comb$sc.geneset.n

my_palette.camera<-colorRampPalette(rev(brewer.pal(11,"RdBu")))(219)

colors.camera <- c(seq(-12,-5.1,length=20),seq(-5,-1.3,length=80),seq(-1.29,1.29,length=20),seq(1.3,5,length=80),seq(5.1,12,length=20))

png(filename = "scmarker.OvY.png",height = 6,width =12,res = 300,units = "in",type="cairo")

c.sc.OvY.comb%>%
  dplyr::select(sc.geneset.n,dir.P.ZT0,starts_with("dir.P.SS"))%>%
  column_to_rownames(.,var="sc.geneset.n")%>%
  t%>%
  pheatmap(., breaks = colors.camera, color=my_palette.camera, cluster_cols=FALSE,cluster_rows=FALSE,border_color= NA, annotation_col=annotation.col, gaps_col=12, labels_row=c("ZT0","ZT3","ZT6","ZT9","ZT12"),fontsize = 14.5,fontsize_col=15,fontsize_row=14, legend_breaks=c(-8,-1.3,1.3,8),legend_labels =c("e-8","0.05","0.05","e-8"), main="(a) Enrichment of cell-type specific gene sets in young and old comparison")

dev.off()


```












```{r}
#plot select circadian core clock genes

cir.core<-c("ENSMUSG00000020893","ENSMUSG00000055866","ENSMUSG00000028957","ENSMUSG00000055116","ENSMUSG00000020038","ENSMUSG00000068742","ENSMUSG00000059824","ENSMUSG00000022389","ENSMUSG00000003949","ENSMUSG00000021775","ENSMUSG00000020889","ENSMUSG00000029238")

  
cir.core.plot<-cbind(id=v.norm$genes$id,as.data.frame(v.norm$E))%>%
  filter(id%in%cir.core)%>%
  left_join(.,v.norm$genes,by="id")%>%
  rename(.,Young.SS.Z0.B2=Young.Z0.66902.B2,Young.SS.Z0.G4=Young.Z0.66919.G4,Young.SS.Z0.D8=Young.Z0.66938.D8,Young.SS.Z0.F4=Young.Z0.66972.F4,Young.SS.Z0.E7=Young.Z0.66988.E7,Young.SS.Z0.E11=Young.Z0.67008.E11,Old.SS.Z0.D2=Old.Z0.66904.D2,Old.SS.Z0.C5=Old.Z0.66921.C5,Old.SS.Z0.F8=Old.Z0.66940.F8,Old.SS.Z0.B5=Old.Z0.66974.B5,Old.SS.Z0.B8=Old.Z0.66990.B8,Old.SS.Z0.F11=Old.Z0.67009.F11)%>%
  mutate(.,Young.SD.Z0.B2=Young.SS.Z0.B2,Young.SD.Z0.G4=Young.SS.Z0.G4,Young.SD.Z0.D8=Young.SS.Z0.D8,Young.SD.Z0.F4=Young.SS.Z0.F4,Young.SD.Z0.E7=Young.SS.Z0.E7,Young.SD.Z0.E11=Young.SS.Z0.E11,Old.SD.Z0.D2=Old.SS.Z0.D2,Old.SD.Z0.C5=Old.SS.Z0.C5,Old.SD.Z0.F8=Old.SS.Z0.F8,Old.SD.Z0.B5=Old.SS.Z0.B5,Old.SD.Z0.B8=Old.SS.Z0.B8,Old.SD.Z0.F11=Old.SS.Z0.F11)%>%
    dplyr::select(geneSymbol,starts_with("Young"),starts_with("Old"))%>%
gather(sampleID,CPM,-geneSymbol)%>%
  mutate(age=ifelse(grepl("Young",sampleID),"Y","O"), sleep=ifelse(grepl("SS",sampleID),"SS","SD"),
         time=ifelse(grepl("Z0",sampleID),"0",ifelse(grepl("3hr",sampleID),"3",ifelse(grepl("6hr",sampleID),"6",ifelse(grepl("9hr",sampleID),"9","12")))))%>%
     unite(age.sleep,age,sleep,sep = ".",remove=F)%>%
  mutate(geneSymbol=factor(geneSymbol,levels=c("Arntl","Clock","Cry1","Cry2","Per1","Per2","Per3","Dbp","Hlf","Tef","Nr1d1","Nr1d2")),age.sleep=factor(age.sleep,levels=c("Y.SS","O.SS","Y.SD","O.SD")))

 exp.palette<-c(brewer.pal(9,"OrRd")[c(6,9)],brewer.pal(9,"GnBu")[c(6,9)])

     
cir.plot<-function(i){plot<-cir.core.plot%>%
  filter(geneSymbol==levels(cir.core.plot$geneSymbol)[i])%>%
  ggline(.,x="time",y="CPM",color="age.sleep",palette =exp.palette, linetype="age.sleep",add ="mean_se",size=0.8,plot_type="l",xlab ="",ylab = "",title=levels(cir.core.plot$geneSymbol)[i])+scale_linetype_manual(values=c("solid","longdash","solid","longdash"))+theme(legend.title=element_blank(),legend.position="right",plot.title = element_text(size=10,face="bold",hjust = 0.5), axis.text=element_text(size=8))+ scale_y_continuous(labels = scales::number_format(accuracy = 0.1))
  return(plot)}

cir.pp<-ggarrange(cir.plot(1),cir.plot(2),cir.plot(3),cir.plot(4),cir.plot(5),cir.plot(6),cir.plot(7),cir.plot(8),cir.plot(9),cir.plot(10),cir.plot(11),cir.plot(12),ncol = 4, nrow =3,common.legend = TRUE, legend = "right")

png(filename = "circadian.core.png",height = 6,width =8,res = 300,units = "in",type="cairo")
annotate_figure(cir.pp,top = text_grob("Circadian core clock genes", color = "black", face = "bold", size = 14),bottom = text_grob("Zeitgeber time (ZT)", color = "black",hjust = 0.5, face = "bold", size = 12),left = text_grob("Gene expression (logCPM)", color = "black", face = "bold",rot = 90, size = 12),fig.lab.size=14,fig.lab.face="bold")
dev.off()


cir.table<-cbind(id=v.norm$genes$id,as.data.frame(efit.main$p.value))%>%
  dplyr::select(id,starts_with("OvY.SS"))%>%
  left_join(.,DEG.combine.all[,c(1,5,12:13,33:36,44:45)],by="id")%>%
  filter(id%in%cir.core)%>%
  left_join(.,OvY.combine.result[,c(1,7:10)])


write.table(cir.table,"cir.table.txt",sep="\t",row.names = FALSE)

```
TF
Fos, ENSMUSG00000021250
Fosb, ENSMUSG00000003545
Egr1, ENSMUSG00000038418
Egr2, ENSMUSG00000037868
Egr3, ENSMUSG00000033730
Arc, ENSMUSG00000022602

Synaptic plasticity
Bdnf, ENSMUSG00000048482
Ntrk2, ENSMUSG00000055254
Gria4, ENSMUSG00000025892
(Gria2, ENSMUSG00000033981; replaced by Gria1)
Gria1, ENSMUSG00000020524
Grin3a, ENSMUSG00000039579
Grin2a, ENSMUSG00000059003
Grin2c, ENSMUSG00000020734
Cacna1b, ENSMUSG00000004113
Cacna2d1, ENSMUSG00000040118
Ephb2, ENSMUSG00000028664

synaptogenesis
Pclo, ENSMUSG00000061601
Syn2, ENSMUSG00000009394
Syn1, ENSMUSG00000037217
Nlgn1, ENSMUSG00000063887
Nrxn1, ENSMUSG00000024109
Nrxn3, ENSMUSG00000066392
Slit2, ENSMUSG00000031558
Reln, ENSMUSG00000042453
Lrp8, ENSMUSG00000028613
Vldlr, ENSMUSG00000024924

Lsg1, ENSMUSG00000022538
Gfap, ENSMUSG00000020932
Slc22a3, ENSMUSG00000023828

```{r}

#plot select synaptic homestasis genes

syn.home<-c("ENSMUSG00000021250","ENSMUSG00000003545","ENSMUSG00000038418","ENSMUSG00000037868","ENSMUSG00000033730","ENSMUSG00000022602","ENSMUSG00000048482","ENSMUSG00000055254","ENSMUSG00000025892","ENSMUSG00000020524","ENSMUSG00000039579","ENSMUSG00000059003","ENSMUSG00000020734","ENSMUSG00000004113","ENSMUSG00000040118","ENSMUSG00000028664","ENSMUSG00000061601","ENSMUSG00000009394","ENSMUSG00000037217","ENSMUSG00000063887","ENSMUSG00000024109","ENSMUSG00000066392","ENSMUSG00000031558","ENSMUSG00000042453","ENSMUSG00000028613","ENSMUSG00000024924","ENSMUSG00000022538","ENSMUSG00000020932","ENSMUSG00000023828")

  
synhome.plot<-cbind(id=v.norm$genes$id,as.data.frame(v.norm$E))%>%
  filter(id%in%syn.home)%>%
  left_join(.,v.norm$genes,by="id")%>%
  rename(.,Young.SS.Z0.B2=Young.Z0.66902.B2,Young.SS.Z0.G4=Young.Z0.66919.G4,Young.SS.Z0.D8=Young.Z0.66938.D8,Young.SS.Z0.F4=Young.Z0.66972.F4,Young.SS.Z0.E7=Young.Z0.66988.E7,Young.SS.Z0.E11=Young.Z0.67008.E11,Old.SS.Z0.D2=Old.Z0.66904.D2,Old.SS.Z0.C5=Old.Z0.66921.C5,Old.SS.Z0.F8=Old.Z0.66940.F8,Old.SS.Z0.B5=Old.Z0.66974.B5,Old.SS.Z0.B8=Old.Z0.66990.B8,Old.SS.Z0.F11=Old.Z0.67009.F11)%>%
  mutate(.,Young.SD.Z0.B2=Young.SS.Z0.B2,Young.SD.Z0.G4=Young.SS.Z0.G4,Young.SD.Z0.D8=Young.SS.Z0.D8,Young.SD.Z0.F4=Young.SS.Z0.F4,Young.SD.Z0.E7=Young.SS.Z0.E7,Young.SD.Z0.E11=Young.SS.Z0.E11,Old.SD.Z0.D2=Old.SS.Z0.D2,Old.SD.Z0.C5=Old.SS.Z0.C5,Old.SD.Z0.F8=Old.SS.Z0.F8,Old.SD.Z0.B5=Old.SS.Z0.B5,Old.SD.Z0.B8=Old.SS.Z0.B8,Old.SD.Z0.F11=Old.SS.Z0.F11)%>%
    dplyr::select(geneSymbol,starts_with("Young"),starts_with("Old"))%>%
gather(sampleID,CPM,-geneSymbol)%>%
  mutate(age=ifelse(grepl("Young",sampleID),"Y","O"), sleep=ifelse(grepl("SS",sampleID),"SS","SD"),
         time=ifelse(grepl("Z0",sampleID),"0",ifelse(grepl("3hr",sampleID),"3",ifelse(grepl("6hr",sampleID),"6",ifelse(grepl("9hr",sampleID),"9","12")))))%>%
     unite(age.sleep,age,sleep,sep = ".",remove=F)%>%
  mutate(geneSymbol=factor(geneSymbol,levels=c("Fos","Fosb","Egr1","Egr2","Egr3","Arc","Bdnf","Ntrk2","Gria1","Gria4","Grin2a","Grin2c","Grin3a","Cacna1b","Cacna2d1","Ephb2","Pclo","Syn1","Syn2","Nlgn1","Nrxn1","Nrxn3","Slit2","Reln","Lrp8","Vldlr","Lsg1","Gfap","Slc22a3")),age.sleep=factor(age.sleep,levels=c("Y.SS","O.SS","Y.SD","O.SD")))

 exp.palette<-c(brewer.pal(9,"OrRd")[c(6,9)],brewer.pal(9,"GnBu")[c(6,9)])

 SH.plot<-function(i){plot<-synhome.plot%>%
  filter(geneSymbol==levels(synhome.plot$geneSymbol)[i])%>%
  ggline(.,x="time",y="CPM",color="age.sleep",palette =exp.palette, linetype="age.sleep",add ="mean_se",size=1,plot_type="l",xlab ="",ylab = "",title=levels(synhome.plot$geneSymbol)[i])+scale_linetype_manual(values=c("solid","dashed","solid","dashed"))+theme(legend.title=element_blank(),legend.position="right",plot.title = element_text(size=14,face="bold",hjust = 0.5), axis.text=element_text(size=12))+ scale_y_continuous(labels = scales::number_format(accuracy = 0.1))
  return(plot)}

SH.TF<-ggarrange(SH.plot(1),SH.plot(2),SH.plot(3),SH.plot(4),SH.plot(6),ncol = 5, nrow =1,common.legend = F, legend ="none")

SH.SP<-ggarrange(SH.plot(7),SH.plot(8),SH.plot(9),SH.plot(10),SH.plot(11),SH.plot(12),SH.plot(13),SH.plot(14),SH.plot(15),SH.plot(16),ncol = 5, nrow =2,common.legend = TRUE, legend = "none")

SH.SG<-ggarrange(SH.plot(17),SH.plot(18),SH.plot(19),SH.plot(20),SH.plot(21),SH.plot(22),SH.plot(23),SH.plot(24),SH.plot(25),SH.plot(26),ncol = 5, nrow =2,common.legend = TRUE, legend = "none")

SH.rest<-ggarrange(SH.plot(27),SH.plot(28),SH.plot(29),ncol = 3, nrow =1,common.legend = TRUE, legend = "right")

blk<-ggplot(synhome.plot, aes(x =time, y = CPM)) + geom_blank()+theme(axis.title=element_blank(),axis.text=element_blank(),panel.background = element_blank(),axis.ticks=element_blank())

SH.pp<-ggarrange(SH.TF,SH.SP,SH.SG,ggarrange(SH.rest,blk,ncol=2,widths=c(2.1,1),labels=c("D","")),nrow=4,heights=c(1,2,2,1),labels=c("A","B","C"))



png(filename = "synaptichome.png",height = 15,width =12,res = 300,units = "in",type="cairo")
annotate_figure(SH.pp,top = text_grob("Synaptic homeostasis related genes", color = "black", face = "bold", size = 20),bottom = text_grob("Zeitgeber time (ZT)", color = "black",hjust = 0.5, face = "bold", size = 18),left = text_grob("Gene expression (logCPM)", color = "black", face = "bold",rot = 90, size = 20),fig.lab.size=20,fig.lab.face="bold")
dev.off()


sH.table<-cbind(id=v.norm$genes$id,as.data.frame(efit.main$p.value))%>%
  dplyr::select(id,starts_with("OvY.SS"))%>%
  left_join(.,DEG.combine.all[,c(1,5,12:13,33:36,44:45)],by="id")%>%
  filter(id%in%syn.home)%>%
  left_join(.,OvY.combine.result[,c(1,7:10)])


write.table(sH.table,"synapticHomeo.table.txt",sep="\t",row.names = FALSE)

```


```{r}
# compare with DEG from other microarray studies

HP.Abel<-read.delim("HP.Abel.FDR5.txt",sep="\t",header=T,stringsAsFactors = F)

overlap.HPAbel<-intersect(HP.Abel$Entrez.Gene.ID, DEG.combine.all[DEG.combine.all$FDR.Y<0.01,]$ENTREZID)
overlap.FC.HPAbel<-intersect(HP.Abel$Entrez.Gene.ID, DEG.combine.all[DEG.combine.all$FDR.Y<0.01&abs(DEG.combine.all$aveFC.SDvSS.Y)>0.2,]$ENTREZID)
overlap.FC.HPAbelFC<-intersect(HP.Abel[abs(HP.Abel$logFC.SD.NSD)>0.2,]$Entrez.Gene.ID, DEG.combine.all[DEG.combine.all$FDR.Y<0.01&abs(DEG.combine.all$aveFC.SDvSS.Y)>0.2,]$ENTREZID)

validID.HPAbel<-intersect(HP.Abel$Entrez.Gene.ID, DEG.combine.all$ENTREZID)
validID.HPAbel.FC<-intersect(HP.Abel[abs(HP.Abel$logFC.SD.NSD)>0.2,]$Entrez.Gene.ID, DEG.combine.all$ENTREZID)

HP.Abel.match<-HP.Abel%>%
  rename(ENTREZID=Entrez.Gene.ID)%>%
  left_join(.,DEG.combine.all[,c(2,12,13,34,35)],by="ENTREZID")%>%
  distinct(ENTREZID,.keep_all=T)%>%
  mutate(dir=logFC.SD.NSD*aveFC.SDvSS.Y)%>%
  mutate(FC.cut=ifelse(abs(aveFC.SDvSS.Y)>0.2|abs(aveFC.SDvSS.O)>0.2,"high","low"), FDR.cut=ifelse(FDR.Y<0.01|FDR.O<0.01,"sig","NS"),dir.comp=ifelse(dir>0,"same.dir","opposite.dir"))

HP.count<-HP.Abel.match%>%
  mutate(FDR.FC=ifelse(FDR.cut=="NS","NS",ifelse(is.na(FDR.cut),"NA",ifelse(FDR.cut=="sig"&FC.cut=="high","high","low"))))%>%
  group_by(FDR.FC,dir.comp)%>%
  summarise(n())

#51/525 unique entrez ID do not match to my study
#326/525 were also significant (FDR<1%) when no FC cutoff is used
#258/326 had abs.log2FC>0.2 (DEG used in this study; high)
#239/258 "high" had same directional change; 54/68 "low" genes had same directional change


cortex.Pack<-read.delim("Mackiewiez.cortex.txt",sep="\t",header=T,stringsAsFactors = F)

validID.pack<-intersect(cortex.Pack$Entrez.Gene, DEG.combine.all$ENTREZID)
overlap.pack<-intersect(cortex.Pack$Entrez.Gene, DEG.combine.all[DEG.combine.all$FDR.Y<0.01,]$ENTREZID)
overlap.FC.pack<-intersect(cortex.Pack$Entrez.Gene,DEG.combine.all[DEG.combine.all$FDR.Y<0.01&abs(DEG.combine.all$aveFC.SDvSS.Y)>0.2,]$ENTREZID)

cortex.FC.Pack<-read.delim("cortex.aveFC.pack.txt",sep="\t",header=T,stringsAsFactors = F)

cortex.Pack<-cortex.Pack%>%
  left_join(.,cortex.FC.Pack,by="Affymetrix.id")%>%
  mutate(ENTREZID=as.integer(Entrez.Gene))%>%
  left_join(.,DEG.combine.all[,c(2,12,13,34,35)],by="ENTREZID")%>%
  distinct(ENTREZID,.keep_all=T)%>%
  mutate(dir=aveFC*aveFC.SDvSS.Y)%>%
  mutate(FC.cut=ifelse(abs(aveFC.SDvSS.Y)>0.2|abs(aveFC.SDvSS.O)>0.2,"high","low"), FDR.cut=ifelse(FDR.Y<0.01|FDR.O<0.01,"sig","NS"),dir.comp=ifelse(dir>0,"same.dir","opposite.dir"))

cortex.count<-cortex.Pack%>%
  mutate(FDR.FC=ifelse(FDR.cut=="NS","NS",ifelse(is.na(FDR.cut),"NA",ifelse(FDR.cut=="sig"&FC.cut=="high","high","low"))))%>%
  group_by(FDR.FC,dir.comp)%>%
  summarise(n())



HP.Abel.match<-HP.Abel%>%
  rename(ENTREZID=Entrez.Gene.ID)%>%
  left_join(.,DEG.combine.all[,c(2,12,13,34,35)],by="ENTREZID")%>%
  distinct(ENTREZID,.keep_all=T)%>%
  mutate(dir=logFC.SD.NSD*aveFC.SDvSS.Y)%>%
  mutate(FC.cut=ifelse(abs(aveFC.SDvSS.Y)>0.2|abs(aveFC.SDvSS.O)>0.2,"high","low"), FDR.cut=ifelse(FDR.Y<0.01|FDR.O<0.01,"sig","NS"),dir.comp=ifelse(dir>0,"same.dir","opposite.dir"))

HP.count<-HP.Abel.match%>%
  mutate(FDR.FC=ifelse(FDR.cut=="NS","NS",ifelse(is.na(FDR.cut),"NA",ifelse(FDR.cut=="sig"&FC.cut=="high","high","low"))))%>%
  group_by(FDR.FC,dir.comp)%>%
  summarise(n())


validID.pack.FC<-intersect(cortex.Pack[abs(cortex.Pack$aveFC)>0.2,]$Entrez.Gene, DEG.combine.all$ENTREZID)
overlap.FC.packFC<-intersect(cortex.Pack[abs(cortex.Pack$aveFC)>0.2,]$Entrez.Gene, DEG.combine.all[DEG.combine.all$FDR.Y<0.01&abs(DEG.combine.all$aveFC.SDvSS.Y)>0.2,]$ENTREZID)

# %overlap with cortex DEGs (FDR<1%; no FC cutoff)= 1486/3510=42.3%
# %overlap with cortex DEGs (FDR<1%; abs.log2FC>0.2)= 581/1274=45.6%


meta.Wang<-read.delim("Wang.meta.txt",sep="\t",header=T,stringsAsFactors = F)

validID.wang<-intersect(meta.Wang$Gene.Symbol, DEG.combine.all$geneSymbol)
overlap.wang<-intersect(meta.Wang$Gene.Symbol, DEG.combine.all[DEG.combine.all$FDR.Y<0.01,]$geneSymbol)
overlap.FC.wang<-intersect(meta.Wang$Gene.Symbol, DEG.combine.all[DEG.combine.all$FDR.Y<0.01&abs(DEG.combine.all$aveFC.SDvSS.Y)>0.2,]$geneSymbol)

# %overlap with Wang meta (91 genes)= 63/76=83%
# %overlap with Wang meta (91 genes) with FC cutoff= 55/76=72.3%

wang.DEGmatch<-meta.Wang%>%
  rename(geneSymbol=Gene.Symbol)%>%
  left_join(.,DEG.combine.all[,c(5,12,13,34,35)],by="geneSymbol")%>%
  dplyr::select(-starts_with("X"))%>%
  distinct(geneSymbol,.keep_all=T)%>%
  mutate(FC.cut=ifelse(abs(aveFC.SDvSS.Y)>0.2|abs(aveFC.SDvSS.O)>0.2,"high","low"), FDR.cut=ifelse(FDR.Y<0.01|FDR.O<0.01,"sig","NS"))

write.table(wang.DEGmatch,"wang.DEGmatch.txt",sep="\t",row.names = FALSE)


```




```{r}

#age comparison
contrasts.main<-makeContrasts(
  OvY.SS3h=O.SS.3hr-Y.SS.3hr,
  OvY.SS6h=O.SS.6hr-Y.SS.6hr,
  OvY.SS9h=O.SS.9hr-Y.SS.9hr,
  OvY.SS12h=O.SS.12hr-Y.SS.12hr,
  OvY.SD3h=O.SD.3hr-Y.SD.3hr,
  OvY.SD6h=O.SD.6hr-Y.SD.6hr,
  OvY.SD9h=O.SD.9hr-Y.SD.9hr,
  OvY.SD12h=O.SD.12hr-Y.SD.12hr,
  OvY.Z0=O.Z0.Z0-Y.Z0.Z0,
  SDvsSS_3hr.Y = Y.SD.3hr-Y.SS.3hr, 
   SDvsSS_6hr.Y = Y.SD.6hr-Y.SS.6hr,
   SDvsSS_9hr.Y = Y.SD.9hr-Y.SS.9hr,
   SDvsSS_12hr.Y = Y.SD.12hr-Y.SS.12hr,
  SDvsSS_3hr.O = O.SD.3hr-O.SS.3hr, 
   SDvsSS_6hr.O = O.SD.6hr-O.SS.6hr,
   SDvsSS_9hr.O = O.SD.9hr-O.SS.9hr,
   SDvsSS_12hr.O = O.SD.12hr-O.SS.12hr,
  aveYSS.vZ0=(Y.SS.3hr+Y.SS.6hr+Y.SS.9hr+Y.SS.12hr)/4-Y.Z0.Z0,
  aveOSS.vZ0=(O.SS.3hr+O.SS.6hr+O.SS.9hr+O.SS.12hr)/4-O.Z0.Z0,
  SSvZ0.Y3h=Y.SS.3hr-Y.Z0.Z0,
  SSvZ0.Y6h=Y.SS.6hr-Y.Z0.Z0,
  SSvZ0.Y9h=Y.SS.9hr-Y.Z0.Z0,
  SSvZ0.Y12h=Y.SS.12hr-Y.Z0.Z0,
  SSvZ0.O3h=O.SS.3hr-O.Z0.Z0,
  SSvZ0.O6h=O.SS.6hr-O.Z0.Z0,
  SSvZ0.O9h=O.SS.9hr-O.Z0.Z0,
  SSvZ0.O12h=O.SS.12hr-O.Z0.Z0,
  levels=design)

efit.main<- eBayes(contrasts.fit(vfit, contrasts=contrasts.main)) 

#DEGs with F-test over four timepoints FDR<1%

DEG.OvY.SS<-topTable(efit.main,coef=1:4,n=Inf,p.value=0.01) #654 DEG with FDR<1% for Old vs. Young during sleep (SS)
DEG.OvY.SD<-topTable(efit.main,coef=5:8,n=Inf,p.value=0.01) #951 DEG with FDR<1% for Old vs. Young during sleep deprivation (SD)
DEG.OvY.ZT0<-topTable(efit.main,coef=9,n=Inf,p.value=0.01) #255 DEG with FDR<1% for Old vs. Young at ZT0 (7am)

#define old up-regulation as aveFC.SS.OvY>0.1

DEG.OvY.SS<-topTable(efit.main,coef=1:4,n=Inf) #654 DEG with FDR<1% for Old vs. Young during sleep (SS)
DEG.OvY.ZT0<-topTable(efit.main,coef=9,n=Inf) #255 DEG with FDR<1% for Old vs. Young at ZT0 (7am)


OvY.combine.result<-DEG.OvY.SS%>%
  dplyr::select(id,starts_with("OvY"),P.Value,adj.P.Val)%>%
  left_join(DEG.OvY.ZT0,.,by="id",suffix=c(".ZT0",".SS"))%>%
  dplyr::select(id,ENTREZID,description,starts_with("gene"),starts_with("P.Value"),starts_with("adj."),starts_with("OvY"),logFC)%>%
  dplyr::rename(FDR.SS=adj.P.Val.SS,FDR.ZT0=adj.P.Val.ZT0,OvY.ZT0=logFC)%>%
  mutate(aveFC.OvY.SS=rowMeans(.[,11:14]))%>%
  mutate(dir.SS=ifelse(FDR.SS<0.01&aveFC.OvY.SS>0.2,"O-up",ifelse(FDR.SS<0.01&aveFC.OvY.SS<(-0.2),"O-down",ifelse(FDR.SS<0.01,"smallFC","NS"))),dir.Z0=ifelse(FDR.ZT0<0.01&OvY.ZT0>0.2,"O-up",ifelse(FDR.ZT0<0.01&OvY.ZT0<(-0.2),"O-down",ifelse(FDR.ZT0<0.01,"smallFC","NS"))))%>%
  mutate(dir=ifelse(dir.SS=="O-down"|dir.Z0=="O-down","O-down",ifelse(dir.SS=="O-up"|dir.Z0=="O-up","O-up","NS")),class=ifelse(dir.SS=="O-down"&dir.Z0=="O-down","common",ifelse(dir.SS=="O-up"&dir.Z0=="O-up","common",ifelse(dir.SS=="O-down"&dir.Z0!="O-down","SS.only",ifelse(dir.SS=="O-up"&dir.Z0!="O-up","SS.only",ifelse(dir.SS!="O-down"&dir.Z0=="O-down","Z0.only",ifelse(dir.SS!="O-up"&dir.Z0=="O-up","Z0.only","NS")))))))
  

write.table(OvY.combine.result,"OvY.combine.result.txt",sep="\t",row.names = F)


#volcano plot

vol.SS<-OvY.combine.result%>%
  mutate(class.SS=ifelse(FDR.SS>0.01,"ns",ifelse(abs(aveFC.OvY.SS)>0.2,"high","low")),class.Z0=ifelse(FDR.ZT0>0.01,"ns",ifelse(abs(OvY.ZT0)>0.2,"high","low")))%>%
  mutate_at(vars(starts_with("FDR")),funs(-log10(.)))%>%
ggscatter(.,x="aveFC.OvY.SS",y="FDR.SS",size=1,color="class.SS", palette=c("sienna2","gray50","gray80"),xlab=expression('Average log'[2]*'fold change between old and young'), ylab=expression('-log'[10]*'FDR'),title="Age affected DEGs during SS",xlim=c(-5,5))+
geom_hline(yintercept=2, linetype="longdash", color = "black",size=0.5)+
  geom_vline(xintercept=c(-0.2,0.2), linetype="longdash", color = c("green4","red"),size=0.5)+
  scale_x_continuous(breaks=c(-4,-2,-0.2,0.2,2,4))+scale_y_continuous(breaks=c(0,2,10,20,30))+
  theme(legend.position = "none",axis.title=element_text(size=10,face="bold"), axis.text.x=element_text(angle = 60,hjust = 1),plot.title=element_text(face="bold",size=16,hjust=0.5))+
  annotate("segment", x = -0.2, xend = -1, y = 24, yend = 24, colour = "green4", size=1, alpha=0.5, arrow=arrow(length = unit(0.5, "cm")))+
  annotate("segment", x = 0.2, xend = 1, y = 24, yend = 24, colour = "red", size=1, alpha=0.5, arrow=arrow(length = unit(0.5, "cm")))+
  annotate("text", x = c(-3.3,3), y = c(24,24), label = c("195", "341") , color=c("green4","red"), size=7)+
  annotate("text", x = c(-3.3,3), y = c(20,20), label = c("Down", "Up") , color=c("green4","red"), size=8,fontface="bold")

vol.Z0<-OvY.combine.result%>%
  mutate(class.SS=ifelse(FDR.SS>0.01,"ns",ifelse(abs(aveFC.OvY.SS)>0.2,"high","low")),class.Z0=ifelse(FDR.ZT0>0.01,"ns",ifelse(abs(OvY.ZT0)>0.2,"high","low")))%>%
  mutate_at(vars(starts_with("FDR")),funs(-log10(.)))%>%
ggscatter(.,x="OvY.ZT0",y="FDR.ZT0",size=1,color="class.Z0", palette=c("purple","gray50","gray80"),xlab=expression('log'[2]*'fold change between old and young'), ylab=expression('-log'[10]*'FDR'),title="Age affected DEGs at ZT0",xlim=c(-5,5))+
geom_hline(yintercept=2, linetype="longdash", color = "black",size=0.5)+
  geom_vline(xintercept=c(-0.2,0.2), linetype="longdash", color = c("green4","red"),size=0.5)+
  scale_x_continuous(breaks=c(-4,-2,-0.2,0.2,2,4))+scale_y_continuous(breaks=c(0,2,5,10,15))+
  theme(legend.position = "none",axis.title=element_text(size=10,face="bold"), axis.text.x=element_text(angle = 60,hjust = 1),plot.title=element_text(face="bold",size=16,hjust=0.5))+
  annotate("segment", x = -0.2, xend = -1, y = 11, yend = 11, colour = "green4", size=1, alpha=0.5, arrow=arrow(length = unit(0.5, "cm")))+
  annotate("segment", x = 0.2, xend = 1, y = 11, yend = 11, colour = "red", size=1, alpha=0.5, arrow=arrow(length = unit(0.5, "cm")))+
  annotate("text", x = c(-3.3,3), y = c(11,11), label = c("108", "144") , color=c("green4","red"), size=7)+
  annotate("text", x = c(-3.3,3), y = c(9,9), label = c("Down", "Up") , color=c("green4","red"), size=8,fontface="bold")


png(filename ="volcano.YvsO.png",height = 4,width =8,res = 300,units = "in",type="cairo")
ggarrange(vol.SS,vol.Z0,ncol=2,labels=c("(f)",""))
   dev.off() 
 

#O-up; O-down genes at SS and ZT0; without FC>0.2 cutoff

venn.diagram(list(OvY.combine.result[OvY.combine.result$FDR.ZT0<0.01,]$id,OvY.combine.result[OvY.combine.result$FDR.SS<0.01,]$id), "OvY.SSandZ0.tiff", fill=brewer.pal(5,"Set1")[4:5], cex = 3.5, main="(e) DEGs between young and old", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.85,main.cex=2.6, cat.cex=3,category.names=c("ZT0","SS"),cat.pos=c(-140,120))

venn.diagram(list(OvY.combine.result[OvY.combine.result$FDR.SS<0.01&OvY.combine.result$aveFC.OvY.SS>0,]$id,OvY.combine.result[OvY.combine.result$FDR.ZT0<0.01&OvY.combine.result$OvY.ZT0>0,]$id), "Oup.SSandZ0.tiff", fill=brewer.pal(3,"Dark2")[1:2], cex = 3.5, main="Age up-regulated genes", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.85,main.cex=3, cat.cex=3,category.names=c("SS","ZT0"),cat.pos=c(-55,50))

venn.diagram(list(OvY.combine.result[OvY.combine.result$FDR.SS<0.01&OvY.combine.result$aveFC.OvY.SS<0,]$id,OvY.combine.result[OvY.combine.result$FDR.ZT0<0.01&OvY.combine.result$OvY.ZT0<0,]$id), "Odown.SSandZ0.tiff", fill=brewer.pal(4,"Dark2")[3:4], cex = 3.5, main="Age down-regulated genes", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.85,main.cex=3, cat.cex=3,category.names=c("SS","ZT0"),cat.pos=c(-55,50))

#O-up; O-down genes at SS and ZT0; with FC>0.2 cutoff

venn.diagram(list(OvY.combine.result[OvY.combine.result$dir.SS=="O-up",]$id,OvY.combine.result[OvY.combine.result$dir.Z0=="O-up",]$id), "Oup.SSandZ0.FC0.2.tiff", fill=brewer.pal(9,"Purples")[c(5,8)], cex = 3.5, main="Age up-regulated genes", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.85,main.cex=3, cat.cex=3,category.names=c("SS","ZT0"),cat.pos=c(-55,50))


venn.diagram(list(OvY.combine.result[OvY.combine.result$dir.SS=="O-down",]$id,OvY.combine.result[OvY.combine.result$dir.Z0=="O-down",]$id), "Odown.SSandZ0.FC0.2.tiff", fill=brewer.pal(9,"Greens")[c(5,8)], cex = 3.5, main="Age down-regulated genes", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.85,main.cex=3, cat.cex=3,category.names=c("SS","ZT0"),cat.pos=c(-45,45))

#replace ensembl ID from the DAVID result to gene symbol
OvY.GO.id<-read.delim("OvY.BP.all.txt",sep="\t",header=T,stringsAsFactors = F)



OvY.GO.genename<-OvY.GO.id%>%
  unite(search.term,search,Term,remove=F)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,OvY.combine.result[,c(1,5)],by="id")%>%
  dplyr::select(-Genes)%>%
  group_by(search.term)%>%
  mutate(geneSymbol=paste(geneSymbol,collapse=","))%>%
  distinct(Term,.keep_all=T)

OvY.GO.FDRandFC<-OvY.GO.id%>%  
  unite(search.term,search,Term,remove=F)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,OvY.combine.result[,c(1,3,5,9:10,15:16,20)],by="id")%>%
  group_by(search.term)%>%
  distinct(geneSymbol,.keep_all=T)


write.table(OvY.GO.genename,"OvY.GO.genename.txt",sep="\t",row.names = FALSE)
write.table(OvY.GO.FDRandFC,"OvY.GO.FDRandFC.txt",sep="\t",row.names = FALSE)





#choose top unique GO terms and take their unique genes for heatmap
Oup.GO<-DAVID.OvY.enseml%>%
  mutate(dir=ifelse(grepl("up",search),"Oup","Odown"),con=ifelse(grepl("SS",search),"SS","Z0"),cat=ifelse(grepl("BP",Category),"BP","KEGG"))%>%
  filter(cat=="BP"&PValue<0.01&dir=="Oup")%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,5)],by="id")%>%
  distinct(id,.keep_all=T)%>%
  mutate(Term.b=ifelse(geneSymbol=="C3","GO:0002376~immune system process",Term))%>%
  mutate(Term.factor=as.factor(Term.b))%>%
mutate(class=rep("Oup.GO"))

SSvsZT0.FC<-cbind(as.data.frame(efit.main$genes),as.data.frame(efit.main$coefficients))%>%dplyr::select(id,geneSymbol,starts_with("SSvZ0"))

Oup.heatmap<-Oup.GO%>%
  filter(as.numeric(Term.factor)%in%c(1,2,3,4,5,6,7,8,9,10))%>%
  left_join(.,OvY.combine.result[,c(1,11:16)],by="id")%>%
  filter(as.numeric(Term.factor)==10|aveFC.OvY.SS>0.5|OvY.ZT0>0.5)

  
Oup.h1<-Oup.heatmap%>%
filter(as.numeric(Term.factor)%in%c(1,2,3,5,6,7,9))%>%
  dplyr::select(geneSymbol,OvY.ZT0,starts_with("OvY"))%>%
column_to_rownames(.,var="geneSymbol")

Oup.h2<-Oup.heatmap%>%
filter(as.numeric(Term.factor)==8)%>%
  dplyr::select(geneSymbol,OvY.ZT0,starts_with("OvY"))%>%
column_to_rownames(.,var="geneSymbol")

Oup.h3<-Oup.heatmap%>%
filter(as.numeric(Term.factor)==4)%>%
  dplyr::select(geneSymbol,OvY.ZT0,starts_with("OvY"))%>%
column_to_rownames(.,var="geneSymbol")

Oup.h4<-Oup.heatmap%>%
filter(as.numeric(Term.factor)==10)%>%
  dplyr::select(geneSymbol,OvY.ZT0,starts_with("OvY"))%>%
column_to_rownames(.,var="geneSymbol")

Oup.fit1<-hclust(dist(as.matrix(Oup.h1)))
Oup.fit2<-hclust(dist(as.matrix(Oup.h2)))
Oup.fit3<-hclust(dist(as.matrix(Oup.h3)))
Oup.fit4<-hclust(dist(as.matrix(Oup.h4)))

Oup.hc <- as.hclust(merge(merge(merge(as.dendrogram(Oup.fit1), as.dendrogram(Oup.fit2)), as.dendrogram(Oup.fit3)),as.dendrogram(Oup.fit4)))


Oup.table<-rbind(Oup.h1,Oup.h2,Oup.h3,Oup.h4)

colors <- c(seq(-2.5,-0.11,length=50),seq(-0.1,0.1,length=10),seq(0.11,2.5,length=50))
my_palette<-colorRampPalette(rev(brewer.pal(11,"RdGy")))(110)

rowannotation <- data.frame(GO = c(rep("Immune pathways", 27), 
                                rep("Homophilic cell adhesion", 11),
                                rep("Lipid metabolic", 10),
                                rep("Pos.reg.myelination", 5)),
                                row.names = row.names(Oup.table))
rowannotation$GO=factor(rowannotation$GO,levels=c("Immune pathways","Homophilic cell adhesion","Lipid metabolic","Pos.reg.myelination"))

png(filename = "OvY.up.selectGO.flip.png",height = 3,width =15,res = 300,units = "in",type="cairo")

pheatmap(t(as.matrix(Oup.table)),cluster_rows =F ,cluster_col=Oup.hc, annotation_col = rowannotation,breaks = colors,color=my_palette, border_color= NA,labels_row = c("0","3","6","9","12"), gaps_row=1,fontsize = 14, cutree_cols = 4, treeheight_col=0,legend_breaks=c(-2,0,2),legend_labels =c("-2","0","2"))

dev.off()





#choose top unique GO terms and take their unique genes for heatmap
Odown.GO<-DAVID.OvY.enseml%>%
  mutate(dir=ifelse(grepl("up",search),"Oup","Odown"),con=ifelse(grepl("SS",search),"SS","Z0"),cat=ifelse(grepl("BP",Category),"BP","KEGG"))%>%
  filter(cat=="BP"&PValue<0.03&dir=="Odown")%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,5)],by="id")%>%
  mutate(Term.factor=as.factor(Term))%>%
  filter(as.numeric(Term.factor)%in%c(2,6,7,8,9,12,19,22,24,25,31))%>%
  distinct(id,.keep_all=T)%>%
  mutate(Term=ifelse(geneSymbol=="Egr2","GO:0048168~regulation of neuronal synaptic plasticity",Term))%>%
  mutate(Term.factor=as.factor(Term))



SSvsZT0.FC<-cbind(as.data.frame(efit.main$genes),as.data.frame(efit.main$coefficients))%>%dplyr::select(id,geneSymbol,starts_with("SSvZ0"))

Odown.heatmap<-Odown.GO%>%
  left_join(.,OvY.combine.result[,c(1,11:16)],by="id")%>%
  filter(as.numeric(Term.factor)==9|search=="Odown.SS"&abs(aveFC.OvY.SS)>0.25|search=="Odown.Z0"&abs(OvY.ZT0)>0.3)

  
Odown.h1<-Odown.heatmap%>%
filter(as.numeric(Term.factor)==7)%>%
  dplyr::select(geneSymbol,OvY.ZT0,starts_with("OvY"))%>%
column_to_rownames(.,var="geneSymbol")

Odown.h2<-Odown.heatmap%>%
filter(as.numeric(Term.factor)==8)%>%
  dplyr::select(geneSymbol,OvY.ZT0,starts_with("OvY"))%>%
column_to_rownames(.,var="geneSymbol")

Odown.h3<-Odown.heatmap%>%
filter(as.numeric(Term.factor)==1)%>%
  dplyr::select(geneSymbol,OvY.ZT0,starts_with("OvY"))%>%
column_to_rownames(.,var="geneSymbol")

Odown.h4<-Odown.heatmap%>%
filter(as.numeric(Term.factor)==9)%>%
  dplyr::select(geneSymbol,OvY.ZT0,starts_with("OvY"))%>%
column_to_rownames(.,var="geneSymbol")

Odown.h5<-Odown.heatmap%>%
filter(as.numeric(Term.factor)==2)%>%
  dplyr::select(geneSymbol,OvY.ZT0,starts_with("OvY"))%>%
column_to_rownames(.,var="geneSymbol")

Odown.h6<-Odown.heatmap%>%
filter(as.numeric(Term.factor)==6)%>%
  dplyr::select(geneSymbol,OvY.ZT0,starts_with("OvY"))%>%
column_to_rownames(.,var="geneSymbol")

Odown.h7<-Odown.heatmap%>%
filter(as.numeric(Term.factor)%in%c(3,4,5,10,11))%>%
  dplyr::select(geneSymbol,OvY.ZT0,starts_with("OvY"))%>%
column_to_rownames(.,var="geneSymbol")

Odown.fit1<-hclust(dist(as.matrix(Odown.h1)))
Odown.fit2<-hclust(dist(as.matrix(Odown.h2)))
Odown.fit3<-hclust(dist(as.matrix(Odown.h3)))
Odown.fit4<-hclust(dist(as.matrix(Odown.h4)))
Odown.fit5<-hclust(dist(as.matrix(Odown.h5)))
Odown.fit6<-hclust(dist(as.matrix(Odown.h6)))
Odown.fit7<-hclust(dist(as.matrix(Odown.h7)))


Odown.hc <- as.hclust(merge(merge(merge(merge(merge(merge(as.dendrogram(Odown.fit1), as.dendrogram(Odown.fit2)), as.dendrogram(Odown.fit3)),as.dendrogram(Odown.fit4)),as.dendrogram(Odown.fit5)),as.dendrogram(Odown.fit6)),as.dendrogram(Odown.fit7)))


Odown.table<-rbind(Odown.h1,Odown.h2,Odown.h3,Odown.h4,Odown.h5,Odown.h6,Odown.h7)

colors <- c(seq(-2,-0.11,length=50),seq(-0.1,0.1,length=10),seq(0.11,2,length=50))
my_palette<-colorRampPalette(rev(brewer.pal(11,"PRGn")))(110)

rowannotation <- data.frame(GO = c(rep("Down in Old at ZT0", 27), rep("Down in Old at SS", 29)),row.names = row.names(Odown.table))
rowannotation$GO=factor(rowannotation$GO,levels=c("Down in Old at ZT0","Down in Old at SS"))

png(filename = "OvY.down.selectGO.flip.png",height = 3,width =15,res = 300,units = "in",type="cairo")

pheatmap(t(as.matrix(Odown.table)),cluster_rows =F ,cluster_col=Odown.hc, annotation_col = rowannotation,breaks = colors,color=my_palette, border_color= NA,labels_row = c("0","3","6","9","12"), gaps_row=1,fontsize = 14, treeheight_col=0,legend_breaks=c(-1,0,1),legend_labels =c("-1","0","1"))

dev.off()

```



```{r}



contrasts.ZT0<-makeContrasts(
  SSvYZ0.Y3h=Y.SS.3hr-Y.Z0.Z0,
  SSvYZ0.Y6h=Y.SS.6hr-Y.Z0.Z0,
  SSvYZ0.Y9h=Y.SS.9hr-Y.Z0.Z0,
  SSvYZ0.Y12h=Y.SS.12hr-Y.Z0.Z0,
  SDvYZ0.Y3h=Y.SD.3hr-Y.Z0.Z0,
  SDvYZ0.Y6h=Y.SD.6hr-Y.Z0.Z0,
  SDvYZ0.Y9h=Y.SD.9hr-Y.Z0.Z0,
  SDvYZ0.Y12h=Y.SD.12hr-Y.Z0.Z0,
  SSvYZ0.O3h=O.SS.3hr-Y.Z0.Z0,
  SSvYZ0.O6h=O.SS.6hr-Y.Z0.Z0,
  SSvYZ0.O9h=O.SS.9hr-Y.Z0.Z0,
  SSvYZ0.O12h=O.SS.12hr-Y.Z0.Z0,
  SDvYZ0.O3h=O.SD.3hr-Y.Z0.Z0,
  SDvYZ0.O6h=O.SD.6hr-Y.Z0.Z0,
  SDvYZ0.O9h=O.SD.9hr-Y.Z0.Z0,
  SDvYZ0.O12h=O.SD.12hr-Y.Z0.Z0,
  Z0.OvY=O.Z0.Z0-Y.Z0.Z0,
  SSvOZ0.O3h=O.SS.3hr-O.Z0.Z0,
  SSvOZ0.O6h=O.SS.6hr-O.Z0.Z0,
  SSvOZ0.O9h=O.SS.9hr-O.Z0.Z0,
  SSvOZ0.O12h=O.SS.12hr-O.Z0.Z0,
  SDvOZ0.O3h=O.SD.3hr-O.Z0.Z0,
  SDvOZ0.O6h=O.SD.6hr-O.Z0.Z0,
  SDvOZ0.O9h=O.SD.9hr-O.Z0.Z0,
  SDvOZ0.O12h=O.SD.12hr-O.Z0.Z0,
  levels=design)

efit.newZT0<- eBayes(contrasts.fit(vfit, contrasts=contrasts.ZT0)) 

#DEGs with F-test over four timepoints FDR<1%


exp.time.YZT0<-test%>%
  dplyr::select(id,wake.0.2,sleep.0.2)%>%
  mutate(dir=ifelse(wake.0.2!="na","wake",ifelse(sleep.0.2!="na","sleep","na")), class=ifelse(wake.0.2=="common.wake"|sleep.0.2=="common.sleep","common",ifelse(wake.0.2=="Yspec.wake"|sleep.0.2=="Yspec.sleep","Yspec",ifelse(wake.0.2=="Ospec.wake"|sleep.0.2=="Ospec.sleep","Ospec","na"))))%>%
  left_join(.,cbind(id=v.norm$genes$id,as.data.frame(efit.newZT0$coefficients)),by="id")%>%
  mutate(SSvYZ0.Y0h=rep(0),SDvYZ0.Y0h=rep(0),SSvYZ0.O0h=Z0.OvY,SDvYZ0.O0h=Z0.OvY)%>%
  dplyr::select(-Z0.OvY,-ends_with("0.2"),-contains("vOZ0"))%>%
  gather(con,FCvZ0,-id,-dir,-class)%>%
  mutate(condition=ifelse(grepl("SSvYZ0.Y",con),"Y.SS",ifelse(grepl("SDvYZ0.Y",con),"Y.SD",ifelse(grepl("SSvYZ0.O",con),"O.SS","O.SD"))),
time=ifelse(grepl("0h",con),0,ifelse(grepl("3h",con),3,ifelse(grepl("6h",con),6,ifelse(grepl("9h",con),9,12)))))%>%
    mutate(FCvZ0.pc=((2^(FCvZ0)-1)*100))%>% #convert log2FC to % change relative to ZT0
  mutate(condition=factor(condition,levels=c("Y.SS","O.SS","Y.SD","O.SD")),class=factor(class,levels=c("Yspec","common","Ospec"),labels=c("Y-specific","Common","O-specific")))
  
exp.palette<-c(brewer.pal(9,"OrRd")[c(5,9)],brewer.pal(9,"GnBu")[c(5,9)])

p1<-exp.time.YZT0%>%
  filter(dir=="sleep")%>%
ggscatter(.,x="time",y="FCvZ0.pc",point=F,color="condition",palette =exp.palette,add="loess",conf.int=T, conf.int.level=0.95, facet.by="class",xlab="Zeitgeber time (ZT)", ylab="Relative Expression Change (%)")+
scale_x_continuous(breaks=c(0,3,6,9,12))+geom_hline(yintercept = 0,color="black",linetype="longdash",size=0.5)+theme(legend.position = "bottom",legend.title =element_blank(),strip.background.x = element_rect(fill = brewer.pal(9,"YlOrBr")[8]),strip.text.x = element_text(colour = "white", face = "bold"),panel.background = element_rect(colour = "black"), axis.title=element_text(size=8,face="bold"))

p2<-exp.time.YZT0%>%
  filter(dir=="wake")%>%
  ggscatter(.,x="time",y="FCvZ0.pc",point=F,color="condition",palette =exp.palette,add="loess",conf.int=T, conf.int.level=0.95, facet.by="class",xlab="Zeitgeber time (ZT)", ylab="Relative Expression Change (%)")+
scale_x_continuous(breaks=c(0,3,6,9,12))+geom_hline(yintercept = 0,color="black",linetype="longdash",size=0.5)+theme(legend.position = "bottom",legend.title =element_blank(),strip.background.x = element_rect(fill = brewer.pal(9,"PuBu")[8]),strip.text.x = element_text(colour = "white", face = "bold"),panel.background = element_rect(colour = "black"), axis.title=element_text(size=8,face="bold"))

png(filename ="exp.vYZ0.png",height = 3,width =8,res = 300,units = "in",type="cairo")
ggarrange(p1,p2,ncol=2,common.legend = T, legend="bottom")

dev.off()



exp.time.ZT0<-test%>%
  dplyr::select(id,wake.0.2,sleep.0.2)%>%
  mutate(dir=ifelse(wake.0.2!="na","wake",ifelse(sleep.0.2!="na","sleep","na")), class=ifelse(wake.0.2=="common.wake"|sleep.0.2=="common.sleep","common",ifelse(wake.0.2=="Yspec.wake"|sleep.0.2=="Yspec.sleep","Yspec",ifelse(wake.0.2=="Ospec.wake"|sleep.0.2=="Ospec.sleep","Ospec","na"))))%>%
  left_join(.,cbind(id=v.norm$genes$id,as.data.frame(efit.newZT0$coefficients)),by="id")%>%
  mutate(SSvYZ0.Y0h=rep(0),SDvYZ0.Y0h=rep(0),SSvOZ0.O0h=rep(0),SDvOZ0.O0h=rep(0))%>%
  dplyr::select(-Z0.OvY,-ends_with("0.2"),-contains("vYZ0.O"))%>%
  gather(con,FCvZ0,-id,-dir,-class)%>%
  mutate(condition=ifelse(grepl("SSvYZ0.Y",con),"Y.SS",ifelse(grepl("SDvYZ0.Y",con),"Y.SD",ifelse(grepl("SSvOZ0.O",con),"O.SS","O.SD"))),
time=ifelse(grepl("0h",con),0,ifelse(grepl("3h",con),3,ifelse(grepl("6h",con),6,ifelse(grepl("9h",con),9,12)))))%>%
    mutate(FCvZ0.pc=((2^(FCvZ0)-1)*100))%>% #convert log2FC to % change relative to ZT0
  mutate(condition=factor(condition,levels=c("Y.SS","O.SS","Y.SD","O.SD")),class=factor(class,levels=c("Yspec","common","Ospec"),labels=c("Y-specific","Common","O-specific")))
  
exp.palette<-c(brewer.pal(9,"OrRd")[c(5,9)],brewer.pal(9,"GnBu")[c(5,9)])

p1<-exp.time.ZT0%>%
  filter(dir=="sleep")%>%
ggscatter(.,x="time",y="FCvZ0.pc",point=F,color="condition",palette =exp.palette,add="loess",conf.int=T, conf.int.level=0.95, facet.by="class",xlab="Zeitgeber time (ZT)", ylab="Relative Expression Change (%)")+
scale_x_continuous(breaks=c(0,3,6,9,12))+geom_hline(yintercept = 0,color="black",linetype="longdash",size=0.5)+theme(legend.position = "bottom",legend.title =element_blank(),strip.background.x = element_rect(fill = brewer.pal(9,"YlOrBr")[8]),strip.text.x = element_text(colour = "white", face = "bold"),panel.background = element_rect(colour = "black"), axis.title=element_text(size=8,face="bold"))

p2<-exp.time.ZT0%>%
  filter(dir=="wake")%>%
  ggscatter(.,x="time",y="FCvZ0.pc",point=F,color="condition",palette =exp.palette,add="loess",conf.int=T, conf.int.level=0.95, facet.by="class",xlab="Zeitgeber time (ZT)", ylab="Relative Expression Change (%)")+
scale_x_continuous(breaks=c(0,3,6,9,12))+geom_hline(yintercept = 0,color="black",linetype="longdash",size=0.5)+theme(legend.position = "bottom",legend.title =element_blank(),strip.background.x = element_rect(fill = brewer.pal(9,"PuBu")[8]),strip.text.x = element_text(colour = "white", face = "bold"),panel.background = element_rect(colour = "black"), axis.title=element_text(size=8,face="bold"))

png(filename ="exp.Z0.png",height = 3,width =8,res = 300,units = "in",type="cairo")
ggarrange(p1,p2,ncol=2,common.legend = T, legend="bottom")

dev.off()


p1<-exp.time.ZT0%>%
  mutate(age=ifelse(grepl(".Y",con),"Young","Old"))%>%
  mutate(age=factor(age,levels=c("Young","Old")))%>%
  filter(dir=="sleep")%>%
ggscatter(.,x="time",y="FCvZ0.pc",point=F,color="condition",palette =exp.palette,add="loess",conf.int=T, conf.int.level=0.95,xlab="Zeitgeber time (ZT)", ylab="Relative Expression Change (%)")+
  facet_grid(age~class)+
scale_x_continuous(breaks=c(0,3,6,9,12))+theme(legend.position = "bottom",legend.title =element_blank(),strip.background.x = element_rect(fill = brewer.pal(9,"YlOrBr")[8]),strip.background.y = element_rect(fill ="white"),strip.text.x = element_text(colour = "white", face = "bold"),strip.text.y = element_text(colour = "black", face = "bold"),panel.background = element_rect(colour = "black"), axis.title=element_text(size=12,face="bold"))+ geom_hline(yintercept = 0,color="black",linetype="longdash",size=0.5)

p2<-exp.time.ZT0%>%
  mutate(age=ifelse(grepl(".Y",con),"Young","Old"))%>%
  mutate(age=factor(age,levels=c("Young","Old")))%>%
  filter(dir=="wake")%>%
  ggscatter(.,x="time",y="FCvZ0.pc",point=F,color="condition",palette =exp.palette,add="loess",conf.int=T, conf.int.level=0.95, xlab="Zeitgeber time (ZT)", ylab="Relative Expression Change (%)")+
facet_grid(age~class)+
scale_x_continuous(breaks=c(0,3,6,9,12))+theme(legend.position = "bottom",legend.title =element_blank(),strip.background.x = element_rect(fill = brewer.pal(9,"PuBu")[8]),strip.background.y = element_rect(fill ="white"), strip.text.x = element_text(colour = "white", face = "bold"), strip.text.y = element_text(colour = "black", face = "bold"), panel.background = element_rect(colour = "black"), axis.title=element_text(size=12,face="bold"))+ geom_hline(yintercept = 0,color="black",linetype="longdash",size=0.5)

png(filename ="exp.Z0.age.png",height = 4,width =8,res = 300,units = "in",type="cairo")
ggarrange(p1,p2,ncol=2,common.legend = T, legend="bottom")

dev.off()


#OvY temporal graphs


DEG.combine.all.OvY<-DEG.SDvSS.O.full%>%
  dplyr::select(id,P.Value,adj.P.Val,starts_with("SDvsSS"))%>%
  left_join(DEG.SDvSS.Y.full,.,by="id",suffix=c(".Y",".O"))%>%
  dplyr::select(id,starts_with("P.Value"),starts_with("adj."),starts_with("SDvsSS"))%>%
  left_join(.,DEG.OvY.SS.full,by="id")%>%
  dplyr::select(id,starts_with("P.Value"),starts_with("adj."),starts_with("SDvsSS"),starts_with("OvY"))%>%
  left_join(.,DEG.OvY.SD.full,by="id",suffix=c(".SS",".SD"))%>%
  dplyr::select(id,starts_with("P.Value"),starts_with("adj."),starts_with("SDvsSS"),starts_with("OvY"))%>%
  left_join(.,DEG.OvY.ZT0.full,by="id")%>%
  dplyr::select(id,ENTREZID,description,starts_with("gene"),starts_with("P.Value"),starts_with("adj."),starts_with("SDvsSS"),starts_with("OvY"),logFC)%>%
    dplyr::rename(FDR.Y=adj.P.Val.Y,FDR.O=adj.P.Val.O,FDR.SS.OvY=adj.P.Val.SS,FDR.SD.OvY=adj.P.Val.SD,FDR.Z0.OvY=adj.P.Val,P.Value.Z0=P.Value,OvY.Z0=logFC)%>%
  mutate(aveFC.SDvSS.Y=rowMeans(.[,17:20]),aveFC.SDvSS.O=rowMeans(.[,21:24]), aveFC.OvY.SS=rowMeans(.[,25:28]), aveFC.OvY.SD=rowMeans(.[,29:32]))%>%
  mutate(SD.up=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0&FDR.O<0.01&aveFC.SDvSS.O>0,"common",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0,"Yspec",ifelse(FDR.O<0.01&aveFC.SDvSS.O>0,"Ospec","na"))), SD.down=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<0&FDR.O<0.01&aveFC.SDvSS.O<0,"common",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<0,"Yspec",ifelse(FDR.O<0.01&aveFC.SDvSS.O<0,"Ospec","na"))),
         wake=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0.2&FDR.O<0.01&aveFC.SDvSS.O>0.2,"common",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0.2,"Yspec",ifelse(FDR.O<0.01&aveFC.SDvSS.O>0.2,"Ospec","na"))),
         sleep=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<(-0.2)&FDR.O<0.01&aveFC.SDvSS.O<(-0.2),"common",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<(-0.2),"Yspec",ifelse(FDR.O<0.01&aveFC.SDvSS.O<(-0.2),"Ospec","na"))))%>%
  mutate(dir=ifelse(wake!="na","wake",ifelse(sleep!="na","sleep","na")), class=ifelse(wake=="common"|sleep=="common","common",ifelse(wake=="Yspec"|sleep=="Yspec","Yspec",ifelse(wake=="Ospec"|sleep=="Ospec","Ospec","na"))))%>%
  left_join(.,aveFCvsZT0,by="id")



exp.time.OvYDEG.YZT0<-test%>%
  dplyr::select(id,wake.0.2,sleep.0.2)%>%
  mutate(dir=ifelse(wake.0.2!="na","wake",ifelse(sleep.0.2!="na","sleep","na")), class=ifelse(wake.0.2=="common.wake"|sleep.0.2=="common.sleep","common",ifelse(wake.0.2=="Yspec.wake"|sleep.0.2=="Yspec.sleep","Yspec",ifelse(wake.0.2=="Ospec.wake"|sleep.0.2=="Ospec.sleep","Ospec","na"))))%>%
  left_join(.,cbind(id=v.norm$genes$id,as.data.frame(efit.newZT0$coefficients)),by="id")%>%
  mutate(SSvYZ0.Y0h=rep(0),SDvYZ0.Y0h=rep(0),SSvYZ0.O0h=Z0.OvY,SDvYZ0.O0h=Z0.OvY)%>%
  dplyr::select(-Z0.OvY,-ends_with("0.2"),-contains("vOZ0"))%>%
  gather(con,FCvZ0,-id,-dir,-class)%>%
  mutate(condition=ifelse(grepl("SSvYZ0.Y",con),"Y.SS",ifelse(grepl("SDvYZ0.Y",con),"Y.SD",ifelse(grepl("SSvYZ0.O",con),"O.SS","O.SD"))),
time=ifelse(grepl("0h",con),0,ifelse(grepl("3h",con),3,ifelse(grepl("6h",con),6,ifelse(grepl("9h",con),9,12)))))%>%
    mutate(FCvZ0.pc=((2^(FCvZ0)-1)*100))%>% #convert log2FC to % change relative to ZT0
  mutate(condition=factor(condition,levels=c("Y.SS","O.SS","Y.SD","O.SD")),class=factor(class,levels=c("Yspec","common","Ospec"),labels=c("Y-specific","Common","O-specific")))



SDVSS.combine.result<-DEG.SDvSS.O.full%>%
  dplyr::select(id,starts_with("SDvsSS"),P.Value,adj.P.Val)%>%
  left_join(DEG.SDvSS.Y.full,.,by="id",suffix=c(".Y",".O"))%>%
  dplyr::select(id,ENTREZID,description,starts_with("gene"),starts_with("P.Value"),starts_with("adj."),starts_with("SDvsSS"))%>%
  dplyr::rename(FDR.Y=adj.P.Val.Y,FDR.O=adj.P.Val.O)%>%
  mutate(aveFC.SDvSS.Y=rowMeans(.[,11:14]),aveFC.SDvSS.O=rowMeans(.[,15:18]))%>%
  left_join(.,aveFCvsZT0,by="id")%>%
  mutate(class.Y=ifelse(FDR.Y<0.01&aveFC.SDvSS.Y<0&aveYSS.vZ0>0.1,"SleepGene",ifelse(FDR.Y<0.01&aveFC.SDvSS.Y>0.2,"WakeGene","na")),class.O=ifelse(FDR.O<0.01&aveFC.SDvSS.O<0&aveOSS.vZ0>0.1,"SleepGene",ifelse(FDR.O<0.01&aveFC.SDvSS.O>0.2,"WakeGene","na")))%>%
  rename(aveFC.SSvZ0.Y=aveYSS.vZ0,aveFC.SSvZ0.O=aveOSS.vZ0)%>%
  mutate(class.sum=ifelse(class.Y=="SleepGene"&class.O=="SleepGene","S.Common",ifelse(class.Y=="SleepGene","S.Y-spe",ifelse(class.O=="SleepGene","S.O-spe",ifelse(class.Y=="WakeGene"&class.O=="WakeGene","W.Common",ifelse(class.Y=="WakeGene","W.Y-spe",ifelse(class.O=="WakeGene","W.O-spe","na")))))))%>%
  mutate(dir=ifelse(grepl("S",class.sum),"SleepGenes",ifelse(grepl("W",class.sum),"WakeGenes","na")),class=ifelse(grepl("Common",class.sum),"Common",ifelse(grepl("Y-spe",class.sum),"Y-specific",ifelse(grepl("O-spe",class.sum),"O-specific","na"))))%>%
  mutate(dir=factor(dir,levels=c("SleepGenes","WakeGenes","na")),class=factor(class,levels=c("Common","Y-specific","O-specific","na")))

write.table(SDVSS.combine.result,"SDVSS.combine.result.txt",sep="\t",row.names = F)
  
library(VennDiagram)

#DEG.Y vs. DEG.O

venn.diagram(list(SDVSS.combine.result[SDVSS.combine.result$FDR.Y<0.01,]$id,SDVSS.combine.result[SDVSS.combine.result$FDR.O<0.01,]$id), "DEG.YvsO.tiff", fill=brewer.pal(9,"BuGn")[c(4,8)], cex = 2, main="DEGs between SD and SS", main.pos=c(0.5,0.64), cat.fontface=1.5, margin=0.8,main.cex=3, cat.cex=3,category.names=c("Young","Old"),cat.pos=c(-130,120))


#Sleep genes Yvs.O
venn.diagram(list(SDVSS.combine.result[SDVSS.combine.result$class.Y=="SleepGene",]$id,SDVSS.combine.result[SDVSS.combine.result$class.O=="SleepGene",]$id), "sleep.YvsO.tiff", fill=brewer.pal(9,"OrRd")[c(5,9)], cex = 3.5, main="Sleep Genes", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.85,main.cex=5, cat.cex=3,category.names=c("Young","Old"),cat.pos=c(-45,50))


#Wake genes Yvs.O
venn.diagram(list(SDVSS.combine.result[SDVSS.combine.result$class.Y=="WakeGene",]$id,SDVSS.combine.result[SDVSS.combine.result$class.O=="WakeGene",]$id), "wake.YvsO.tiff", fill=brewer.pal(9,"GnBu")[c(5,9)], cex = 3.5, main="Wake Genes", main.pos=c(0.5,0.64), cat.fontface=2, margin=0.8,main.cex=5, cat.cex=3,category.names=c("Young","Old"),cat.pos=c(-50,55))


sleep.cpm<-cbind(id=v.norm$genes$id,as.data.frame(v.norm$E))%>%
  filter(id%in%c(SDVSS.combine.result[SDVSS.combine.result$class.Y=="SleepGene",]$id,SDVSS.combine.result[SDVSS.combine.result$class.O=="SleepGene",]$id))%>%
  rename(.,Young.SS.Z0.B2=Young.Z0.66902.B2,Young.SS.Z0.G4=Young.Z0.66919.G4,Young.SS.Z0.D8=Young.Z0.66938.D8,Young.SS.Z0.F4=Young.Z0.66972.F4,Young.SS.Z0.E7=Young.Z0.66988.E7,Young.SS.Z0.E11=Young.Z0.67008.E11,Old.SS.Z0.D2=Old.Z0.66904.D2,Old.SS.Z0.C5=Old.Z0.66921.C5,Old.SS.Z0.F8=Old.Z0.66940.F8,Old.SS.Z0.B5=Old.Z0.66974.B5,Old.SS.Z0.B8=Old.Z0.66990.B8,Old.SS.Z0.F11=Old.Z0.67009.F11)%>%
  gather(sampleID,cpm,-id)%>%
  spread(id,cpm)%>%
  mutate(age=ifelse(grepl("Young",sampleID),"Y","O"), sleep=ifelse(grepl("SS",sampleID),"SS","SD"),
time=ifelse(grepl("Z0",sampleID),"0",ifelse(grepl("3hr",sampleID),"3",ifelse(grepl("6hr",sampleID),"6",ifelse(grepl("9hr",sampleID),"9","12")))))%>%
  unite(condition,age:time)%>%
  group_by(condition)%>%
  summarise_at(vars(starts_with("ENSMUSG")),mean)%>%
gather(id,meancpm,-condition)%>%
  spread(condition,meancpm)%>%
mutate(O_SD_0=O_SS_0,Y_SD_0=Y_SS_0)%>%
  mutate_at(vars(contains("Y_")),funs(.-Y_SD_0))%>%
  mutate_at(vars(contains("O_")),funs(.-O_SD_0))%>%
  left_join(.,SDVSS.combine.result[,c(1,23:24)],by="id")%>%
  mutate(class=ifelse(class.Y=="SleepGene"&class.O=="SleepGene","Common",ifelse(class.Y=="SleepGene","Y-specific","O-specific")))%>%
  dplyr::select(-starts_with("class."))%>%
  gather(con,norm.cpm,-id,-class)%>%
  separate(con,c("age","sleep","time"),sep="_")%>%
  unite(age.sleep,age:sleep)%>%
  mutate(age.sleep=factor(age.sleep,levels=c("Y_SS","O_SS","Y_SD","O_SD"),labels=c("Young.sleep","Old.sleep","Young.SD","Old.SD")),time=as.numeric(as.character(time)),class=factor(class,levels=c("Y-specific","Common","O-specific")))


  mutate(age=factor(age,levels=c("Y","O"),labels=c("Young","Old")),sleep=factor(sleep,levels=c("SS","SD")),time=as.numeric(as.character(time)),class=factor(class,levels=c("Y-specific","Common","O-specific")))
  
  mutate(age.sleep=factor(age.sleep,levels=c("Y_SS","O_SS","Y_SD","O_SD"),labels=c("Young.sleep","Old.sleep","Young.SD","Old.SD")),
         time=as.numeric(as.character(time)),class=factor(class,levels=c("Y-specific","Common","O-specific")))



wake.cpm<-cbind(id=v.norm$genes$id,as.data.frame(v.norm$E))%>%
  filter(id%in%c(SDVSS.combine.result[SDVSS.combine.result$class.Y=="WakeGene",]$id,SDVSS.combine.result[SDVSS.combine.result$class.O=="WakeGene",]$id))%>%
  rename(.,Young.SS.Z0.B2=Young.Z0.66902.B2,Young.SS.Z0.G4=Young.Z0.66919.G4,Young.SS.Z0.D8=Young.Z0.66938.D8,Young.SS.Z0.F4=Young.Z0.66972.F4,Young.SS.Z0.E7=Young.Z0.66988.E7,Young.SS.Z0.E11=Young.Z0.67008.E11,Old.SS.Z0.D2=Old.Z0.66904.D2,Old.SS.Z0.C5=Old.Z0.66921.C5,Old.SS.Z0.F8=Old.Z0.66940.F8,Old.SS.Z0.B5=Old.Z0.66974.B5,Old.SS.Z0.B8=Old.Z0.66990.B8,Old.SS.Z0.F11=Old.Z0.67009.F11)%>%
  gather(sampleID,cpm,-id)%>%
  spread(id,cpm)%>%
  mutate(age=ifelse(grepl("Young",sampleID),"Y","O"), sleep=ifelse(grepl("SS",sampleID),"SS","SD"),
time=ifelse(grepl("Z0",sampleID),"0",ifelse(grepl("3hr",sampleID),"3",ifelse(grepl("6hr",sampleID),"6",ifelse(grepl("9hr",sampleID),"9","12")))))%>%
  unite(condition,age:time)%>%
  group_by(condition)%>%
  summarise_at(vars(starts_with("ENSMUSG")),mean)%>%
gather(id,meancpm,-condition)%>%
  spread(condition,meancpm)%>%
mutate(O_SD_0=O_SS_0,Y_SD_0=Y_SS_0)%>%
  mutate_at(vars(contains("Y_")),funs(.-Y_SD_0))%>%
  mutate_at(vars(contains("O_")),funs(.-O_SD_0))%>%
  left_join(.,SDVSS.combine.result[,c(1,23:24)],by="id")%>%
  mutate(class=ifelse(class.Y=="WakeGene"&class.O=="WakeGene","Common",ifelse(class.Y=="WakeGene","Y-specific","O-specific")))%>%
  dplyr::select(-starts_with("class."))%>%
  gather(con,norm.cpm,-id,-class)%>%
  separate(con,c("age","sleep","time"),sep="_")%>%
  unite(age.sleep,age:sleep)%>%
  mutate(age.sleep=factor(age.sleep,levels=c("Y_SS","O_SS","Y_SD","O_SD"),labels=c("Young.sleep","Old.sleep","Young.SD","Old.SD")),time=as.numeric(as.character(time)),class=factor(class,levels=c("Y-specific","Common","O-specific")))


exp.palette<-c(brewer.pal(9,"OrRd")[c(5,9)],brewer.pal(9,"GnBu")[c(5,9)])

#figure 4; 12-14-18
png(filename ="exp.profile.png",height = 3,width =8,res = 300,units = "in",type="cairo")

p1<-ggscatter(sleep.cpm,x="time",y="norm.cpm",point=F,color="age.sleep",palette =exp.palette,add="loess",conf.int=T, conf.int.level=0.95, facet.by="class",xlab="Zeitgeber time (ZT)", ylab="Normalized Gene Expression", ylim=c(-0.5,0.5))+
scale_x_continuous(breaks=c(0,3,6,9,12))+geom_hline(yintercept = 0,linetype="longdash",color="black",size=0.5)+theme(legend.position = "bottom",legend.title =element_blank(),strip.background.x = element_rect(fill = brewer.pal(9,"YlOrBr")[8]),strip.text.x = element_text(colour = "white", face = "bold"),panel.background = element_rect(colour = "black"))

p2<-ggscatter(wake.cpm,x="time",y="norm.cpm",point=F,color="age.sleep",palette =exp.palette,add="loess",conf.int=T, conf.int.level=0.95, xlab="Zeitgeber time (ZT)", ylab="Normalized Gene Expression",ylim=c(-0.5,0.5))+
facet_wrap(~class)+
  scale_x_continuous(breaks=c(0,3,6,9,12))+geom_hline(yintercept = 0,linetype="longdash",color="black",size=0.5)+theme(legend.position = "bottom",legend.title =element_blank(),strip.background = element_rect(fill = brewer.pal(9,"PuBu")[8]),strip.text = element_text(colour = "white", face = "bold"),panel.background = element_rect(colour = "black"))

ggarrange(p1,p2,ncol=2,common.legend = T, legend="bottom")

dev.off()






  
expressionchange.time<-cbind(id=v.norm$genes$id,as.data.frame(efit.main$coefficients))%>%
  filter(id%in%SDVSS.combine.result[SDVSS.combine.result$dir!="na",]$id)%>%
  dplyr::select(id,starts_with("SSvZ0"),starts_with("SDvZ0"))%>%
  mutate_at(vars(-id),funs((2^(.)-1)*100))%>% #convert log2FC to % change relative to ZT0
  mutate(SSvZ0.Y0h=rep(0),SSvZ0.O0h=rep(0),SDvZ0.Y0h=rep(0),SDvZ0.O0h=rep(0))%>% #add ZT0 values as zero
  left_join(.,SDVSS.combine.result[,c(1,26:27)],by="id")%>%
  gather(con,per.change,-id,-class,-dir)%>%
  mutate(condition=ifelse(grepl("SSvZ0.Y",con),"Y.SS",ifelse(grepl("SDvZ0.Y",con),"Y.SD",ifelse(grepl("SSvZ0.O",con),"O.SS","O.SD"))),
time=ifelse(grepl("0h",con),0,ifelse(grepl("3h",con),3,ifelse(grepl("6h",con),6,ifelse(grepl("9h",con),9,12)))))%>%
  mutate(condition=factor(condition,levels=c("Y.SS","O.SS","Y.SD","O.SD")),class=factor(class,levels=c("Y-specific","Common","O-specific")))

exp.palette<-c(brewer.pal(9,"OrRd")[c(5,9)],brewer.pal(9,"GnBu")[c(5,9)])


p1<-expressionchange.time%>%
  filter(dir=="SleepGenes")%>%
ggscatter(.,x="time",y="per.change",point=F,color="condition",palette =exp.palette,add="loess",conf.int=T, conf.int.level=0.95, facet.by="class",xlab="Zeitgeber time (ZT)", ylab="Relative Expression Change (%)",ylim=c(-20,42))+
scale_x_continuous(breaks=c(0,3,6,9,12))+geom_hline(yintercept = 0,color="black",linetype="longdash",size=0.5)+theme(legend.position = "bottom",legend.title =element_blank(),strip.background.x = element_rect(fill = brewer.pal(9,"YlOrBr")[8]),strip.text.x = element_text(colour = "white", face = "bold"),panel.background = element_rect(colour = "black"), axis.title=element_text(size=8,face="bold"))

p2<-expressionchange.time%>%
  filter(dir=="WakeGenes")%>%
  ggscatter(.,x="time",y="per.change",point=F,color="condition",palette =exp.palette,add="loess",conf.int=T, conf.int.level=0.95, facet.by="class",xlab="Zeitgeber time (ZT)", ylab="Relative Expression Change (%)",ylim=c(-20,42))+
scale_x_continuous(breaks=c(0,3,6,9,12))+geom_hline(yintercept = 0,color="black",linetype="longdash",size=0.5)+theme(legend.position = "bottom",legend.title =element_blank(),strip.background.x = element_rect(fill = brewer.pal(9,"PuBu")[8]),strip.text.x = element_text(colour = "white", face = "bold"),panel.background = element_rect(colour = "black"), axis.title=element_text(size=8,face="bold"))

png(filename ="exp.profile.png",height = 3,width =8,res = 300,units = "in",type="cairo")
ggarrange(p1,p2,ncol=2,common.legend = T, legend="bottom")

dev.off()


library(png)
library(grid)
library(gridExtra)

thePlots <- lapply (2:length(names(mtcars)), function(i) {
  png("testgraph.png")
  plot(mtcars[,1], mtcars[,i])

  dev.off()
  rasterGrob(readPNG("testgraph.png", native = FALSE),
    interpolate = FALSE)
})

pdf("testgraph.pdf")
do.call(grid.arrange, c(thePlots, ncol = 3))
dev.off()

#comparing the FDR ranks of common genes vs. Y-specific and O-specific genes

P.val.rank<-SDVSS.combine.result%>%
  dplyr::select(id,class.sum,starts_with("P.Value"))%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  mutate(rank.Y=rank(P.Value.Y),rank.O=rank(P.Value.O))%>%
  filter(class.sum!="na")%>%
  gather(age,rank,rank.Y:rank.O)%>%
  mutate(age=ifelse(grepl("Y",age),"Y","O"),dir=ifelse(grepl("S",class.sum),"Sleep","Wake"),class=ifelse(grepl("Common",class.sum),"Common",ifelse(grepl("Y-spe",class.sum),"Y-specific","O-specific")))%>%
mutate(age=factor(age,levels=c("Y","O"),labels=c("Young","Old")),dir=factor(dir,levels=c("Sleep","Wake"),labels=c("Sleep Genes","Wake Genes")),class=factor(class,levels=c("Common","Y-specific","O-specific")))

class.contrast <- list( c("Common", "Y-specific"), c("Common","O-specific") )

png(filename ="rank.Pval.png",height = 5,width =5,res = 300,units = "in",type="cairo")

ggviolin(P.val.rank,x="class",y="rank",fill = "class",
   palette = c("#00AFBB", "#E7B800", "#FC4E07"), add="boxplot",add.params = list(fill = "white"), facet.by=c("age","dir"),ylim=c(0,20000),xlab="",ylab="Ranking of the -log10 (P.Value)")+
  stat_compare_means(comparisons = class.contrast,label="p.format",label.y = c(17000,19000))+
  scale_y_continuous(breaks=c(0,5000,10000,15000,20000),labels=c("0","5k","10k","15k","20k"))+
theme(legend.position="none",strip.text = element_text(face="bold", size=12),axis.text.x = element_text(angle = 60,hjust = 1,face="bold",size=11))
dev.off()

#comparing the P.Value between Y and O for the common, Y-specific and O-specific genes

SDVSS.combine.result%>%
  dplyr::select(id,class.sum,starts_with("P.Value"))%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  filter(class.sum!="na")%>%
  gather(age,p.val,P.Value.Y:P.Value.O)%>%
  mutate(age=ifelse(grepl("Y",age),"Y","O"),dir=ifelse(grepl("S",class.sum),"Sleep","Wake"),class=ifelse(grepl("Common",class.sum),"Common",ifelse(grepl("Y-spe",class.sum),"Y-specific","O-specific")))%>%
mutate(age=factor(age,levels=c("Y","O"),labels=c("Young","Old")),dir=factor(dir,levels=c("Sleep","Wake"),labels=c("Sleep Genes","Wake Genes")),class=factor(class,levels=c("Common","Y-specific","O-specific")))%>%
  ggviolin(.,x="age",y="p.val",fill = "age",add="boxplot",add.params = list(fill = "white"), facet.by=c("dir","class"),xlab="",ylab="-log10 (P.Value)")+
  stat_compare_means(label="p.format",paired = TRUE,label.x=1.3)+
theme(legend.position="none",strip.text = element_text(face="bold", size=12),axis.text.x = element_text(face="bold",size=11))

png(filename ="Pval.compare.png",height = 5,width =10,res = 300,units = "in",type="cairo")
p1<-SDVSS.combine.result%>%
  filter(class.sum%in%c("S.Common","S.Y-spe","S.O-spe"))%>%
  dplyr::select(id,class.sum,starts_with("P.Value"))%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  gather(age,p.val,P.Value.Y:P.Value.O)%>%
  mutate(age=ifelse(grepl("Y",age),"Y","O"),dir=ifelse(grepl("S",class.sum),"Sleep","Wake"),class=ifelse(grepl("Common",class.sum),"Common",ifelse(grepl("Y-spe",class.sum),"Y-specific","O-specific")))%>%
mutate(age=factor(age,levels=c("Y","O"),labels=c("Young","Old")),dir=factor(dir,levels=c("Sleep","Wake"),labels=c("Sleep Genes","Wake Genes")),class=factor(class,levels=c("Common","Y-specific","O-specific")))%>%
  ggviolin(.,x="age",y="p.val",fill = "class",palette = c("#00AFBB", "#E7B800", "#FC4E07"), add="boxplot",add.params = list(fill = "white"), xlab="",ylab="-log10 (P.Value)",title="Sleep Genes")+
  facet_wrap(~class,strip.position = "bottom")+
  stat_compare_means(label="p.signif",paired = TRUE,label.x=1.4,label.y=40)+
theme(legend.position="none",strip.text = element_text(face="bold",size=12,color="white"), axis.text.x = element_text(face="bold",size=11),strip.background = element_rect(color="white",fill =brewer.pal(9,"YlOrBr")[8]), plot.title=element_text(face="bold",size=14,hjust=0.5))+
  scale_y_continuous(expand=c(0,1))

p2<-SDVSS.combine.result%>%
  filter(class.sum%in%c("W.Common","W.Y-spe","W.O-spe"))%>%
  dplyr::select(id,class.sum,starts_with("P.Value"))%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  gather(age,p.val,P.Value.Y:P.Value.O)%>%
  mutate(age=ifelse(grepl("Y",age),"Y","O"),dir=ifelse(grepl("S",class.sum),"Sleep","Wake"),class=ifelse(grepl("Common",class.sum),"Common",ifelse(grepl("Y-spe",class.sum),"Y-specific","O-specific")))%>%
mutate(age=factor(age,levels=c("Y","O"),labels=c("Young","Old")),dir=factor(dir,levels=c("Sleep","Wake"),labels=c("Sleep Genes","Wake Genes")),class=factor(class,levels=c("Common","Y-specific","O-specific")))%>%
  ggviolin(.,x="age",y="p.val",fill = "class",palette = c("#00AFBB", "#E7B800", "#FC4E07"), add="boxplot",add.params = list(fill = "white"), xlab="",ylab="-log10 (P.Value)",title="Wake Genes")+
  facet_wrap(~class,strip.position = "bottom")+
  stat_compare_means(label="p.signif",paired = TRUE,label.x=1.4,label.y=40)+
theme(legend.position="none",strip.text = element_text(face="bold",size=12,color="white"), axis.text.x = element_text(face="bold",size=11),strip.background = element_rect(color="white",fill = brewer.pal(9,"PuBu")[8]), plot.title=element_text(face="bold",size=14,hjust=0.5))+
  scale_y_continuous(expand=c(0,1))

ggarrange(p1,p2,ncol=2)

  dev.off()
  
  
png(filename ="deltaPval.png",height = 5,width =5,res = 300,units = "in",type="cairo")
  
SDVSS.combine.result%>%
    filter(class.sum!="na")%>%
    dplyr::select(id,class,dir,starts_with("P.Value"))%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  mutate(deltaP=P.Value.O-P.Value.Y)%>%
ggscatter(.,x="P.Value.Y",y="deltaP",size=1,facet.by="dir",color="class",palette = c("#00AFBB", "#E7B800", "#FC4E07"), ylim=c(-12,12),xlab=expression('log'[10]*'P.Value of SD vs. SS in young mice'),ylab=expression('delta-log'[10]*'P.Value (Old-Young)'),title="Shift of P.values in old compared to the young")+
    geom_hline(yintercept = 0,linetype="dashed",color="black")+
    theme(plot.title=element_text(size=12,face="bold",hjust=0.5),legend.title=element_blank(),legend.position="bottom",strip.text=element_text(size=14,face="bold"))

  dev.off()

 #pie chart look at the shift in P.Value 
 
  blank_theme <- theme_minimal()+
  theme(
  axis.title.x = element_blank(),
  axis.title.y = element_blank(),
  panel.border = element_blank(),
  panel.grid=element_blank(),
  axis.ticks = element_blank(),
  axis.text.x=element_blank(),
  axis.text.y=element_blank(),
  plot.title=element_text(size=14, face="bold")
  )
  
png(filename ="Pie.deltaPval.png",height = 5,width =5,res = 300,units = "in",type="cairo")

SDVSS.combine.result%>%
  filter(class.sum!="na")%>%
  dplyr::select(id,class,dir,starts_with("P.Value"))%>%
  mutate_at(vars(starts_with("P.Value")),funs(-log10(.)))%>%
  mutate(deltaP=P.Value.O-P.Value.Y)%>%
  mutate(deltaP.dir=ifelse(deltaP>0,"O-up","O-down"))%>%
  group_by(dir,deltaP.dir)%>%
  summarise(n=n())%>%
ggplot(., aes(x="", y=n, fill=deltaP.dir))+geom_bar(width = 1, stat="identity",color="black")+ coord_polar("y", start=0)+ scale_fill_manual(values=c("green","red"))+
  facet_grid(dir ~ .)+
 blank_theme +
  geom_text(aes(label=n),size=8,color=c("white","black","white","black"),position = position_stack(vjust = 0.5))
    dev.off()


png(filename ="deltaFC.png",height = 5,width =5,res = 300,units = "in",type="cairo")
  
SDVSS.combine.result%>%
    filter(class.sum!="na")%>%
    dplyr::select(id,class,dir,starts_with("aveFC.SDvSS"))%>%
  mutate(deltaFC=abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y))%>%
ggscatter(.,x="aveFC.SDvSS.Y",y="deltaFC",size=1,color="class",palette = c("#00AFBB", "#E7B800", "#FC4E07"), xlab=expression('log'[2]*'Fold change between SD and SS in young mice'),ylab=expression('abs.delta [log'[2]*'Fold change (Old-Young)]'),title="Shift of fold change in old compared to the young")+
  facet_wrap(~dir,scale="free_x")+
    geom_hline(yintercept = 0,linetype="dashed",color="black")+
    theme(plot.title=element_text(size=12,face="bold",hjust=0.5),legend.title=element_blank(),legend.position="bottom",strip.text=element_text(size=14,face="bold"),panel.background = element_rect(color="black"))

  dev.off()

  
png(filename ="deltaFC.class.png",height = 4,width =6,res = 300,units = "in",type="cairo")

SDVSS.combine.result%>%
    filter(class.sum!="na")%>%
    dplyr::select(id,class,dir,starts_with("aveFC.SDvSS"))%>%
  mutate(deltaFC=abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y))%>%
ggviolin(.,x="class",y="deltaFC",color="class",size=1,palette = c("#00AFBB", "#E7B800", "#FC4E07"),add="jitter", add.params = list(size =0.05,color="grey") , xlab="",ylab=expression('delta abs.log'[2]*'FC (Old-Young)'))+
    facet_wrap(~dir)+
    geom_hline(yintercept = 0,linetype="dashed",color="black")+
    theme(plot.title=element_text(size=12,face="bold",hjust=0.5),legend.title=element_blank(),legend.position="bottom",strip.text=element_text(size=14,face="bold"),panel.background = element_rect(color="black"), axis.text.x = element_text(angle = 60,hjust = 1,face="bold",size=11), axis.text = element_text(face="bold",size=11))

  dev.off()

  
  
#pie chart look at the shift in ave.FC 
 
png(filename ="Pie.deltaFC.class.png",height = 4,width =6,res = 300,units = "in",type="cairo")

test<-SDVSS.combine.result%>%
  filter(class.sum!="na")%>%
    dplyr::select(id,class,dir,starts_with("aveFC.SDvSS"))%>%
  mutate(deltaFC=abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y))%>%
  mutate(deltaFC.dir=ifelse(deltaFC>0,"O-up","O-down"))%>%
  group_by(dir,class,deltaFC.dir)%>%
  summarise(n=n())%>%
  mutate(freq = n / sum(n))%>%
ggplot(., aes(x="", y=freq, fill=deltaFC.dir))+geom_bar(width = 1, stat="identity",color="black")+ coord_polar("y", start=0)+ scale_fill_manual(values=c("green","red"))+
  facet_grid(dir~class)+
 blank_theme +
  geom_text(aes(label=n),size=6,position = position_stack(vjust = 0.5))

dev.off()

#calculate CV among biological replicate in each condition
CV.DEGs<-cbind(as.data.frame(v.norm$genes),as.data.frame(v.norm$E))%>%
        filter(.$id%in%SDVSS.combine.result[SDVSS.combine.result$class.sum!="na",]$id)%>%
        dplyr::select(-starts_with("gene"),-ENTREZID,-description,-transcript_length)%>%
        gather(key=sample,value=logCPM,-id)%>%
        mutate(CPM=2^logCPM)%>%
  mutate(age=ifelse(grepl("Young",sample),"Y","O"),sleep=ifelse(grepl("SS",sample),"SS","SD"),time=ifelse(grepl("Z0",sample),"Z0",ifelse(grepl("3hr",sample),"3",ifelse(grepl("6hr",sample),"6",ifelse(grepl("9hr",sample),"9","12")))))%>%
  filter(time!="Z0")%>%
  unite(con,age:time,sep=".")%>%  
  unite(con.id,c("id","con"),sep="_")%>%
  group_by(con.id)%>%
  summarise(meanCPM=mean(CPM),sdCPM=sd(CPM))%>%
  mutate(CV=sdCPM/meanCPM)%>%
  mutate(logCV=log(CV))

png(filename ="CV.corr.png",height = 7,width =4.5,res = 300,units = "in",type="cairo")
CV.DEGs%>%
  separate(con.id,c("id","con"),sep="_")%>%
    dplyr::select(id,con,logCV)%>%
    separate(con,c("age","sleep","time"))%>%
    spread(age,logCV)%>%
    mutate(sleep=factor(sleep,levels=c("SS","SD")),time=factor(time,levels=c("3","6","9","12"),labels=c("3h","6h","9h","12h")))%>%
ggscatter(.,x="Y",y="O",size=0.5,facet.by=c("time","sleep"),ellipse = F,ellipse.alpha=0.8,xlab="logCV of young mice", ylab="logCV of old mice",title="Correlation plots of CV between young and old mice")+geom_abline(slope=1,linetype="dashed",size=1)+stat_cor(method = "pearson", label.x = -4.5, label.y = 0)+theme(plot.title=element_text(face="bold",size=11,hjust=0.5),axis.text = element_text(face="bold",size=12))
  
 dev.off()


 png(filename ="deltaCV.png",height = 5,width =6,res = 300,units = "in",type="cairo")

 CV.DEGs%>%
  separate(con.id,c("id","con"),sep="_")%>%
    dplyr::select(id,con,logCV)%>%
    separate(con,c("age","sleep","time"))%>%
    spread(age,logCV)%>%
    mutate(deltaCV=O-Y,sleep=factor(sleep,levels=c("SS","SD")),time=factor(time,levels=c("3","6","9","12"),labels=c("3h","6h","9h","12h")))%>%
ggboxplot(.,x="time",y="deltaCV",size=0.5,facet.by="sleep",xlab="Hours of SS or SD", ylab="deltalogCV (old-young)",title="Shift in CV in old mice compared to the young mice")+geom_hline(yintercept = 0,linetype="longdash",size=1,color="blue")+theme(plot.title=element_text(face="bold",size=11,hjust=0.5),axis.text = element_text(face="bold",size=12))
   dev.off()

 
lm(O.SS.3~Y.SS.3,data=test)
lm(Y.SS.3~O.SS.3,data=test)
spearman(x=O.SS.3,y=Y.SS.3,data=test)

#cluster profiler
sleep.common<-SDVSS.combine.result%>%
  filter(dir=="SleepGenes"&class=="Common")%>%
  .$ENTREZID

sleep.Yonly<-SDVSS.combine.result%>%
  filter(dir=="SleepGenes"&class=="Y-specific")%>%
  .$ENTREZID
sleep.Oonly<-SDVSS.combine.result%>%
  filter(dir=="SleepGenes"&class=="O-specific")%>%
  .$ENTREZID
  
Sleep.ClusterPro<-list(sleep.common,sleep.Yonly,sleep.Oonly)
names(Sleep.ClusterPro)<-c("Common","Y-specific","O-specific")

source("https://bioconductor.org/biocLite.R")
biocLite("clusterProfiler")
install.packages("data.table", dependencies=TRUE)
require(clusterProfiler)
library(org.Mm.eg.db)
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("AnnotationDbi", version = "3.8")

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("org.Mm.eg.db", version = "3.8")

sleep.bp <- compareCluster(geneCluster = Sleep.ClusterPro, universe = as.character(v.norm$genes$ENTREZID),
fun = "enrichGO",OrgDb= org.Mm.eg.db, ont= "BP",pAdjustMethod = "none",pvalueCutoff  = 0.001,qvalueCutoff =1,minGSSize = 10)

sleep.bp.sim <- clusterProfiler::simplify(sleep.bp, cutoff=0.6, by="pvalue", select_fun=min)

dotplot(sleep.bp,showCategory=10,title="sleep.BP")





#interaction with sleep and wake genes combined from y and o
combine.slwake.id<-x_norm$genes$id%in%SDVSS.combine.result[SDVSS.combine.result$class.sum!="na",]$id

v.int.slwake<-v.norm[combine.slwake.id,]

contr.int<-makeContrasts(
  SDvsSS_3hr.YvO=(Y.SD.3hr-Y.SS.3hr)-(O.SD.3hr-O.SS.3hr),
  SDvsSS_6hr.YvO=(Y.SD.6hr-Y.SS.6hr)-(O.SD.6hr-O.SS.6hr),
  SDvsSS_9hr.YvO=(Y.SD.9hr-Y.SS.9hr)-(O.SD.9hr-O.SS.9hr),
  SDvsSS_12hr.YvO=(Y.SD.12hr-Y.SS.12hr)-(O.SD.12hr-O.SS.12hr),
  ave.int=((Y.SD.3hr-Y.SS.3hr)+(Y.SD.6hr-Y.SS.6hr)+(Y.SD.9hr-Y.SS.9hr)+(Y.SD.12hr-Y.SS.12hr))-((O.SD.3hr-O.SS.3hr)+(O.SD.6hr-O.SS.6hr)+(O.SD.9hr-O.SS.9hr)+(O.SD.12hr-O.SS.12hr)),
  levels=design)

vfit.int.slwake<- eBayes(contrasts.fit(lmFit(v.int.slwake, design),contrasts=contr.int))

SDvSS.combine.int<-topTable(vfit.int.slwake,coef=c(1:4),n=Inf)%>%
  dplyr::select(id,adj.P.Val)%>%
  rename(FDR.int=adj.P.Val)%>%
  left_join(SDVSS.combine.result,.,by="id")#21 FDR<0.05 from combined sleep and wake genes 

write.table(SDvSS.combine.int,"SDvSS.combine.int.txt",sep="\t",row.names = F)


int.ave.DEG<-topTable(vfit.int.slwake,coef=5,n=Inf,p.value=0.05)%>%
  dplyr::select(id,geneSymbol,adj.P.Val)%>%
  left_join(.,SDVSS.combine.result[,c(1,23:27)],by="id")#21 FDR<0.05 from combined sleep and wake genes 





#replace ensembl ID from the DAVID result to gene symbol
DAVID.comb.EnID<-read.delim("DAVID.comb.EnID.txt",sep="\t",header=T,stringsAsFactors = F)

DAVID.comb.genename<-DAVID.comb.EnID%>%
  unite(search.term,search,Term,remove=F)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,5)],by="id")%>%
  dplyr::select(-Genes)%>%
  group_by(search.term)%>%
  mutate(geneSymbol=paste(geneSymbol,collapse=","))%>%
  distinct(Term,.keep_all=T)

write.table(DAVID.comb.genename,"DAVID.comb.genename.txt",sep="\t",row.names = FALSE)

#calculate percentage change in fold change comparing old to young for each sleep and wake pathway
sleep.Yspec<-DAVID.comb.EnID%>%
  filter(search=="sleep.Yspec"&PValue<0.01)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,21,22)],by="id")%>%
  mutate(GO=ifelse(grepl("GO",Term),"GO","KEGG"))%>%
  filter(GO=="GO")%>%
  mutate(Term.factor=as.factor(Term))%>%
  filter(as.numeric(Term.factor)%in%c(2,3,4,6,7,13))%>%
    dplyr::select(Term,PValue,id,starts_with("aveFC"))%>%
mutate(class=rep("Yspec"))%>%
  mutate(GO=ifelse(grepl("adhesion",Term), "Cell adhesion", ifelse(grepl("GO:0007268",Term),"Chemial synaptic transmission", ifelse(grepl("GO:0001764",Term),"Neuron migration", ifelse(grepl("GO:0007411",Term),"Axon guidance","Potassium ion transport")))))%>%
  group_by(GO)%>%
  distinct(id,.keep_all=T)%>%
  mutate(FC.ratio=(2^(abs(aveFC.SSvZ0.O)-abs(aveFC.SSvZ0.Y))-1)*100)

sleep.common<-DAVID.comb.EnID%>%
  filter(search=="sleep.common"&PValue<0.02)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,21,22)],by="id")%>%
  mutate(GO=ifelse(grepl("GO",Term),"GO","KEGG"))%>%
  filter(GO=="GO")%>%
  mutate(Term.factor=as.factor(Term))%>%
  filter(as.numeric(Term.factor)%in%c(4,2,8))%>%
  dplyr::select(Term,PValue,id,starts_with("aveFC"))%>%
mutate(class=rep("common"))%>%
  mutate(GO=ifelse(grepl("DNA",Term),"DNArepair","proteinprocessing"))%>%
  group_by(GO)%>%
  distinct(id,.keep_all=T)%>%
  mutate(FC.ratio=(2^(abs(aveFC.SSvZ0.O)-abs(aveFC.SSvZ0.Y))-1)*100)

sleep.Ospec<-DAVID.comb.EnID%>%
  filter(search=="sleep.Ospec"&PValue<0.003)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,21,22)],by="id")%>%
  mutate(GO=ifelse(grepl("GO",Term),"GO","KEGG"))%>%
  filter(GO=="GO")%>%
  mutate(Term.factor=as.factor(Term))%>%
  filter(as.numeric(Term.factor)%in%c(1,2))%>%
  dplyr::select(Term,PValue,id,starts_with("aveFC"))%>%
mutate(class=rep("Ospec"),GO="cholesterol synthesis")%>%
  group_by(GO)%>%
  distinct(id,.keep_all=T)%>%
  mutate(FC.ratio=(2^(abs(aveFC.SSvZ0.O)-abs(aveFC.SSvZ0.Y))-1)*100)

png(filename = "Sleep.GO.FCshift.png",height = 4,width =5,res = 300,units = "in",type="cairo")

do.call("rbind", list(sleep.common, sleep.Ospec, sleep.Yspec))%>%
  mutate(class=factor(class,levels=c("common","Ospec","Yspec"),labels=c("Common","O-specific","Y-specific")))%>%
ggbarplot(.,x="GO",y="FC.ratio",size=1,color="class",palette = c("#00AFBB","#FC4E07","#E7B800"),add=c("jitter","mean_se"),add.params=list(size=0.5),xlab="",ylab="Relative FC shift in old (%)",ylim=c(-40,25))+
  geom_hline(yintercept = 0,linetype="dashed",color="black")+
  theme(axis.text.x=element_text(angle = 30,hjust = 1,size=12), legend.position = c(0.85,0.88),legend.title=element_blank())

dev.off()



wake.common<-DAVID.comb.EnID%>%
  filter(search=="wake.common"&PValue<0.02)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,19,20)],by="id")%>%
  mutate(Term.factor=as.factor(Term))%>%
  filter(as.numeric(Term.factor)%in%c(24,28,13,31,33))%>%
  dplyr::select(Term,PValue,id,starts_with("aveFC"))%>%
  mutate(GO=ifelse(grepl("MAPK",Term), "MAPK signaling", ifelse(grepl("mmu04141",Term),"Protein processing in ER", ifelse(grepl("GO:0051591",Term),"Response to cAMP", ifelse(grepl("GO:0045944",Term),"Pos.reg.transcription","Phosphorylation")))))%>%
  mutate(class=rep("common"))%>%
  mutate(FC.ratio=(2^(abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y))-1)*100)


wake.Ospec<-DAVID.comb.EnID%>%
  filter(search=="wake.Ospec"&PValue<0.01)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,19,20)],by="id")%>%
  mutate(Term.factor=as.factor(Term))%>%
  dplyr::select(Term,PValue,id,starts_with("aveFC"))%>%
 mutate(class=rep("Ospec"),GO=rep("Pos.reg cell proliferation"))%>%
  mutate(FC.ratio=(2^(abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y))-1)*100)


wake.Yspec<-DAVID.comb.EnID%>%
  filter(search=="wake.Yspec"&PValue<0.002)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,19,20)],by="id")%>%
  mutate(Term.factor=as.factor(Term))%>%
  dplyr::select(Term,PValue,id,starts_with("aveFC"))%>%
 mutate(class=rep("Yspec"),GO=rep("Actin cytoskeleton organization"))%>%
  mutate(FC.ratio=(2^(abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y))-1)*100)


png(filename = "Wake.GO.FCshift.png",height = 4,width =5,res = 300,units = "in",type="cairo")

do.call("rbind", list(wake.common, wake.Ospec, wake.Yspec))%>%
  mutate(class=factor(class,levels=c("common","Ospec","Yspec"),labels=c("Common","O-specific","Y-specific")))%>%
ggbarplot(.,x="GO",y="FC.ratio",size=1,color="class",palette = c("#00AFBB","#FC4E07","#E7B800"),add=c("jitter","mean_se"),add.params=list(size=0.5),xlab="",ylab="Relative FC shift in old (%)",ylim=c(-40,30))+
  geom_hline(yintercept = 0,linetype="dashed",color="black")+
  theme(axis.text.x=element_text(angle = 35,hjust = 1,size=11), legend.position = c(0.92,0.87),legend.title=element_blank(),legend.key.size = unit(1,"line"))

dev.off()  
  
  
wake.GO.FC<-do.call("rbind", list(wake.common, wake.Ospec, wake.Yspec))%>%
  left_join(.,SDvsSS.FC,by="id")%>%
  dplyr::select(-Term)%>%
  mutate(aveY=rowMeans(.[,6:9]),aveO=rowMeans(.[,10:13]))%>%
  filter(class!="common"|aveY>0.5|aveO>0.5)%>%
  filter(aveY>0.3|aveO>0.3)


#choose top unique GO terms and take their unique genes for heatmap
sleep.Yspec<-DAVID.comb.EnID%>%
  filter(search=="sleep.Yspec"&PValue<0.01)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,19,20)],by="id")%>%
  mutate(GO=ifelse(grepl("GO",Term),"GO","KEGG"))%>%
  filter(GO=="GO")%>%
  mutate(Term.factor=as.factor(Term))%>%
  filter(as.numeric(Term.factor)%in%c(2,3,4,6,7,13))%>%
    dplyr::select(Term,PValue,id,starts_with("aveFC"))%>%
mutate(class=rep("Yspec"))%>%
  mutate(GO=ifelse(grepl("adhesion",Term), "Cell adhesion", ifelse(grepl("GO:0007268",Term),"Chemial synaptic transmission", ifelse(grepl("GO:0001764",Term),"Neuron migration", ifelse(grepl("GO:0007411",Term),"Axon guidance","Potassium ion transport")))))%>%
  group_by(GO)%>%
  distinct(id,.keep_all=T)%>%
  mutate(FC.ratio=2^(abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y))-1)

ggboxplot(sleep.Yspec,x="GO",y="FC.ratio",ylim=c(0,150),add="jitter")


  test<-sleep.Yspec%>%
  dplyr::select(Term,PValue,id)%>%
  mutate(GO=ifelse(grepl("GO",Term),"GO","KEGG"))%>%
  filter(GO=="GO")%>%
  unite(Term.p,Term,PValue,sep=".")%>%
  mutate(yes=rep(1))%>%
  dplyr::select(-GO)%>%
  spread(Term.p,yes)%>%
replace(., is.na(.), 0)

pheatmap(as.matrix(test), fontsize = 6, fontsize_row=10,fontsize_col=10)

sleep.common<-DAVID.comb.EnID%>%
  filter(search=="sleep.common"&PValue<0.02)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,19,20)],by="id")%>%
  mutate(GO=ifelse(grepl("GO",Term),"GO","KEGG"))%>%
  filter(GO=="GO")%>%
  mutate(Term.factor=as.factor(Term))%>%
  filter(as.numeric(Term.factor)%in%c(4,2,8))%>%
  dplyr::select(Term,PValue,id,starts_with("aveFC"))%>%
mutate(class=rep("common"))%>%
  mutate(GO=ifelse(grepl("DNA",Term),"DNArepair","proteinprocessing"))%>%
  group_by(GO)%>%
  distinct(id,.keep_all=T)%>%
  mutate(FC.ratio=2^(abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y))-1)

  
  mutate(Term.factor=as.factor(Term))%>%
  filter(as.numeric(Term.factor)%in%c(4,2,8))%>%
  distinct(id,.keep_all=T)%>%
  mutate(class=rep("common"))

sleep.Ospec<-DAVID.comb.EnID%>%
  filter(search=="sleep.Ospec"&PValue<0.003)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,5)],by="id")%>%
  dplyr::select(Term,PValue,id)%>%
  mutate(GO=ifelse(grepl("GO",Term),"GO","KEGG"))%>%
  filter(GO=="GO")%>%
  mutate(Term.factor=as.factor(Term))%>%
  distinct(id,.keep_all=T)%>%
  mutate(class=rep("Ospec"))

sleep.Ospec<-DAVID.comb.EnID%>%
  filter(search=="sleep.Ospec"&PValue<0.003)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,19,20)],by="id")%>%
  mutate(GO=ifelse(grepl("GO",Term),"GO","KEGG"))%>%
  filter(GO=="GO")%>%
  mutate(Term.factor=as.factor(Term))%>%
  filter(as.numeric(Term.factor)%in%c(1,2))%>%
  dplyr::select(Term,PValue,id,starts_with("aveFC"))%>%
mutate(class=rep("Ospec"),GO="cholesterol synthesis")%>%
  group_by(GO)%>%
  distinct(id,.keep_all=T)%>%
  mutate(FC.ratio=2^(abs(aveFC.SDvSS.O)-abs(aveFC.SDvSS.Y))-1)


do.call("rbind", list(sleep.common, sleep.Ospec, sleep.Yspec))%>%
ggboxplot(.,x="GO",y="FC.ratio",add=c("jitter","mean"),add.params=list(size=0.5))


do.call("rbind", list(sleep.common, sleep.Ospec, sleep.Yspec))%>%
ggbarplot(.,x="GO",y="FC.ratio",ylim=c(-0.5,0.5),add=c("jitter","mean_se"),add.params=list(size=0.5))

SSvsZT0.FC<-cbind(as.data.frame(efit.main$genes),as.data.frame(efit.main$coefficients))%>%dplyr::select(id,geneSymbol,starts_with("SSvZ0"))

Sleep.GO.FC<-do.call("rbind", list(sleep.common, sleep.Ospec, sleep.Yspec))%>%
  left_join(.,SSvsZT0.FC,by="id")%>%
  dplyr::select(-GO,-Term)%>%
  mutate(aveY=rowMeans(.[,6:9]),aveO=rowMeans(.[,10:13]))%>%
  filter(class=="Ospec"|aveY>0.2|aveO>0.2)
  
h.common1<-Sleep.GO.FC%>%
filter(class=="common"&grepl("GO:0006974",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SSvZ0"))%>%
column_to_rownames(.,var="geneSymbol")

h.common2<-Sleep.GO.FC%>%
filter(class=="common"&grepl("GO:0016485",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SSvZ0"))%>%
column_to_rownames(.,var="geneSymbol")

h.Ospec<-Sleep.GO.FC%>%
filter(class=="Ospec")%>%
  dplyr::select(geneSymbol,starts_with("SSvZ0"))%>%
column_to_rownames(.,var="geneSymbol")

h.Yspec1<-Sleep.GO.FC%>%
filter(class=="Yspec"&grepl("GO:0007156",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SSvZ0"))%>%
column_to_rownames(.,var="geneSymbol")

h.Yspec2<-Sleep.GO.FC%>%
filter(class=="Yspec"&grepl("GO:0007155",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SSvZ0"))%>%
column_to_rownames(.,var="geneSymbol")

h.Yspec3<-Sleep.GO.FC%>%
filter(class=="Yspec"&grepl("GO:0007268",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SSvZ0"))%>%
column_to_rownames(.,var="geneSymbol")

h.Yspec4<-Sleep.GO.FC%>%
filter(class=="Yspec"&grepl("GO:0001764",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SSvZ0"))%>%
column_to_rownames(.,var="geneSymbol")

h.Yspec5<-Sleep.GO.FC%>%
filter(class=="Yspec"&grepl("GO:0007411",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SSvZ0"))%>%
column_to_rownames(.,var="geneSymbol")

h.Yspec6<-Sleep.GO.FC%>%
filter(class=="Yspec"&grepl("GO:0071805",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SSvZ0"))%>%
column_to_rownames(.,var="geneSymbol")

fit1<-hclust(dist(as.matrix(h.common1)))
fit2<-hclust(dist(as.matrix(h.common2)))
fit3<-hclust(dist(as.matrix(h.Ospec)))
fit4<-hclust(dist(as.matrix(h.Yspec1)))
fit5<-hclust(dist(as.matrix(h.Yspec2)))
fit6<-hclust(dist(as.matrix(h.Yspec3)))
fit7<-hclust(dist(as.matrix(h.Yspec4)))
fit8<-hclust(dist(as.matrix(h.Yspec5)))
fit9<-hclust(dist(as.matrix(h.Yspec6)))


hc <- as.hclust(merge(merge(merge(merge(merge(merge(merge(as.dendrogram(fit1), as.dendrogram(fit2)), as.dendrogram(fit3)),as.dendrogram(fit5)),as.dendrogram(fit6)),as.dendrogram(fit7)),as.dendrogram(fit8)),as.dendrogram(fit9)))


table<-rbind(h.common1,h.common2,h.Ospec,h.Yspec2,h.Yspec3,h.Yspec4,h.Yspec5,h.Yspec6)

rowannotation <- data.frame(class = c(rep("Common", 36), 
                                rep("O.specific", nrow(h.Ospec)),
                                rep("Y.specific", 95)),
                                row.names = row.names(table))

colors <- c(seq(-1.5,-0.81,length=10),seq(-0.8,-0.11,length=40),seq(-0.1,0.1,length=10),seq(0.11,0.8,length=40),seq(0.81,1.5,length=10))
my_palette<-colorRampPalette(rev(brewer.pal(11,"RdGy")))(110)



ann.color = list(class=c(Common="#00AFBB",O.specific= "#FC4E07",Y.specific = "#E7B800"))


png(filename = "Sleep.selectGO.png",height = 15,width =5,res = 300,units = "in",type="cairo")

pheatmap(as.matrix(table),cluster_rows = hc,cluster_col=FALSE,annotation_row = rowannotation, annotation_colors = ann.color, breaks = colors,color=my_palette, border_color= NA,labels_col = c("3","6","9","12","3","6","9","12"), gaps_col=4,fontsize = 10,treeheight_row=0,legend_breaks=c(-1,0,1),legend_labels =c("-1","0","1"))

dev.off()



wake.common<-DAVID.comb.EnID%>%
  filter(search=="wake.common"&PValue<0.02)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,5)],by="id")%>%
  dplyr::select(Term,PValue,id)%>%
  mutate(Term.factor=as.factor(Term))%>%
  filter(as.numeric(Term.factor)%in%c(24,28,13,31,33))%>%
  distinct(id,.keep_all=T)%>%
  mutate(class=rep("common"))

wake.Ospec<-DAVID.comb.EnID%>%
  filter(search=="wake.Ospec"&PValue<0.01)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,5)],by="id")%>%
  dplyr::select(Term,PValue,id)%>%
  mutate(Term.factor=as.factor(Term))%>%
  distinct(id,.keep_all=T)%>%
  mutate(class=rep("Ospec"))

wake.Yspec<-DAVID.comb.EnID%>%
  filter(search=="wake.Yspec"&PValue<0.002)%>%
  transform(Genes=strsplit(Genes, ","))%>%
  unnest(Genes)%>%
  mutate(id=gsub(" ", "", Genes))%>%
  left_join(.,SDVSS.combine.result[,c(1,5)],by="id")%>%
  dplyr::select(Term,PValue,id)%>%
  mutate(Term.factor=as.factor(Term))%>%
  distinct(id,.keep_all=T)%>%
  mutate(class=rep("Yspec"))


SDvsSS.FC<-cbind(as.data.frame(efit.main$genes),as.data.frame(efit.main$coefficients))%>%dplyr::select(id,geneSymbol,starts_with("SDvsSS"))

wake.GO.FC<-do.call("rbind", list(wake.common, wake.Ospec, wake.Yspec))%>%
  left_join(.,SDvsSS.FC,by="id")%>%
  dplyr::select(-Term)%>%
  mutate(aveY=rowMeans(.[,6:9]),aveO=rowMeans(.[,10:13]))%>%
  filter(class!="common"|aveY>0.5|aveO>0.5)%>%
  filter(aveY>0.3|aveO>0.3)

write.table(wake.GO.FC,"wake.GO.FC.txt",sep="\t",row.names = F)
  
  
h.common1<-wake.GO.FC%>%
filter(class=="common"&grepl("mmu04010",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SDvsSS"))%>%
column_to_rownames(.,var="geneSymbol")

h.common2<-wake.GO.FC%>%
filter(class=="common"&grepl("mmu04141",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SDvsSS"))%>%
column_to_rownames(.,var="geneSymbol")

h.common3<-wake.GO.FC%>%
filter(class=="common"&grepl("GO:0051591",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SDvsSS"))%>%
column_to_rownames(.,var="geneSymbol")

h.common4<-wake.GO.FC%>%
filter(class=="common"&grepl("GO:0045944",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SDvsSS"))%>%
column_to_rownames(.,var="geneSymbol")

h.common5<-wake.GO.FC%>%
filter(class=="common"&grepl("GO:0016310",Term.factor))%>%
  dplyr::select(geneSymbol,starts_with("SDvsSS"))%>%
column_to_rownames(.,var="geneSymbol")

h.Ospec<-wake.GO.FC%>%
filter(class=="Ospec")%>%
  dplyr::select(geneSymbol,starts_with("SDvsSS"))%>%
column_to_rownames(.,var="geneSymbol")

h.Yspec<-wake.GO.FC%>%
filter(class=="Yspec")%>%
  dplyr::select(geneSymbol,starts_with("SDvsSS"))%>%
column_to_rownames(.,var="geneSymbol")


fit1<-hclust(dist(as.matrix(h.common1)))
fit2<-hclust(dist(as.matrix(h.common2)))
fit3<-hclust(dist(as.matrix(h.common3)))
fit4<-hclust(dist(as.matrix(h.common4)))
fit5<-hclust(dist(as.matrix(h.common5)))
fit6<-hclust(dist(as.matrix(h.Ospec)))
fit7<-hclust(dist(as.matrix(h.Yspec)))



hc <- as.hclust(merge(merge(merge(merge(merge(merge(as.dendrogram(fit1), as.dendrogram(fit2)), as.dendrogram(fit3)),as.dendrogram(fit4)),as.dendrogram(fit5)),as.dendrogram(fit6)),as.dendrogram(fit7)))


table<-rbind(h.common1,h.common2,h.common3,h.common4,h.common5,h.Ospec,h.Yspec)

rowannotation <- data.frame(class = c(rep("Common", 89), 
                                rep("O.specific", 6),
                                rep("Y.specific", 7)),
                                row.names = row.names(table))

colors <- c(seq(-5,-2.1,length=10),seq(-2,-0.11,length=40),seq(-0.1,0.1,length=10),seq(0.11,2,length=40),seq(2.1,5,length=10))
my_palette<-colorRampPalette(brewer.pal(11,"PRGn"))(110)



ann.color = list(class=c(Common="#00AFBB",O.specific= "#FC4E07",Y.specific = "#E7B800"))


png(filename = "wake.selectGO.png",height = 15,width =5,res = 300,units = "in",type="cairo")

pheatmap(as.matrix(table),cluster_rows = hc,cluster_col=FALSE,annotation_row = rowannotation, annotation_colors = ann.color, breaks = colors,color=my_palette, border_color= NA,labels_col = c("3","6","9","12","3","6","9","12"), gaps_col=4,fontsize = 10,treeheight_row=0,legend_breaks=c(-2,0,2),legend_labels =c("-2","0","2"))

dev.off()

```




```{r}
#cell type specific marker; gene set enrichment

SC.marker<-read.delim("singlecell_marker.txt",header = T,stringsAsFactors = F,row.names = NULL,na.strings = "NA")

SC.marker.long<-SC.marker%>%
  gather(.,key=geneset,value=geneSymbol)%>%
  left_join(.,v.norm$genes,by="geneSymbol")%>%
  filter(!is.na(ENTREZID))%>%
  group_by(geneset)%>%
  summarise(marker=list(unique(id)))
  
SC.marker.id<-SC.marker.long$marker
names(SC.marker.id)<-SC.marker.long$geneset
  
SC.marker.ind <-ids2indices(SC.marker.id, v.norm$gene$id)

contrasts.main<-makeContrasts(
  OvY.SS3h=O.SS.3hr-Y.SS.3hr,
  OvY.SS6h=O.SS.6hr-Y.SS.6hr,
  OvY.SS9h=O.SS.9hr-Y.SS.9hr,
  OvY.SS12h=O.SS.12hr-Y.SS.12hr,
  OvY.SD3h=O.SD.3hr-Y.SD.3hr,
  OvY.SD6h=O.SD.6hr-Y.SD.6hr,
  OvY.SD9h=O.SD.9hr-Y.SD.9hr,
  OvY.SD12h=O.SD.12hr-Y.SD.12hr,
  OvY.Z0=O.Z0.Z0-Y.Z0.Z0,
  SDvsSS_3hr.Y = Y.SD.3hr-Y.SS.3hr, 
   SDvsSS_6hr.Y = Y.SD.6hr-Y.SS.6hr,
   SDvsSS_9hr.Y = Y.SD.9hr-Y.SS.9hr,
   SDvsSS_12hr.Y = Y.SD.12hr-Y.SS.12hr,
  SDvsSS_3hr.O = O.SD.3hr-O.SS.3hr, 
   SDvsSS_6hr.O = O.SD.6hr-O.SS.6hr,
   SDvsSS_9hr.O = O.SD.9hr-O.SS.9hr,
   SDvsSS_12hr.O = O.SD.12hr-O.SS.12hr,
  aveYSS.vZ0=(Y.SS.3hr+Y.SS.6hr+Y.SS.9hr+Y.SS.12hr)/4-Y.Z0.Z0,
  aveOSS.vZ0=(O.SS.3hr+O.SS.6hr+O.SS.9hr+O.SS.12hr)/4-O.Z0.Z0,
  SSvZ0.Y3h=Y.SS.3hr-Y.Z0.Z0,
  SSvZ0.Y6h=Y.SS.6hr-Y.Z0.Z0,
  SSvZ0.Y9h=Y.SS.9hr-Y.Z0.Z0,
  SSvZ0.Y12h=Y.SS.12hr-Y.Z0.Z0,
  SSvZ0.O3h=O.SS.3hr-O.Z0.Z0,
  SSvZ0.O6h=O.SS.6hr-O.Z0.Z0,
  SSvZ0.O9h=O.SS.9hr-O.Z0.Z0,
  SSvZ0.O12h=O.SS.12hr-O.Z0.Z0,
  levels=design)

#camera results with single-cell marker (SC.marker.ind); name starts with c.sc. (stands for camera.single-cell)

c.sc.SDvSS.Y.3h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,10],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.Y.6h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,11],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.Y.9h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,12],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.Y.12h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,13],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.O.3h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,14],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.O.6h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,15],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.O.9h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,16],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.SDvSS.O.12h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,17],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")


c.sc.SDvSS.comb<-c.sc.SDvSS.Y.3h%>%
  left_join(.,c.sc.SDvSS.Y.6h,by="sc.geneset",suffix=c(".Y3h",".Y6h"))%>%
  left_join(.,c.sc.SDvSS.Y.9h,by="sc.geneset")%>%
  left_join(.,c.sc.SDvSS.Y.12h,by="sc.geneset",suffix=c(".Y9h",".Y12h"))%>%
  left_join(.,c.sc.SDvSS.O.3h,by="sc.geneset")%>%
  left_join(.,c.sc.SDvSS.O.6h,by="sc.geneset",suffix=c(".O3h",".O6h"))%>%
  left_join(.,c.sc.SDvSS.O.9h,by="sc.geneset")%>%
  left_join(.,c.sc.SDvSS.O.12h,by="sc.geneset",suffix=c(".O9h",".O12h"))%>%
  mutate(Direction.Y3h=ifelse(Direction.Y3h=="Up",1,-1),Direction.Y6h=ifelse(Direction.Y6h=="Up",1,-1),Direction.Y9h=ifelse(Direction.Y9h=="Up",1,-1),Direction.Y12h=ifelse(Direction.Y12h=="Up",1,-1),Direction.O3h=ifelse(Direction.O3h=="Up",1,-1),Direction.O6h=ifelse(Direction.O6h=="Up",1,-1),Direction.O9h=ifelse(Direction.O9h=="Up",1,-1),Direction.O12h=ifelse(Direction.O12h=="Up",1,-1))%>%
  mutate(dir.P.Y3h=-log10(FDR.Y3h)*Direction.Y3h,dir.P.Y6h=-log10(FDR.Y6h)*Direction.Y6h,dir.P.Y9h=-log10(FDR.Y9h)*Direction.Y9h,dir.P.Y12h=-log10(FDR.Y12h)*Direction.Y12h,dir.P.O3h=-log10(FDR.O3h)*Direction.O3h,dir.P.O6h=-log10(FDR.O6h)*Direction.O6h,dir.P.O9h=-log10(FDR.O9h)*Direction.O9h,dir.P.O12h=-log10(FDR.O12h)*Direction.O12h)%>%
  dplyr::select(sc.geneset,NGenes.Y3h,starts_with("dir.P."))%>%
  mutate(sc.geneset=ifelse(grepl("Tasic2016_Endothelial",sc.geneset),"Tasic2016_Endothelial",ifelse(grepl("Hrvatin2018_Endothelial",sc.geneset),"Hrvatin2018_Endothelial",sc.geneset)))%>% #shorten the names
  mutate(sc.geneset.n=paste0(sc.geneset,"(",NGenes.Y3h,")"))%>% #add gene number
  dplyr::select(-sc.geneset,-NGenes.Y3h)%>% 
  mutate(celltype=ifelse(grepl("OPC",sc.geneset.n),"OPC",ifelse(grepl("OLs",sc.geneset.n),"OL",ifelse(grepl("Microglia",sc.geneset.n),"Microglia",ifelse(grepl("neuron",sc.geneset.n),"neuron",ifelse(grepl("Astrocytes",sc.geneset.n),"Astrocyte",ifelse(grepl("Endothelial",sc.geneset.n),"Endo",ifelse(grepl("Inhi",sc.geneset.n),"Inhibitory","Excitatory"))))))))%>%
  mutate(celltype=factor(celltype,levels=c("Microglia","OL","OPC","Astrocyte","Endo","neuron","Inhibitory","Excitatory")))%>%
  arrange(celltype)

annotation.col=data.frame(celltype=c.sc.SDvSS.comb$celltype)
rownames(annotation.col)=c.sc.SDvSS.comb$sc.geneset.n

annotation.row=data.frame(comparison=rep(c("SDvSS.Y","SDvSS.O"),c(4,4)))
annotation.row$comparison=factor(annotation.row$comparison,levels=c("SDvSS.Y","SDvSS.O"))
rownames(annotation.row)=colnames(c.sc.SDvSS.comb)[1:8]

library("pheatmap")
my_palette.camera<-colorRampPalette(rev(brewer.pal(11,"RdBu")))(219)

colors.camera <- c(seq(-7,-5.1,length=20),seq(-5,-1.3,length=80),seq(-1.29,1.29,length=20),seq(1.3,5,length=80),seq(5.1,7,length=20))


png(filename = "scmarker.SDvSS.png",height = 3,width =5.5,res = 300,units = "in",type="cairo")
c.sc.SDvSS.comb%>%
  dplyr::select(-celltype)%>%
  column_to_rownames(.,var="sc.geneset.n")%>%
  t%>%
  pheatmap(., breaks = colors.camera, color=my_palette.camera, cluster_cols=FALSE,cluster_rows=FALSE,border_color= NA,annotation_row =annotation.row, annotation_col=annotation.col, gaps_row=4, gaps_col=12,labels_row=c("ZT3","ZT6","ZT9","ZT12","ZT3","ZT6","ZT9","ZT12"),fontsize = 6.5,legend_breaks=c(-5,-1.3,1.3,5),legend_labels =c("e-5","0.05","0.05","e-5"), main="Enrichment of cell-type specific gene sets in sleep and wake regulation")

dev.off()


c.sc.OvY.SS.3h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,1],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SS.6h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,2],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SS.9h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,3],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SS.12h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,4],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SD.3h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,5],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SD.6h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,6],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SD.9h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,7],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")
c.sc.OvY.SD.12h<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,8],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")

c.sc.OvY.ZT0<-camera(v.norm,index=SC.marker.ind,design=design,contrast=contrasts.main[,9],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="sc.geneset")

c.sc.OvY.comb<-c.sc.OvY.SS.3h%>%
  left_join(.,c.sc.OvY.SS.6h,by="sc.geneset",suffix=c(".SS3h",".SS6h"))%>%
  left_join(.,c.sc.OvY.SS.9h,by="sc.geneset")%>%
  left_join(.,c.sc.OvY.SS.12h,by="sc.geneset",suffix=c(".SS9h",".SS12h"))%>%
  left_join(.,c.sc.OvY.SD.3h,by="sc.geneset")%>%
  left_join(.,c.sc.OvY.SD.6h,by="sc.geneset",suffix=c(".SD3h",".SD6h"))%>%
  left_join(.,c.sc.OvY.SD.9h,by="sc.geneset")%>%
  left_join(.,c.sc.OvY.SD.12h,by="sc.geneset",suffix=c(".SD9h",".SD12h"))%>%
  left_join(.,c.sc.OvY.ZT0,by="sc.geneset")%>%
  mutate(Direction.SS3h=ifelse(Direction.SS3h=="Up",1,-1),Direction.SS6h=ifelse(Direction.SS6h=="Up",1,-1),Direction.SS9h=ifelse(Direction.SS9h=="Up",1,-1),Direction.SS12h=ifelse(Direction.SS12h=="Up",1,-1),Direction.SD3h=ifelse(Direction.SD3h=="Up",1,-1),Direction.SD6h=ifelse(Direction.SD6h=="Up",1,-1),Direction.SD9h=ifelse(Direction.SD9h=="Up",1,-1),Direction.SD12h=ifelse(Direction.SD12h=="Up",1,-1),Direction.ZT0=ifelse(Direction=="Up",1,-1))%>%
  mutate(dir.P.SS3h=-log10(PValue.SS3h)*Direction.SS3h,dir.P.SS6h=-log10(PValue.SS6h)*Direction.SS6h,dir.P.SS9h=-log10(PValue.SS9h)*Direction.SS9h,dir.P.SS12h=-log10(PValue.SS12h)*Direction.SS12h,dir.P.SD3h=-log10(PValue.SD3h)*Direction.SD3h,dir.P.SD6h=-log10(PValue.SD6h)*Direction.SD6h,dir.P.SD9h=-log10(PValue.SD9h)*Direction.SD9h,dir.P.SD12h=-log10(PValue.SD12h)*Direction.SD12h,dir.P.ZT0=-log10(PValue)*Direction.ZT0)%>%
  dplyr::select(sc.geneset,NGenes,starts_with("dir.P."))%>%
  mutate(sc.geneset=ifelse(grepl("Tasic2016_Endothelial",sc.geneset),"Tasic2016_Endothelial",ifelse(grepl("Hrvatin2018_Endothelial",sc.geneset),"Hrvatin2018_Endothelial",sc.geneset)))%>%
  mutate(sc.geneset.n=paste0(sc.geneset,"(",NGenes,")"))%>%
  dplyr::select(-sc.geneset,-NGenes)%>%
  mutate(celltype=ifelse(grepl("OPC",sc.geneset.n),"OPC",ifelse(grepl("OLs",sc.geneset.n),"OL",ifelse(grepl("Microglia",sc.geneset.n),"Microglia",ifelse(grepl("neuron",sc.geneset.n),"neuron",ifelse(grepl("Astrocytes",sc.geneset.n),"Astrocyte",ifelse(grepl("Endothelial",sc.geneset.n),"Endo",ifelse(grepl("Inhi",sc.geneset.n),"Inhibitory","Excitatory"))))))))%>%
  mutate(celltype=factor(celltype,levels=c("Microglia","OL","OPC","Astrocyte","Endo","neuron","Inhibitory","Excitatory")))%>%
  arrange(celltype)

annotation.col=data.frame(celltype=c.sc.OvY.comb$celltype)
rownames(annotation.col)=c.sc.OvY.comb$sc.geneset.n

my_palette.camera<-colorRampPalette(rev(brewer.pal(11,"RdBu")))(219)

colors.camera <- c(seq(-12,-5.1,length=20),seq(-5,-1.3,length=80),seq(-1.29,1.29,length=20),seq(1.3,5,length=80),seq(5.1,12,length=20))


png(filename = "scmarker.OvY.png",height = 2.5,width =5.5,res = 300,units = "in",type="cairo")
c.sc.OvY.comb%>%
  dplyr::select(sc.geneset.n,dir.P.ZT0,starts_with("dir.P.SS"))%>%
  column_to_rownames(.,var="sc.geneset.n")%>%
  t%>%
  pheatmap(., breaks = colors.camera, color=my_palette.camera, cluster_cols=FALSE,cluster_rows=FALSE,border_color= NA, annotation_col=annotation.col, gaps_col=12, labels_row=c("ZT0","ZT3","ZT6","ZT9","ZT12"),fontsize = 6.5,legend_breaks=c(-10,-5,-1.3,1.3,5,10),legend_labels =c("e-10","e-5","0.05","0.05","e-5","e-10"), main="Enrichment of cell-type specific gene sets in young and old comparison")

dev.off()



#gene set enrichment of the TF (transcription factor targets) and miRNA (MicroRNA) targets off C3 dataset


load("mouse_c3_v5p2.rdata")

C3.ind <-ids2indices(Mm.c3, v.norm$gene$ENTREZID)
C3.MiRNA <- grep("MIR-",names(C3.ind)) #221 C3 genes sets are miRNA targets


c.miRNA.SDvSS.Y.3h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,10],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.SDvSS.Y.6h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,11],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.SDvSS.Y.9h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,12],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.SDvSS.Y.12h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,13],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.SDvSS.O.3h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,14],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.SDvSS.O.6h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,15],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.SDvSS.O.9h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,16],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.SDvSS.O.12h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,17],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")


c.miRNA.SDvSS.comb<-c.miRNA.SDvSS.Y.3h%>%
  left_join(.,c.miRNA.SDvSS.Y.6h,by="miRNA.geneset",suffix=c(".Y3h",".Y6h"))%>%
  left_join(.,c.miRNA.SDvSS.Y.9h,by="miRNA.geneset")%>%
  left_join(.,c.miRNA.SDvSS.Y.12h,by="miRNA.geneset",suffix=c(".Y9h",".Y12h"))%>%
  left_join(.,c.miRNA.SDvSS.O.3h,by="miRNA.geneset")%>%
  left_join(.,c.miRNA.SDvSS.O.6h,by="miRNA.geneset",suffix=c(".O3h",".O6h"))%>%
  left_join(.,c.miRNA.SDvSS.O.9h,by="miRNA.geneset")%>%
  left_join(.,c.miRNA.SDvSS.O.12h,by="miRNA.geneset",suffix=c(".O9h",".O12h"))%>%
  mutate(Direction.Y3h=ifelse(Direction.Y3h=="Up",1,-1),Direction.Y6h=ifelse(Direction.Y6h=="Up",1,-1),Direction.Y9h=ifelse(Direction.Y9h=="Up",1,-1),Direction.Y12h=ifelse(Direction.Y12h=="Up",1,-1),Direction.O3h=ifelse(Direction.O3h=="Up",1,-1),Direction.O6h=ifelse(Direction.O6h=="Up",1,-1),Direction.O9h=ifelse(Direction.O9h=="Up",1,-1),Direction.O12h=ifelse(Direction.O12h=="Up",1,-1))%>%
  mutate(dir.P.Y3h=-log10(PValue.Y3h)*Direction.Y3h,dir.P.Y6h=-log10(PValue.Y6h)*Direction.Y6h,dir.P.Y9h=-log10(PValue.Y9h)*Direction.Y9h,dir.P.Y12h=-log10(PValue.Y12h)*Direction.Y12h,dir.P.O3h=-log10(PValue.O3h)*Direction.O3h,dir.P.O6h=-log10(PValue.O6h)*Direction.O6h,dir.P.O9h=-log10(PValue.O9h)*Direction.O9h,dir.P.O12h=-log10(PValue.O12h)*Direction.O12h)%>%
 filter(FDR.Y3h<0.05|FDR.Y6h<0.05|FDR.Y9h<0.05|FDR.Y12h<0.05|FDR.O3h<0.05|FDR.O6h<0.05|FDR.O9h<0.05|FDR.O12h<0.05)%>%
   dplyr::select(miRNA.geneset,NGenes.Y3h,starts_with("dir.P."))%>%
mutate(miRNA.geneset.n=paste0(miRNA.geneset,"(",NGenes.Y3h,")"))%>%
  dplyr::select(-miRNA.geneset,-NGenes.Y3h)%>%
  column_to_rownames(.,var="miRNA.geneset.n")



annotation.row=data.frame(comparison=rep(c("SDvSS.Y","SDvSS.O"),c(4,4)))
annotation.row$comparison=factor(annotation.row$comparison,levels=c("SDvSS.Y","SDvSS.O"))
rownames(annotation.row)=colnames(c.miRNA.SDvSS.comb)

my_palette.camera<-colorRampPalette(rev(brewer.pal(11,"RdBu")))(219)

colors.camera <- c(seq(-6,-3.1,length=20),seq(-3,-1.3,length=80),seq(-1.29,1.29,length=20),seq(1.3,3,length=80),seq(3.1,6,length=20))


png(filename = "miRNA.SDvSS.png",height = 3,width =5,res = 300,units = "in",type="cairo")

pheatmap(as.matrix(c.miRNA.SDvSS.comb), breaks = colors.camera, color=my_palette.camera, cluster_cols=FALSE,border_color= NA,annotation_col=annotation.row, treeheight_row=0,  labels_col=c("ZT3","ZT6","ZT9","ZT12","ZT3","ZT6","ZT9","ZT12"), fontsize = 8,gaps_col = 4, fontsize_col=10,fontsize_row=6,legend_breaks=c(-4,-1.3,1.3,4),legend_labels =c("p1e-4","p0.05","p0.05","p1e-4"), main="miRNA target gene set enrichment")

dev.off()


  

c.miRNA.OvY.SS.3h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,1],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.OvY.SS.6h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,2],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.OvY.SS.9h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,3],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.OvY.SS.12h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,4],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.OvY.SD.3h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,5],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.OvY.SD.6h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,6],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.OvY.SD.9h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,7],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")
c.miRNA.OvY.SD.12h<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,8],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")

c.miRNA.OvY.ZT0<-camera(v.norm,index=C3.ind[C3.MiRNA],design=design,contrast=contrasts.main[,9],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="miRNA.geneset")

c.miRNA.comb<-c.miRNA.OvY.SS.3h%>%
  left_join(.,c.miRNA.OvY.SS.6h,by="miRNA.geneset",suffix=c(".SS3h",".SS6h"))%>%
  left_join(.,c.miRNA.OvY.SS.9h,by="miRNA.geneset")%>%
  left_join(.,c.miRNA.OvY.SS.12h,by="miRNA.geneset",suffix=c(".SS9h",".SS12h"))%>%
  left_join(.,c.miRNA.OvY.SD.3h,by="miRNA.geneset")%>%
  left_join(.,c.miRNA.OvY.SD.6h,by="miRNA.geneset",suffix=c(".SD3h",".SD6h"))%>%
  left_join(.,c.miRNA.OvY.SD.9h,by="miRNA.geneset")%>%
  left_join(.,c.miRNA.OvY.SD.12h,by="miRNA.geneset",suffix=c(".SD9h",".SD12h"))%>%
  left_join(.,c.miRNA.OvY.ZT0,by="miRNA.geneset")%>%
  mutate(Direction.SS3h=ifelse(Direction.SS3h=="Up",1,-1),Direction.SS6h=ifelse(Direction.SS6h=="Up",1,-1),Direction.SS9h=ifelse(Direction.SS9h=="Up",1,-1),Direction.SS12h=ifelse(Direction.SS12h=="Up",1,-1),Direction.SD3h=ifelse(Direction.SD3h=="Up",1,-1),Direction.SD6h=ifelse(Direction.SD6h=="Up",1,-1),Direction.SD9h=ifelse(Direction.SD9h=="Up",1,-1),Direction.SD12h=ifelse(Direction.SD12h=="Up",1,-1),Direction.ZT0=ifelse(Direction=="Up",1,-1))%>%
  mutate(dir.P.SS3h=-log10(PValue.SS3h)*Direction.SS3h,dir.P.SS6h=-log10(PValue.SS6h)*Direction.SS6h,dir.P.SS9h=-log10(PValue.SS9h)*Direction.SS9h,dir.P.SS12h=-log10(PValue.SS12h)*Direction.SS12h,dir.P.SD3h=-log10(PValue.SD3h)*Direction.SD3h,dir.P.SD6h=-log10(PValue.SD6h)*Direction.SD6h,dir.P.SD9h=-log10(PValue.SD9h)*Direction.SD9h,dir.P.SD12h=-log10(PValue.SD12h)*Direction.SD12h,dir.P.ZT0=-log10(PValue)*Direction.ZT0)%>%
    full_join(.,c.miRNA.SDvSS.comb,by="miRNA.geneset")%>%
  filter(FDR.Y3h<0.05|FDR.Y6h<0.05|FDR.Y9h<0.05|FDR.Y12h<0.05|FDR.O3h<0.05|FDR.O6h<0.05|FDR.O9h<0.05|FDR.O12h<0.05|FDR.SS3h<0.05|FDR.SS6h<0.05|FDR.SS9h<0.05|FDR.SS12h<0.05|FDR<0.05)%>%
  dplyr::select(miRNA.geneset,NGenes,dir.P.ZT0,starts_with("dir.P.SS"),starts_with("dir.P.Y"),starts_with("dir.P.O"))%>%
 mutate(miRNA.geneset.n=paste0(miRNA.geneset,"(",NGenes,")"))%>%
  dplyr::select(-miRNA.geneset,-NGenes)%>%
  column_to_rownames(.,var="miRNA.geneset.n")


annotation.col.miRNA=data.frame(comparison=rep(c("OldvsYoung","SDvSS.Y","SDvSS.O"),c(5,4,4)))
annotation.col.miRNA$comparison=factor(annotation.col.miRNA$comparison,levels=c("OldvsYoung","SDvSS.Y","SDvSS.O"))
rownames(annotation.col.miRNA)=colnames(c.miRNA.comb)


library("pheatmap")
my_palette.camera<-colorRampPalette(rev(brewer.pal(11,"RdBu")))(219)

colors.camera.miRNA <- c(seq(min(c.miRNA.comb),-1.3,length=100),seq(-1.29,1.29,length=20),seq(1.3,max(c.miRNA.comb),length=100))

png(filename = "camera.miRNA.png",height = 6,width =8,res = 300,units = "in",type="cairo")

pheatmap(as.matrix(c.miRNA.comb), breaks = colors.camera.miRNA, color=my_palette.camera, cluster_cols=FALSE,border_color= NA,annotation_col=annotation.col.miRNA, labels_col=c("ZT0","ZT3","ZT6","ZT9","ZT12","ZT3","ZT6","ZT9","ZT12","ZT3","ZT6","ZT9","ZT12"),fontsize = 8,gaps_col = c(5,9), fontsize_col=8,fontsize_row=8,legend_breaks=c(-4,-1.3,1.3,4),legend_labels =c("p1e-4","p0.05","p0.05","p1e-4"), main="miRNA target gene set enrichment")

dev.off()




#transcription factor targets

c.TF.SDvSS.Y.3h<-camera(v.norm,index=C3.ind[-C3.MiRNA],design=design,contrast=contrasts.main[,10],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="TF.geneset")
c.TF.SDvSS.Y.6h<-camera(v.norm,index=C3.ind[-C3.MiRNA],design=design,contrast=contrasts.main[,11],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="TF.geneset")
c.TF.SDvSS.Y.9h<-camera(v.norm,index=C3.ind[-C3.MiRNA],design=design,contrast=contrasts.main[,12],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="TF.geneset")
c.TF.SDvSS.Y.12h<-camera(v.norm,index=C3.ind[-C3.MiRNA],design=design,contrast=contrasts.main[,13],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="TF.geneset")
c.TF.SDvSS.O.3h<-camera(v.norm,index=C3.ind[-C3.MiRNA],design=design,contrast=contrasts.main[,14],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="TF.geneset")
c.TF.SDvSS.O.6h<-camera(v.norm,index=C3.ind[-C3.MiRNA],design=design,contrast=contrasts.main[,15],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="TF.geneset")
c.TF.SDvSS.O.9h<-camera(v.norm,index=C3.ind[-C3.MiRNA],design=design,contrast=contrasts.main[,16],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="TF.geneset")
c.TF.SDvSS.O.12h<-camera(v.norm,index=C3.ind[-C3.MiRNA],design=design,contrast=contrasts.main[,17],inter.gene.cor=0.01)%>%
  rownames_to_column(.,var="TF.geneset")

c.TF.SDvSS.comb<-c.TF.SDvSS.Y.3h%>%
  left_join(.,c.TF.SDvSS.Y.6h,by="TF.geneset",suffix=c(".Y3h",".Y6h"))%>%
  left_join(.,c.TF.SDvSS.Y.9h,by="TF.geneset")%>%
  left_join(.,c.TF.SDvSS.Y.12h,by="TF.geneset",suffix=c(".Y9h",".Y12h"))%>%
  left_join(.,c.TF.SDvSS.O.3h,by="TF.geneset")%>%
  left_join(.,c.TF.SDvSS.O.6h,by="TF.geneset",suffix=c(".O3h",".O6h"))%>%
  left_join(.,c.TF.SDvSS.O.9h,by="TF.geneset")%>%
  left_join(.,c.TF.SDvSS.O.12h,by="TF.geneset",suffix=c(".O9h",".O12h"))%>%
  mutate(Direction.Y3h=ifelse(Direction.Y3h=="Up",1,-1),Direction.Y6h=ifelse(Direction.Y6h=="Up",1,-1),Direction.Y9h=ifelse(Direction.Y9h=="Up",1,-1),Direction.Y12h=ifelse(Direction.Y12h=="Up",1,-1),Direction.O3h=ifelse(Direction.O3h=="Up",1,-1),Direction.O6h=ifelse(Direction.O6h=="Up",1,-1),Direction.O9h=ifelse(Direction.O9h=="Up",1,-1),Direction.O12h=ifelse(Direction.O12h=="Up",1,-1))%>%
  mutate(dir.P.Y3h=-log10(PValue.Y3h)*Direction.Y3h,dir.P.Y6h=-log10(PValue.Y6h)*Direction.Y6h,dir.P.Y9h=-log10(PValue.Y9h)*Direction.Y9h,dir.P.Y12h=-log10(PValue.Y12h)*Direction.Y12h,dir.P.O3h=-log10(PValue.O3h)*Direction.O3h,dir.P.O6h=-log10(PValue.O6h)*Direction.O6h,dir.P.O9h=-log10(PValue.O9h)*Direction.O9h,dir.P.O12h=-log10(PValue.O12h)*Direction.O12h)%>%
  dplyr::select(TF.geneset,NGenes.Y3h,starts_with("dir.P."),starts_with("PValue"))%>%
 filter(FDR.Y3h<0.05|FDR.Y6h<0.05|FDR.Y9h<0.05|FDR.Y12h<0.05|FDR.O3h<0.05|FDR.O6h<0.05|FDR.O9h<0.05|FDR.O12h<0.05)%>%
 mutate(TF.geneset.n=paste0(TF.geneset,"(",NGenes.Y3h,")"))%>%
  dplyr::select(-TF.geneset,-NGenes.Y3h)%>%


my_palette.TF<-colorRampPalette(brewer.pal(9,"Reds"))(99)

colors.camera.TF <- c(seq(0,1.3,length=10),seq(1.31,4.5,length=90))

png(filename = "camera.TF.png",height = 3,width =6,res = 300,units = "in",type="cairo")
c.TF.SDvSS.comb%>%
  dplyr::select(TF.geneset.n,starts_with("dir"))%>%
    column_to_rownames(.,var="TF.geneset.n")%>%
  as.matrix%>%
pheatmap(., breaks = colors.camera.TF, color=my_palette.TF, cluster_cols=FALSE,border_color= NA,annotation_col=annotation.col.miRNA, labels_col=c("ZT3","ZT6","ZT9","ZT12","ZT3","ZT6","ZT9","ZT12"),fontsize = 9, fontsize_row=8,fontsize_col = 12, gaps_col = 4, legend_breaks=c(0,1.3,4), treeheight_row=0, legend_labels =c("0","p0.05","E-4"), main="TF target gene set enrichment")

dev.off()
```





```{r}
#plot expressions of selected genes

trans.factor<-c("ENSMUSG00000028163","ENSMUSG00000036931",                "ENSMUSG00000021250", "ENSMUSG00000048482", "ENSMUSG00000022602", "ENSMUSG00000007617","ENSMUSG00000052837")

  
trans.factor.plot<-cbind(id=v.norm$genes$id,as.data.frame(v.norm$E))%>%
  filter(id%in%trans.factor)%>%
  left_join(.,v.norm$genes,by="id")%>%
  rename(.,Young.SS.Z0.B2=Young.Z0.66902.B2,Young.SS.Z0.G4=Young.Z0.66919.G4,Young.SS.Z0.D8=Young.Z0.66938.D8,Young.SS.Z0.F4=Young.Z0.66972.F4,Young.SS.Z0.E7=Young.Z0.66988.E7,Young.SS.Z0.E11=Young.Z0.67008.E11,Old.SS.Z0.D2=Old.Z0.66904.D2,Old.SS.Z0.C5=Old.Z0.66921.C5,Old.SS.Z0.F8=Old.Z0.66940.F8,Old.SS.Z0.B5=Old.Z0.66974.B5,Old.SS.Z0.B8=Old.Z0.66990.B8,Old.SS.Z0.F11=Old.Z0.67009.F11)%>%
  mutate(.,Young.SD.Z0.B2=Young.SS.Z0.B2,Young.SD.Z0.G4=Young.SS.Z0.G4,Young.SD.Z0.D8=Young.SS.Z0.D8,Young.SD.Z0.F4=Young.SS.Z0.F4,Young.SD.Z0.E7=Young.SS.Z0.E7,Young.SD.Z0.E11=Young.SS.Z0.E11,Old.SD.Z0.D2=Old.SS.Z0.D2,Old.SD.Z0.C5=Old.SS.Z0.C5,Old.SD.Z0.F8=Old.SS.Z0.F8,Old.SD.Z0.B5=Old.SS.Z0.B5,Old.SD.Z0.B8=Old.SS.Z0.B8,Old.SD.Z0.F11=Old.SS.Z0.F11)%>%
  tibble::column_to_rownames(var="geneSymbol")%>%
  dplyr::select(-id,-starts_with("gene"),-ENTREZID,-description,-transcript_length)%>%
  t(.)%>%
  as.data.frame(.)%>%
  tibble::rownames_to_column(.,var="sampleID")%>%
  mutate(age=ifelse(grepl("Young",sampleID),"Y","O"), sleep=ifelse(grepl("SS",sampleID),"SS","SD"),
         time=ifelse(grepl("Z0",sampleID),"0",ifelse(grepl("3hr",sampleID),"3",ifelse(grepl("6hr",sampleID),"6",ifelse(grepl("9hr",sampleID),"9","12")))))%>%
  mutate(age1=age,sleep1=sleep)%>%
    unite(age.sleep,age1,sleep1,sep = ".")%>%
  mutate(age.sleep=factor(age.sleep,levels=c("Y.SS","O.SS","Y.SD","O.SD")),time=factor(time,levels=c("0","3","6","9","12")),age=factor(age,levels=c("Y","O")),sleep=factor(sleep,levels=c("SS","SD")))%>%
  gather(.,key=gene,value=logCPM,Homer1:Junb)%>%
  mutate(gene=factor(gene,levels=unique(gene)))
  
line.palette<-brewer.pal(11, "RdYlBu")[c(2,3,10,9)]
     
exp.plot<-function(i,df){plot<-df%>%
  filter(gene==levels(df$gene)[i])%>%
  ggline(.,x="time",y="logCPM",color="age.sleep",palette =line.palette, linetype="age.sleep",add =c("mean_se","jitter"),size=0.8,plot_type="l",xlab ="",ylab = "",title=levels(df$gene)[i])+scale_linetype_manual(values=c("solid","dashed","solid","dashed"))+theme(legend.title=element_blank(),legend.position="right",plot.title = element_text(hjust = 0.5))
  return(plot)}


TF.pp<-ggarrange(exp.plot(1,trans.factor.plot),exp.plot(2,trans.factor.plot),exp.plot(3,trans.factor.plot),exp.plot(4,trans.factor.plot),exp.plot(5,trans.factor.plot),exp.plot(6,trans.factor.plot),exp.plot(7,trans.factor.plot),ncol = 3, nrow =2,common.legend = TRUE, legend = "right")

png(filename = "selectTF.png",height = 9,width =12,res = 300,units = "in",type="cairo")
annotate_figure(TF.pp,top = text_grob("Selective Transcription factor", color = "black", face = "bold", size = 14),bottom = text_grob("Time", color = "black",hjust = 0.5, face = "bold", size = 12),left = text_grob("logCPM", color = "black", face = "bold",rot = 90, size = 12),fig.lab="11",fig.lab.size=14,fig.lab.face="bold")
dev.off()

```